<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>1&nbsp; Species tree reconstruction from organellar genomes – Development of taxonomical biomarkers for Ulva genus</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./cp-and-mt-phylogeny-reconstruction.html" rel="next">
<link href="./index.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<link href="site_libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="site_libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="site_libs/datatables-binding-0.33/datatables.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="site_libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="site_libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="site_libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script>
<link href="site_libs/dt-ext-fixedcolumns-1.13.6/css/fixedColumns.dataTables.min.css" rel="stylesheet">
<script src="site_libs/dt-ext-fixedcolumns-1.13.6/js/dataTables.fixedColumns.min.js"></script>
<link href="site_libs/dt-ext-fixedheader-1.13.6/css/fixedHeader.dataTables.min.css" rel="stylesheet">
<script src="site_libs/dt-ext-fixedheader-1.13.6/js/dataTables.fixedHeader.min.js"></script>
<link href="site_libs/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="site_libs/crosstalk-1.2.1/js/crosstalk.min.js"></script>
<style>
#quarto-sidebar .sidebar-item-text {
  font-weight: bold;
}
#quarto-sidebar .chapter-number {
  font-weight: normal
}
#quarto-sidebar .chapter-title {
  font-weight: normal
}
#.nav-tabs .nav-link.active {
  background-color: rgba(239, 236, 255, 0.5);
}
</style>


</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Development of taxonomical biomarkers for Ulva genus</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./cp-and-mt-dataset-overview.html"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Species tree reconstruction from organellar genomes</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Abstract</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cp-and-mt-dataset-overview.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Species tree reconstruction from organellar genomes</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cp-and-mt-phylogeny-reconstruction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Species tree reconstruction from organellar genomes</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./markers-selection.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data Exploration - Kidney Cancer</span></span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#on-this-page" id="toc-on-this-page" class="nav-link active" data-scroll-target="#on-this-page"><span class="header-section-number">1.1</span> On this page</a></li>
  <li><a href="#dataset-overview" id="toc-dataset-overview" class="nav-link" data-scroll-target="#dataset-overview"><span class="header-section-number">1.2</span> Dataset overview</a>
  <ul>
  <li><a href="#dataset-retrieval" id="toc-dataset-retrieval" class="nav-link" data-scroll-target="#dataset-retrieval"><span class="header-section-number">1.2.1</span> Dataset retrieval</a></li>
  <li><a href="#gene-occupancy" id="toc-gene-occupancy" class="nav-link" data-scroll-target="#gene-occupancy"><span class="header-section-number">1.2.2</span> Gene occupancy</a></li>
  <li><a href="#gene-alignments-distributions" id="toc-gene-alignments-distributions" class="nav-link" data-scroll-target="#gene-alignments-distributions"><span class="header-section-number">1.2.3</span> Gene alignments distributions</a></li>
  </ul></li>
  <li><a href="#reconstruction-of-ulva-species-tree" id="toc-reconstruction-of-ulva-species-tree" class="nav-link" data-scroll-target="#reconstruction-of-ulva-species-tree"><span class="header-section-number">1.3</span> Reconstruction of Ulva species tree</a></li>
  <li><a href="#sec-lesson-learnt-cp-and-mt-dataset-overview" id="toc-sec-lesson-learnt-cp-and-mt-dataset-overview" class="nav-link" data-scroll-target="#sec-lesson-learnt-cp-and-mt-dataset-overview"><span class="header-section-number">1.4</span> Lessons Learnt</a></li>
  <li><a href="#session-information" id="toc-session-information" class="nav-link" data-scroll-target="#session-information"><span class="header-section-number">1.5</span> Session Information</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Species tree reconstruction from organellar genomes</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="on-this-page" class="level2" data-number="1.1">
<h2 data-number="1.1" class="anchored" data-anchor-id="on-this-page"><span class="header-section-number">1.1</span> On this page</h2>
<p>Biological insights and take-home messages are at the bottom of the page at Lesson Learnt: <a href="#sec-lesson-learnt-cp-and-mt-dataset-overview" class="quarto-xref"><span>Section 1.4</span></a>.</p>
<ul>
<li>Here we retrieve Ulva organellar genomes</li>
<li>we investigate the</li>
</ul>
</section>
<section id="dataset-overview" class="level2" data-number="1.2">
<h2 data-number="1.2" class="anchored" data-anchor-id="dataset-overview"><span class="header-section-number">1.2</span> Dataset overview</h2>
<section id="dataset-retrieval" class="level3" data-number="1.2.1">
<h3 data-number="1.2.1" class="anchored" data-anchor-id="dataset-retrieval"><span class="header-section-number">1.2.1</span> Dataset retrieval</h3>
<p>We included the following species, that as for 2022 where the most comprehensive list of cp and mt complete genomes published for Ulva.</p>
<p>Oltmassiellopsis viridi and Pseudendoclonium akinetum were used as an outgroup to root the phylogenetic three.</p>
<div class="cell">
<div class="cell-output-display">
<div class="datatables html-widget html-fill-item" id="htmlwidget-3f50ec31c773ba27c7ac" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-3f50ec31c773ba27c7ac">{"x":{"filter":"none","vertical":false,"extensions":["FixedColumns","FixedHeader"],"caption":"<caption>Table 1: Species used in the study<\/caption>","data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52"],["Oviri","Pakin","Uaust1","Uaust2","Uaust3","Uaust4","Uaust5","Uaust6","Ucali","Ucomp1","Ucomp10","Ucomp2","Ucomp3","Ucomp4","Ucomp5","Ucomp6","Ucomp7","Ucomp8","Ucomp9","Ufene","Uflex1","Uflex2","Uflex3","Ugiga","Uinte","Ulaci1","Ulaci2","Ulact1","Ulact2","Ulact3","Ulinz","Umeri1","Umeri2","Umeri3","Umeri4","Umeri5","Umeri6","Umeri7","Uohno","Uprol1","Uprol2","Uprol3","Uprol4","Uprol5","Urigi1","Urigi2","Urigi3","Usp1","Usp2","Usp3","Usp4","Utort"],["Oltmannsiellopsis viridis","Pseudendoclonium akinetum","Ulva australis","Ulva australis","Ulva australis","Ulva australis","Ulva australis","Ulva australis","Ulva californica","Ulva compressa","Ulva compressa","Ulva compressa","Ulva compressa","Ulva compressa","Ulva compressa","Ulva compressa","Ulva compressa","Ulva compressa","Ulva compressa","Ulva fenestrata","Ulva flexuosa","Ulva flexuosa","Ulva flexuosa","Ulva gigantea","Ulva intestinalis","Ulva lacinulata","Ulva lacinulata","Ulva lactuca","Ulva lactuca","Ulva lactuca","Ulva linza","Ulva medirionalis","Ulva medirionalis","Ulva medirionalis","Ulva medirionalis","Ulva medirionalis","Ulva meridionalis","Ulva meridionalis","Ulva ohnoi","Ulva prolifera","Ulva prolifera","Ulva prolifera","Ulva prolifera","Ulva prolifera","Ulva rigida","Ulva rigida","Ulva rigida","Ulva sp. AF-2021","Ulva sp. Q253","Ulva sp. TM637","Ulva sp. TM708","Ulva torta"],[51324,160070,111616,111616,111616,111616,111616,111616,83794,63659,63659,63659,63659,63659,63659,63659,63659,63659,63659,83795,83791,83791,83791,523338,3116,2806530,2806530,63410,63410,63410,63409,434723,434723,434723,434723,434723,434723,434723,240864,3117,3117,3117,3117,3117,75689,75689,75689,2847172,2608636,2496872,2496873,932731],["","","KU-1658","U100","CNS00533","","Up1","Up2","U484-3","","","","Ucom2","Ucom3","UNA00072687","","","","","U64","","Ufle2","","U57","CNS00531","L0054997","U35","","","L0054996","QD08","LF008","LF010","LF011","LF012","LF018","SHOU2019QD082001","UNA00071828","KU-3321","","U161","","","","HCC102617","LD 14294","U112","AF-2021","Q253","TM637","TM708",""],["NC_008099","NC_008114","LC507117","NC_053611","MW531676","MN853875","","","MZ561475","MK069585","data from Frederik","MT916929","MW344287","MW353781","MW548841","NC_050739","","","NC_043860","NC_053612","NC_035823","","","NC_053613","MZ158703","MW543061","NC_053614","NC_029040","","NC_042255","NC_030312","","","","","","MN889540","KP720616","AP018696","MN853879","MZ571508","NC_036137","","","MN389525","MW543060","NC_053616","NC_053615","MW699788","","","MZ703011"],["NC_008256","NC_005926","","MT179354","","","NC_035722","KX530817","","MK069586","data from Frederik","","MH013469","","","KX595276","MH093740","NC_041082","MK069587","MT179355","NC_035809","MH013470","KY626326","MT179356","MZ571476","","NC_053631","NC_028081","KU182748","MH763013","NC_029701","ON402240","NC_065202","ON402237","ON402238","ON402239","MN861072","KP720617","AP018695","","","KU161104","NC_028538","MN853878","MN389526","","NC_053633","NC_053632","","NC_053253","NC_053254","NC_053255"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>label<\/th>\n      <th>species<\/th>\n      <th>NCBI.taxid<\/th>\n      <th>isolate<\/th>\n      <th>cp.accenssion.number<\/th>\n      <th>mt.accession.number<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":false,"pageLength":10,"columnDefs":[{"className":"dt-right","targets":3},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"label","targets":1},{"name":"species","targets":2},{"name":"NCBI.taxid","targets":3},{"name":"isolate","targets":4},{"name":"cp.accenssion.number","targets":5},{"name":"mt.accession.number","targets":6}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<p>Organellar genome sequences and gene models were retrieved from NCBI, and corrected where necessary.</p>
</section>
<section id="gene-occupancy" class="level3" data-number="1.2.2">
<h3 data-number="1.2.2" class="anchored" data-anchor-id="gene-occupancy"><span class="header-section-number">1.2.2</span> Gene occupancy</h3>
<p>The gene matrix occupancy and gene length distribution across organellar genomes included in this study are presented below. Most selected genes are present/ nearly present in all studied sequences. In 29 selected genes for mitochondrial, there are two genes, rps3 and rps4 that are not presented in more than 10% of the 37 mt genomic sequences. For chloroplast, there are 10 out of 70 genes not presented in more than 10% of the 34 cp genomic sequences.</p>
<div class="cell" data-result="hide">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="co"># read tables</span></span>
<span id="cb1-2"><a href="#cb1-2"></a>cp_mt_genes_occupancy <span class="ot">=</span> <span class="fu">read.delim</span>(<span class="st">"./data/cp_mt_genes_occupancy.tab"</span>)</span>
<span id="cb1-3"><a href="#cb1-3"></a></span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="co"># drop last line which is uninformative</span></span>
<span id="cb1-5"><a href="#cb1-5"></a>cp_mt_genes_occupancy <span class="ot">=</span> <span class="fu">head</span>(cp_mt_genes_occupancy, <span class="sc">-</span><span class="dv">1</span>)</span>
<span id="cb1-6"><a href="#cb1-6"></a></span>
<span id="cb1-7"><a href="#cb1-7"></a><span class="co"># format headers</span></span>
<span id="cb1-8"><a href="#cb1-8"></a><span class="fu">colnames</span>(cp_mt_genes_occupancy) <span class="ot">=</span> stringr<span class="sc">::</span><span class="fu">str_remove_all</span>(<span class="fu">colnames</span>(cp_mt_genes_occupancy), <span class="st">".aln.fa"</span>)</span>
<span id="cb1-9"><a href="#cb1-9"></a><span class="fu">colnames</span>(cp_mt_genes_occupancy) <span class="ot">=</span> stringr<span class="sc">::</span><span class="fu">str_replace_all</span>(<span class="fu">colnames</span>(cp_mt_genes_occupancy), <span class="st">"X..Sequences"</span>, <span class="st">"Species"</span>)</span>
<span id="cb1-10"><a href="#cb1-10"></a></span>
<span id="cb1-11"><a href="#cb1-11"></a><span class="co"># melt and clean</span></span>
<span id="cb1-12"><a href="#cb1-12"></a>cp_mt_genes_occupancy <span class="ot">=</span> reshape2<span class="sc">::</span><span class="fu">melt</span>(cp_mt_genes_occupancy)</span>
<span id="cb1-13"><a href="#cb1-13"></a>cp_mt_genes_occupancy<span class="sc">$</span>organell <span class="ot">=</span> <span class="fu">ifelse</span>(</span>
<span id="cb1-14"><a href="#cb1-14"></a>  stringr<span class="sc">::</span><span class="fu">str_detect</span>(cp_mt_genes_occupancy<span class="sc">$</span>variable, <span class="st">"_cp_CDS_"</span>),</span>
<span id="cb1-15"><a href="#cb1-15"></a>  <span class="st">"Chloroplast"</span>,</span>
<span id="cb1-16"><a href="#cb1-16"></a>  <span class="st">"Mitochondrion"</span></span>
<span id="cb1-17"><a href="#cb1-17"></a>)</span>
<span id="cb1-18"><a href="#cb1-18"></a>cp_mt_genes_occupancy<span class="sc">$</span>variable <span class="ot">=</span> stringr<span class="sc">::</span><span class="fu">str_remove_all</span>(cp_mt_genes_occupancy<span class="sc">$</span>variable, <span class="st">"X02_cp_CDS_align."</span>)</span>
<span id="cb1-19"><a href="#cb1-19"></a>cp_mt_genes_occupancy<span class="sc">$</span>variable <span class="ot">=</span> stringr<span class="sc">::</span><span class="fu">str_remove_all</span>(cp_mt_genes_occupancy<span class="sc">$</span>variable, <span class="st">"X02_mt_CDS_align."</span>)</span>
<span id="cb1-20"><a href="#cb1-20"></a><span class="co"># set different number to assign different color to mitochondrial genes</span></span>
<span id="cb1-21"><a href="#cb1-21"></a>cp_mt_genes_occupancy<span class="sc">$</span>value <span class="ot">=</span> <span class="fu">ifelse</span>(</span>
<span id="cb1-22"><a href="#cb1-22"></a>  cp_mt_genes_occupancy<span class="sc">$</span>value <span class="sc">==</span> <span class="dv">1</span> <span class="sc">&amp;</span> cp_mt_genes_occupancy<span class="sc">$</span>organell <span class="sc">==</span> <span class="st">"Mitochondrion"</span>,</span>
<span id="cb1-23"><a href="#cb1-23"></a>  <span class="dv">2</span>,</span>
<span id="cb1-24"><a href="#cb1-24"></a>  cp_mt_genes_occupancy<span class="sc">$</span>value</span>
<span id="cb1-25"><a href="#cb1-25"></a>)</span>
<span id="cb1-26"><a href="#cb1-26"></a>cp_mt_genes_occupancy<span class="sc">$</span>value <span class="ot">=</span> <span class="fu">as.character</span>(cp_mt_genes_occupancy<span class="sc">$</span>value)</span>
<span id="cb1-27"><a href="#cb1-27"></a></span>
<span id="cb1-28"><a href="#cb1-28"></a><span class="co"># plot</span></span>
<span id="cb1-29"><a href="#cb1-29"></a>p1 <span class="ot">=</span> <span class="fu">ggplot</span>(cp_mt_genes_occupancy) <span class="sc">+</span></span>
<span id="cb1-30"><a href="#cb1-30"></a>  <span class="fu">geom_tile</span>(<span class="fu">aes</span>(<span class="at">x =</span> variable, <span class="at">y =</span> Species, <span class="at">fill =</span> value), <span class="at">color =</span> <span class="st">"grey75"</span>) <span class="sc">+</span></span>
<span id="cb1-31"><a href="#cb1-31"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"#469d89"</span>, <span class="st">"steelblue"</span>)) <span class="sc">+</span></span>
<span id="cb1-32"><a href="#cb1-32"></a>  <span class="fu">facet_grid</span>(<span class="sc">~</span>organell, <span class="at">scales =</span> <span class="st">"free"</span>, <span class="at">space =</span> <span class="st">"free_x"</span>) <span class="sc">+</span></span>
<span id="cb1-33"><a href="#cb1-33"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Genes"</span>,</span>
<span id="cb1-34"><a href="#cb1-34"></a>       <span class="at">y =</span> <span class="st">"Species"</span>) <span class="sc">+</span></span>
<span id="cb1-35"><a href="#cb1-35"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">24</span>, <span class="at">hjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb1-36"><a href="#cb1-36"></a>        <span class="at">axis.ticks.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb1-37"><a href="#cb1-37"></a>        <span class="at">axis.ticks.y =</span> <span class="fu">element_blank</span>(), </span>
<span id="cb1-38"><a href="#cb1-38"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">hjust =</span> <span class="dv">1</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb1-39"><a href="#cb1-39"></a>        <span class="at">legend.position =</span> <span class="st">"none"</span>,</span>
<span id="cb1-40"><a href="#cb1-40"></a>        <span class="at">panel.background =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb1-41"><a href="#cb1-41"></a>        <span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">colour =</span> <span class="st">"black"</span>, <span class="at">fill =</span> <span class="cn">NA</span>),</span>
<span id="cb1-42"><a href="#cb1-42"></a>        <span class="at">strip.background =</span> <span class="fu">element_rect</span>(<span class="at">colour =</span> <span class="st">"NA"</span>, <span class="at">fill =</span> <span class="st">"NA"</span>),</span>
<span id="cb1-43"><a href="#cb1-43"></a>        <span class="at">strip.placement =</span> <span class="st">"outside"</span>,</span>
<span id="cb1-44"><a href="#cb1-44"></a>        <span class="at">strip.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">10</span>, <span class="at">angle =</span> <span class="dv">0</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust =</span> <span class="fl">0.5</span>))</span>
<span id="cb1-45"><a href="#cb1-45"></a><span class="fu">plot</span>(p1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="cp-and-mt-dataset-overview_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption>Figure 1: Genes occupancy</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="gene-alignments-distributions" class="level3" data-number="1.2.3">
<h3 data-number="1.2.3" class="anchored" data-anchor-id="gene-alignments-distributions"><span class="header-section-number">1.2.3</span> Gene alignments distributions</h3>
<p>Next, we retrieved the CDS sequences and we align them. To do so, first we extract all CDSs that are longer than 200 bp from the .fasta .gff files pairs for each organellar genome.</p>
<div class="cell" data-result="hide">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="co"># get CP genes list</span></span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="fu">cut</span> <span class="at">-f</span> 9 00_cp_GenBank_RAW/<span class="pp">*</span>.gff <span class="dt">\</span></span>
<span id="cb2-3"><a href="#cb2-3"></a>  <span class="kw">|</span> <span class="fu">sed</span> <span class="st">"s/Name=//g"</span> <span class="dt">\</span></span>
<span id="cb2-4"><a href="#cb2-4"></a>  <span class="kw">|</span> <span class="fu">sed</span> <span class="st">"s/ .*//g"</span> <span class="dt">\</span></span>
<span id="cb2-5"><a href="#cb2-5"></a>  <span class="kw">|</span> <span class="fu">grep</span> <span class="st">"accD\|atp\|ccsA\|cemA\|chlI\|clpP\|ftsH\|infA\|pafI\|pbf1\|pet\|psa\|psb30\|psb\|rbcL\|rnl\|rns\|rpl\|rpo\|rps\|rrn5\|tufA\|ycf"</span> <span class="dt">\</span></span>
<span id="cb2-6"><a href="#cb2-6"></a>  <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-u</span> <span class="dt">\</span></span>
<span id="cb2-7"><a href="#cb2-7"></a>  <span class="op">&gt;</span> cp_gene_list.lst</span>
<span id="cb2-8"><a href="#cb2-8"></a></span>
<span id="cb2-9"><a href="#cb2-9"></a><span class="co"># extract CP fasta sequences</span></span>
<span id="cb2-10"><a href="#cb2-10"></a><span class="cf">while</span> <span class="bu">read</span> <span class="va">GENE</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb2-11"><a href="#cb2-11"></a>    <span class="cf">for</span> file <span class="kw">in</span> 00_cp_GenBank_RAW/<span class="pp">*[</span><span class="ss">0</span><span class="pp">-</span><span class="ss">9</span><span class="pp">]</span>.fasta<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb2-12"><a href="#cb2-12"></a>        <span class="fu">grep</span> <span class="va">$GENE</span> <span class="va">$(</span><span class="fu">basename</span> <span class="va">$file</span> .fasta<span class="va">)</span>.cds.gff <span class="op">&gt;</span> tmp.gff<span class="kw">;</span></span>
<span id="cb2-13"><a href="#cb2-13"></a>        <span class="ex">bedtools</span> getfasta <span class="at">-fi</span> <span class="va">$file</span> <span class="at">-bed</span> tmp.gff <span class="kw">|</span> <span class="fu">tr</span> <span class="st">":"</span> <span class="st">" "</span><span class="kw">;</span></span>
<span id="cb2-14"><a href="#cb2-14"></a>    <span class="cf">done</span> <span class="op">&gt;</span> 01_cp_CDS/<span class="va">$GENE</span>.fa<span class="kw">;</span></span>
<span id="cb2-15"><a href="#cb2-15"></a>    <span class="fu">rm</span> tmp.gff<span class="kw">;</span></span>
<span id="cb2-16"><a href="#cb2-16"></a><span class="cf">done</span> <span class="op">&lt;</span> cp_gene_list.lst</span>
<span id="cb2-17"><a href="#cb2-17"></a></span>
<span id="cb2-18"><a href="#cb2-18"></a></span>
<span id="cb2-19"><a href="#cb2-19"></a><span class="co"># get MT genes list</span></span>
<span id="cb2-20"><a href="#cb2-20"></a><span class="fu">cut</span> <span class="at">-f</span> 9 00_mt_GenBank_RAW/<span class="pp">*</span>.gff <span class="dt">\</span></span>
<span id="cb2-21"><a href="#cb2-21"></a>  <span class="kw">|</span> <span class="fu">sed</span> <span class="st">"s/Name=//g"</span> <span class="dt">\</span></span>
<span id="cb2-22"><a href="#cb2-22"></a>  <span class="kw">|</span> <span class="fu">sed</span> <span class="st">"s/ .*//g"</span> <span class="dt">\</span></span>
<span id="cb2-23"><a href="#cb2-23"></a>  <span class="kw">|</span> <span class="fu">sed</span> <span class="st">"s/ATP/atp/g"</span> <span class="dt">\</span></span>
<span id="cb2-24"><a href="#cb2-24"></a>  <span class="kw">|</span> <span class="fu">sed</span> <span class="st">"s/NAD/nad/"</span>  <span class="dt">\</span></span>
<span id="cb2-25"><a href="#cb2-25"></a>  <span class="kw">|</span> <span class="fu">grep</span> <span class="st">"atp\|cob\|cox\|nad\|rpl\|rps"</span> <span class="dt">\</span></span>
<span id="cb2-26"><a href="#cb2-26"></a>  <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-u</span> <span class="dt">\</span></span>
<span id="cb2-27"><a href="#cb2-27"></a>  <span class="op">&gt;</span> mt_gene_list.lst</span>
<span id="cb2-28"><a href="#cb2-28"></a></span>
<span id="cb2-29"><a href="#cb2-29"></a><span class="co"># extract MT fasta sequences</span></span>
<span id="cb2-30"><a href="#cb2-30"></a><span class="cf">while</span> <span class="bu">read</span> <span class="va">GENE</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb2-31"><a href="#cb2-31"></a>    <span class="cf">for</span> file <span class="kw">in</span> 00_mt_GenBank_RAW/<span class="pp">*[</span><span class="ss">0</span><span class="pp">-</span><span class="ss">9</span><span class="pp">]</span>.fasta<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb2-32"><a href="#cb2-32"></a>        <span class="fu">grep</span> <span class="va">$GENE</span> <span class="va">$(</span><span class="fu">basename</span> <span class="va">$file</span> .fasta<span class="va">)</span>.cds.gff <span class="op">&gt;</span> tmp.gff<span class="kw">;</span></span>
<span id="cb2-33"><a href="#cb2-33"></a>        <span class="ex">bedtools</span> getfasta <span class="at">-fi</span> <span class="va">$file</span> <span class="at">-bed</span> tmp.gff <span class="kw">|</span> <span class="fu">tr</span> <span class="st">":"</span> <span class="st">" "</span><span class="kw">;</span></span>
<span id="cb2-34"><a href="#cb2-34"></a>    <span class="cf">done</span> <span class="op">&gt;</span> 01_mt_CDS/<span class="va">$GENE</span>.fa<span class="kw">;</span></span>
<span id="cb2-35"><a href="#cb2-35"></a>    <span class="fu">rm</span> tmp.gff<span class="kw">;</span></span>
<span id="cb2-36"><a href="#cb2-36"></a><span class="cf">done</span> <span class="op">&lt;</span> mt_gene_list.lst</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We will use <a href="https://open.bioqueue.org/home/knowledge/showKnowledge/sig/ucsc-fatrans">faTrans</a> to translate the nucleotidic sequences to the corresponding amino-acid sequences, <a href="https://academic.oup.com/bib/article/20/4/1160/4106928">MAFFT</a> to align the amino-acid sequences for each gene, and finally <a href="https://academic.oup.com/nar/article/34/suppl_2/W609/2505720">pal2nal</a> to translate the alignemts back at the nucleotidic level. We will do this both for the chloroplast and for the mitochondrial genes.</p>
<div class="cell" data-result="hid">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1"></a><span class="co"># translate CP to amino-acid sequences</span></span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="cf">for</span> file <span class="kw">in</span> 01_cp_CDS/<span class="pp">*</span>.fa<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb3-3"><a href="#cb3-3"></a>  <span class="ex">~/bin/faTrans</span> <span class="va">$file</span> 02_cp_CDS_align/<span class="va">$(</span><span class="fu">basename</span> <span class="va">$file</span> .fa<span class="va">)</span>.aa.fa<span class="kw">;</span></span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="cf">done</span></span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="co"># translate MT to amino-acid sequences</span></span>
<span id="cb3-6"><a href="#cb3-6"></a><span class="cf">for</span> file <span class="kw">in</span> 01_mt_CDS/<span class="pp">*</span>.fa<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb3-7"><a href="#cb3-7"></a>  <span class="ex">~/bin/faTrans</span> <span class="va">$file</span> 02_mt_CDS_align/<span class="va">$(</span><span class="fu">basename</span> <span class="va">$file</span> .fa<span class="va">)</span>.aa.fa<span class="kw">;</span></span>
<span id="cb3-8"><a href="#cb3-8"></a><span class="cf">done</span></span>
<span id="cb3-9"><a href="#cb3-9"></a></span>
<span id="cb3-10"><a href="#cb3-10"></a><span class="co"># align amino-acid sequences</span></span>
<span id="cb3-11"><a href="#cb3-11"></a><span class="fu">ls</span> ./02_cp_CDS_align/ <span class="dt">\</span></span>
<span id="cb3-12"><a href="#cb3-12"></a>  <span class="kw">|</span> <span class="fu">xargs</span> <span class="at">-n</span> 4 <span class="at">-P</span> 8 <span class="at">-I</span> {} sh <span class="at">-c</span> <span class="dt">\</span></span>
<span id="cb3-13"><a href="#cb3-13"></a>  <span class="st">'mafft \</span></span>
<span id="cb3-14"><a href="#cb3-14"></a><span class="st">     --localpair \</span></span>
<span id="cb3-15"><a href="#cb3-15"></a><span class="st">     --maxiterate 1000 \</span></span>
<span id="cb3-16"><a href="#cb3-16"></a><span class="st">     --ep 0.123 \</span></span>
<span id="cb3-17"><a href="#cb3-17"></a><span class="st">     --thread 4 \</span></span>
<span id="cb3-18"><a href="#cb3-18"></a><span class="st">     ./02_cp_CDS_align/{} \</span></span>
<span id="cb3-19"><a href="#cb3-19"></a><span class="st">     &gt; ./02_cp_CDS_align/$(basename {} .aa.fa).aa.aln.fa'</span></span>
<span id="cb3-20"><a href="#cb3-20"></a>     </span>
<span id="cb3-21"><a href="#cb3-21"></a><span class="fu">ls</span> ./02_mt_CDS_align/ <span class="dt">\</span></span>
<span id="cb3-22"><a href="#cb3-22"></a>  <span class="kw">|</span> <span class="fu">xargs</span> <span class="at">-n</span> 4 <span class="at">-P</span> 8 <span class="at">-I</span> {} sh <span class="at">-c</span> <span class="dt">\</span></span>
<span id="cb3-23"><a href="#cb3-23"></a>  <span class="st">'mafft \</span></span>
<span id="cb3-24"><a href="#cb3-24"></a><span class="st">     --localpair \</span></span>
<span id="cb3-25"><a href="#cb3-25"></a><span class="st">     --maxiterate 1000 \</span></span>
<span id="cb3-26"><a href="#cb3-26"></a><span class="st">     --ep 0.123 \</span></span>
<span id="cb3-27"><a href="#cb3-27"></a><span class="st">     --thread 4 \</span></span>
<span id="cb3-28"><a href="#cb3-28"></a><span class="st">     ./02_mt_CDS_align/{} \</span></span>
<span id="cb3-29"><a href="#cb3-29"></a><span class="st">     &gt; ./02_mt_CDS_align/$(basename {} .aa.fa).aa.aln.fa'</span></span>
<span id="cb3-30"><a href="#cb3-30"></a></span>
<span id="cb3-31"><a href="#cb3-31"></a><span class="co"># translate back the alignments to codons</span></span>
<span id="cb3-32"><a href="#cb3-32"></a><span class="cf">for</span> file <span class="kw">in</span> ./02_cp_CDS_align/<span class="pp">*</span>.aa.aln.fa<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb3-33"><a href="#cb3-33"></a>  <span class="fu">perl</span> ~/bin/pal2nal.v14/pal2nal.pl <span class="dt">\</span></span>
<span id="cb3-34"><a href="#cb3-34"></a>    <span class="va">$file</span> <span class="dt">\</span></span>
<span id="cb3-35"><a href="#cb3-35"></a>    ./01_cp_CDS/<span class="va">$(</span><span class="fu">basename</span> <span class="va">$file</span> .aa.aln.fa<span class="va">)</span>.fa <span class="dt">\</span></span>
<span id="cb3-36"><a href="#cb3-36"></a>    <span class="at">-output</span> fasta <span class="dt">\</span></span>
<span id="cb3-37"><a href="#cb3-37"></a>    <span class="op">&gt;</span> ./02_cp_CDS_align/<span class="va">$(</span><span class="fu">basename</span> <span class="va">$file</span> .aa.aln.fa<span class="va">)</span>.fa<span class="kw">;</span></span>
<span id="cb3-38"><a href="#cb3-38"></a><span class="cf">done</span></span>
<span id="cb3-39"><a href="#cb3-39"></a></span>
<span id="cb3-40"><a href="#cb3-40"></a><span class="cf">for</span> file <span class="kw">in</span> ./02_mt_CDS_align/<span class="pp">*</span>.aa.aln.fa<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb3-41"><a href="#cb3-41"></a>  <span class="fu">perl</span> ~/bin/pal2nal.v14/pal2nal.pl <span class="dt">\</span></span>
<span id="cb3-42"><a href="#cb3-42"></a>    <span class="va">$file</span> <span class="dt">\</span></span>
<span id="cb3-43"><a href="#cb3-43"></a>    ./01_mt_CDS/<span class="va">$(</span><span class="fu">basename</span> <span class="va">$file</span> .aa.aln.fa<span class="va">)</span>.fa <span class="dt">\</span></span>
<span id="cb3-44"><a href="#cb3-44"></a>    <span class="at">-output</span> fasta <span class="dt">\</span></span>
<span id="cb3-45"><a href="#cb3-45"></a>    <span class="op">&gt;</span> ./02_mt_CDS_align/<span class="va">$(</span><span class="fu">basename</span> <span class="va">$file</span> .aa.aln.fa<span class="va">)</span>.fa<span class="kw">;</span></span>
<span id="cb3-46"><a href="#cb3-46"></a><span class="cf">done</span></span>
<span id="cb3-47"><a href="#cb3-47"></a></span>
<span id="cb3-48"><a href="#cb3-48"></a><span class="co"># clean up</span></span>
<span id="cb3-49"><a href="#cb3-49"></a><span class="fu">rm</span> ./02_cp_CDS_align/<span class="pp">*</span>.aa.fa <span class="dt">\</span></span>
<span id="cb3-50"><a href="#cb3-50"></a>  ./02_mt_CDS_align/<span class="pp">*</span>.aa.fa <span class="dt">\</span></span>
<span id="cb3-51"><a href="#cb3-51"></a>  ./02_cp_CDS_align/<span class="pp">*</span>.aa.aln.fa <span class="dt">\</span></span>
<span id="cb3-52"><a href="#cb3-52"></a>  ./02_mt_CDS_align/<span class="pp">*</span>.aa.aln.fa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Let’s now fetch the length of the alignments in R.</p>
<div class="cell" data-result="hide">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="co"># create gene distribution matrix</span></span>
<span id="cb4-2"><a href="#cb4-2"></a>gene_length_distributions <span class="ot">=</span> <span class="fu">matrix</span>(<span class="at">ncol =</span> <span class="dv">2</span>)</span>
<span id="cb4-3"><a href="#cb4-3"></a></span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="co"># get cp length distributions</span></span>
<span id="cb4-5"><a href="#cb4-5"></a>filenames <span class="ot">=</span> <span class="fu">list.files</span>(</span>
<span id="cb4-6"><a href="#cb4-6"></a>  <span class="st">"./02_cp_CDS_align/"</span>,</span>
<span id="cb4-7"><a href="#cb4-7"></a>  <span class="at">pattern =</span> <span class="st">"*.aln.fa"</span>,</span>
<span id="cb4-8"><a href="#cb4-8"></a>  <span class="at">full.names =</span> <span class="cn">TRUE</span></span>
<span id="cb4-9"><a href="#cb4-9"></a>)</span>
<span id="cb4-10"><a href="#cb4-10"></a>cp_genes_length <span class="ot">=</span> <span class="fu">lapply</span>(filenames, n.readLines, <span class="at">n =</span> <span class="dv">1</span>, <span class="at">skip =</span> <span class="dv">1</span>)</span>
<span id="cb4-11"><a href="#cb4-11"></a><span class="fu">names</span>(cp_genes_length) <span class="ot">=</span> stringr<span class="sc">::</span><span class="fu">str_remove_all</span>(filenames, <span class="st">"./02_cp_CDS_align//"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb4-12"><a href="#cb4-12"></a>  stringr<span class="sc">::</span><span class="fu">str_remove_all</span>(<span class="st">".aln.fa"</span>)</span>
<span id="cb4-13"><a href="#cb4-13"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(cp_genes_length)){</span>
<span id="cb4-14"><a href="#cb4-14"></a>  gene_length_distributions <span class="ot">=</span> <span class="fu">rbind</span>(gene_length_distributions, <span class="fu">c</span>(<span class="st">"cp"</span>, <span class="fu">nchar</span>(cp_genes_length[[i]][[<span class="dv">1</span>]])))</span>
<span id="cb4-15"><a href="#cb4-15"></a>  <span class="fu">rownames</span>(gene_length_distributions)[<span class="fu">nrow</span>(gene_length_distributions)] <span class="ot">=</span> <span class="fu">names</span>(cp_genes_length)[i]</span>
<span id="cb4-16"><a href="#cb4-16"></a>}</span>
<span id="cb4-17"><a href="#cb4-17"></a></span>
<span id="cb4-18"><a href="#cb4-18"></a><span class="co"># get mt length distributions</span></span>
<span id="cb4-19"><a href="#cb4-19"></a>filenames <span class="ot">=</span> <span class="fu">list.files</span>(</span>
<span id="cb4-20"><a href="#cb4-20"></a>  <span class="st">"./02_mt_CDS_align/"</span>,</span>
<span id="cb4-21"><a href="#cb4-21"></a>  <span class="at">pattern =</span> <span class="st">"*.aln.fa"</span>,</span>
<span id="cb4-22"><a href="#cb4-22"></a>  <span class="at">full.names =</span> <span class="cn">TRUE</span></span>
<span id="cb4-23"><a href="#cb4-23"></a>)</span>
<span id="cb4-24"><a href="#cb4-24"></a>mt_genes_length <span class="ot">=</span> <span class="fu">lapply</span>(filenames, n.readLines, <span class="at">n =</span> <span class="dv">1</span>, <span class="at">skip =</span> <span class="dv">1</span>)</span>
<span id="cb4-25"><a href="#cb4-25"></a><span class="fu">names</span>(mt_genes_length) <span class="ot">=</span> stringr<span class="sc">::</span><span class="fu">str_remove_all</span>(filenames, <span class="st">"./02_mt_CDS_align//"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb4-26"><a href="#cb4-26"></a>  stringr<span class="sc">::</span><span class="fu">str_remove_all</span>(<span class="st">".aln.fa"</span>)</span>
<span id="cb4-27"><a href="#cb4-27"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(mt_genes_length)){</span>
<span id="cb4-28"><a href="#cb4-28"></a>  gene_length_distributions <span class="ot">=</span> <span class="fu">rbind</span>(gene_length_distributions, <span class="fu">c</span>(<span class="st">"mt"</span>, <span class="fu">nchar</span>(mt_genes_length[[i]][[<span class="dv">1</span>]])))</span>
<span id="cb4-29"><a href="#cb4-29"></a>  <span class="fu">rownames</span>(gene_length_distributions)[<span class="fu">nrow</span>(gene_length_distributions)] <span class="ot">=</span> <span class="fu">names</span>(mt_genes_length)[i]</span>
<span id="cb4-30"><a href="#cb4-30"></a>}</span>
<span id="cb4-31"><a href="#cb4-31"></a></span>
<span id="cb4-32"><a href="#cb4-32"></a><span class="co"># reformat table</span></span>
<span id="cb4-33"><a href="#cb4-33"></a>gene_length_distributions <span class="ot">=</span> <span class="fu">as.data.frame</span>(gene_length_distributions)</span>
<span id="cb4-34"><a href="#cb4-34"></a>gene_length_distributions <span class="ot">=</span> gene_length_distributions[<span class="sc">-</span><span class="dv">1</span>, ]</span>
<span id="cb4-35"><a href="#cb4-35"></a><span class="fu">names</span>(gene_length_distributions) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Organell"</span>, <span class="st">"length (bp)"</span>)</span>
<span id="cb4-36"><a href="#cb4-36"></a>gene_length_distributions<span class="sc">$</span><span class="st">`</span><span class="at">length (bp)</span><span class="st">`</span> <span class="ot">=</span> <span class="fu">as.numeric</span>(gene_length_distributions<span class="sc">$</span><span class="st">`</span><span class="at">length (bp)</span><span class="st">`</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>And let’s plot the gene alignment distributions for the chloroplast and the mitochondrion.</p>
<div class="cell" data-result="hide">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="co"># plot</span></span>
<span id="cb5-2"><a href="#cb5-2"></a>p2 <span class="ot">=</span> <span class="fu">ggplot</span>(gene_length_distributions, <span class="fu">aes</span>(<span class="at">x =</span> Organell, <span class="at">y =</span> <span class="st">`</span><span class="at">length (bp)</span><span class="st">`</span>, <span class="at">fill =</span> Organell)) <span class="sc">+</span></span>
<span id="cb5-3"><a href="#cb5-3"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">group =</span> Organell, <span class="at">color =</span> Organell), <span class="at">shape =</span> <span class="st">"|"</span>, <span class="at">size =</span> <span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb5-4"><a href="#cb5-4"></a>  <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">fill =</span> Organell), <span class="at">alpha =</span> <span class="fl">0.75</span>, <span class="at">width =</span> <span class="fl">0.125</span>,</span>
<span id="cb5-5"><a href="#cb5-5"></a>               <span class="at">position =</span> <span class="fu">position_nudge</span>(<span class="at">x =</span> <span class="sc">-</span><span class="fl">0.1</span>),</span>
<span id="cb5-6"><a href="#cb5-6"></a>               <span class="at">outlier.colour =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb5-7"><a href="#cb5-7"></a>  ggdist<span class="sc">::</span><span class="fu">stat_halfeye</span>(<span class="fu">aes</span>(<span class="at">fill =</span> Organell), <span class="at">alpha =</span> <span class="fl">0.75</span>, <span class="at">slab_color =</span> <span class="st">"grey45"</span>, <span class="at">slab_size =</span> <span class="fl">0.5</span>, </span>
<span id="cb5-8"><a href="#cb5-8"></a>                       <span class="at">scale =</span> <span class="fl">0.7</span>, <span class="at">adjust =</span> <span class="dv">1</span>, <span class="at">justification =</span> <span class="sc">-</span><span class="fl">0.05</span>, <span class="at">.width =</span> <span class="dv">0</span>, <span class="at">point_colour =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb5-9"><a href="#cb5-9"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#469d89"</span>, <span class="st">"steelblue"</span>)) <span class="sc">+</span></span>
<span id="cb5-10"><a href="#cb5-10"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#469d89"</span>, <span class="st">"steelblue"</span>)) <span class="sc">+</span></span>
<span id="cb5-11"><a href="#cb5-11"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">labels =</span> <span class="fu">label_number</span>(<span class="at">suffix =</span> <span class="st">" kbp"</span>, <span class="at">scale =</span> <span class="fl">1e-3</span>)) <span class="sc">+</span></span>
<span id="cb5-12"><a href="#cb5-12"></a>  <span class="fu">coord_flip</span>() <span class="sc">+</span></span>
<span id="cb5-13"><a href="#cb5-13"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Gene Length"</span>) <span class="sc">+</span></span>
<span id="cb5-14"><a href="#cb5-14"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">24</span>, <span class="at">hjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb5-15"><a href="#cb5-15"></a>        <span class="at">axis.ticks.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb5-16"><a href="#cb5-16"></a>        <span class="at">axis.title.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">16</span>),</span>
<span id="cb5-17"><a href="#cb5-17"></a>        <span class="at">axis.title.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb5-18"><a href="#cb5-18"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">14</span>),</span>
<span id="cb5-19"><a href="#cb5-19"></a>        <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">18</span>),</span>
<span id="cb5-20"><a href="#cb5-20"></a>        <span class="at">legend.position =</span> <span class="st">"none"</span>,</span>
<span id="cb5-21"><a href="#cb5-21"></a>        <span class="at">panel.background =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb5-22"><a href="#cb5-22"></a>        <span class="at">panel.grid.major =</span> <span class="fu">element_line</span>(<span class="at">colour =</span> <span class="st">"grey85"</span>),</span>
<span id="cb5-23"><a href="#cb5-23"></a>        <span class="at">panel.grid.minor.x =</span> <span class="fu">element_line</span>(<span class="at">colour =</span> <span class="st">"grey85"</span>),</span>
<span id="cb5-24"><a href="#cb5-24"></a>        <span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">colour =</span> <span class="st">"black"</span>, <span class="at">fill =</span> <span class="cn">NA</span>))</span>
<span id="cb5-25"><a href="#cb5-25"></a><span class="fu">plot</span>(p2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="cp-and-mt-dataset-overview_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption>Figure 2: Alignment length distributions</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="reconstruction-of-ulva-species-tree" class="level2" data-number="1.3">
<h2 data-number="1.3" class="anchored" data-anchor-id="reconstruction-of-ulva-species-tree"><span class="header-section-number">1.3</span> Reconstruction of Ulva species tree</h2>
<p>We will reconstruct the Ulva phylogenetic species tree using a Maximum Likelihood approach with <a href="https://doi.org/10.1093/molbev/msaa015">IQ-TREE</a>, and a coalescence-based one with <a href="https://doi.org/10.1186/s12859-018-2129-y">ASTRAL</a>. We will generate the species tree based on:</p>
<ul>
<li>chloroplast only genes</li>
<li>mitochondrial only genes</li>
<li>both chloroplast and mitochondrial genes</li>
</ul>
<p>and we will evaluate the similarities and differences between the reconstructued phylogenies.</p>
</section>
<section id="sec-lesson-learnt-cp-and-mt-dataset-overview" class="level2" data-number="1.4">
<h2 data-number="1.4" class="anchored" data-anchor-id="sec-lesson-learnt-cp-and-mt-dataset-overview"><span class="header-section-number">1.4</span> Lessons Learnt</h2>
<p>So far, we have learnt:</p>
<ul>
<li>TCGA is a public resource with a wide collection of samples from cancer biopsies from patients affected by 32 different cancer types.</li>
<li>TCGA provides abundant molecular and clinical information for both biopsies and patients and it allows for patient stratification and therapeutic target discovery.</li>
<li>Cancers with kidneys as primary sites are the third most abundant samples available in TCGA (n = 1,030 samples), after breast (n = 1,246) and lung biopsies (n = 1,156), and provides an excellent yet challenging case-study.</li>
</ul>
</section>
<section id="session-information" class="level2" data-number="1.5">
<h2 data-number="1.5" class="anchored" data-anchor-id="session-information"><span class="header-section-number">1.5</span> Session Information</h2>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.2 (2023-10-31)
Platform: x86_64-conda-linux-gnu (64-bit)
Running under: openSUSE Tumbleweed

Matrix products: default
BLAS/LAPACK: /home/andrea/miniforge3/envs/moai/lib/libmkl_rt.so.2;  LAPACK version 3.9.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=it_IT.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Europe/Brussels
tzcode source: system (glibc)

attached base packages:
 [1] parallel  stats4    grid      stats     graphics  grDevices utils    
 [8] datasets  methods   base     

other attached packages:
 [1] treeio_1.26.0         TreeDist_2.9.2        stringr_1.5.1        
 [4] scales_1.3.0          RColorBrewer_1.1-3    reshape_0.8.9        
 [7] phytools_2.4-4        maps_3.4.2.1          phylogram_2.1.0      
[10] phangorn_2.12.1       gridExtra_2.3         ggtree_3.10.1        
[13] ggplot2_3.5.1         ggdist_3.3.2          dplyr_1.1.4          
[16] doSNOW_1.0.20         snow_0.4-4            iterators_1.0.14     
[19] foreach_1.5.2         dendextend_1.19.0     DECIPHER_2.30.0      
[22] RSQLite_2.3.9         Biostrings_2.70.3     GenomeInfoDb_1.38.8  
[25] XVector_0.42.0        IRanges_2.36.0        S4Vectors_0.40.2     
[28] BiocGenerics_0.48.1   corrplot_0.95         ComplexHeatmap_2.18.0
[31] circlize_0.4.16       ape_5.8-1            

loaded via a namespace (and not attached):
  [1] jsonlite_1.8.9          shape_1.4.6.1           magrittr_2.0.3         
  [4] farver_2.1.2            rmarkdown_2.29          GlobalOptions_0.1.2    
  [7] fs_1.6.5                zlibbioc_1.48.2         vctrs_0.6.5            
 [10] memoise_2.0.1           RCurl_1.98-1.16         htmltools_0.5.8.1      
 [13] distributional_0.5.0    DEoptim_2.2-8           gridGraphics_0.5-1     
 [16] sass_0.4.9              bslib_0.8.0             htmlwidgets_1.6.4      
 [19] plyr_1.8.9              cachem_1.1.0            igraph_2.1.4           
 [22] mime_0.12               lifecycle_1.0.4         pkgconfig_2.0.3        
 [25] Matrix_1.6-5            R6_2.5.1                fastmap_1.2.0          
 [28] GenomeInfoDbData_1.2.11 rbibutils_2.3           shiny_1.10.0           
 [31] clue_0.3-66             digest_0.6.37           numDeriv_2016.8-1.1    
 [34] aplot_0.2.4             colorspace_2.1-1        patchwork_1.3.0        
 [37] crosstalk_1.2.1         labeling_0.4.3          clusterGeneration_1.3.8
 [40] compiler_4.3.2          bit64_4.6.0-1           withr_3.0.2            
 [43] doParallel_1.0.17       optimParallel_1.0-2     viridis_0.6.5          
 [46] DBI_1.2.3               R.utils_2.12.3          MASS_7.3-60.0.1        
 [49] rjson_0.2.23            scatterplot3d_0.3-44    tools_4.3.2            
 [52] httpuv_1.6.15           TreeTools_1.13.0        R.oo_1.27.0            
 [55] glue_1.8.0              quadprog_1.5-8          nlme_3.1-167           
 [58] R.cache_0.16.0          promises_1.3.2          reshape2_1.4.4         
 [61] cluster_2.1.8           PlotTools_0.3.1         generics_0.1.3         
 [64] gtable_0.3.6            R.methodsS3_1.8.2       tidyr_1.3.1            
 [67] pillar_1.10.1           yulab.utils_0.2.0       later_1.4.1            
 [70] lattice_0.22-6          bit_4.5.0.1             tidyselect_1.2.1       
 [73] knitr_1.49              xfun_0.50               expm_1.0-0             
 [76] matrixStats_1.5.0       DT_0.33                 stringi_1.8.4          
 [79] lazyeval_0.2.2          ggfun_0.1.8             yaml_2.3.10            
 [82] evaluate_1.0.3          codetools_0.2-20        tibble_3.2.1           
 [85] ggplotify_0.1.2         cli_3.6.3               xtable_1.8-4           
 [88] Rdpack_2.6.2            jquerylib_0.1.4         munsell_0.5.1          
 [91] Rcpp_1.0.14             coda_0.19-4.1           png_0.1-8              
 [94] blob_1.2.4              bitops_1.0-9            viridisLite_0.4.2      
 [97] tidytree_0.4.6          purrr_1.0.2             crayon_1.5.3           
[100] combinat_0.0-8          GetoptLong_1.0.5        rlang_1.1.5            
[103] fastmatch_1.1-6         mnormt_2.1.1            shinyjs_2.1.0          </code></pre>
</div>
</div>
</div>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./index.html" class="pagination-link" aria-label="Abstract">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Abstract</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./cp-and-mt-phylogeny-reconstruction.html" class="pagination-link" aria-label="Species tree reconstruction from organellar genomes">
        <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Species tree reconstruction from organellar genomes</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      <ul class="footer-items list-unstyled">
    <li class="nav-item">
 Copyright © 2025 Andrea Del Cortona.
  </li>  
</ul>
    </div>   
    <div class="nav-footer-center">
      <ul class="footer-items list-unstyled">
    <li class="nav-item">
    <a class="nav-link" href="https://quarto.org/">
<p>This page is built with ❤️ and <em>Quarto</em>.</p>
</a>
  </li>  
</ul>
    </div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/19ADC99">
      <i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/andrea-del-cortona-86a75474/">
      <i class="bi bi-linkedin" role="img" aria-label="Lindekin">
</i> 
    </a>
  </li>  
</ul>
    </div>
  </div>
</footer>




</body></html>