<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>3&nbsp; Gene markers selection – Development of taxonomical biomarkers for Ulva genus</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./cp-and-mt-phylogeny-reconstruction.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<link href="site_libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="site_libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="site_libs/datatables-binding-0.33/datatables.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="site_libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="site_libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="site_libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script>
<link href="site_libs/dt-ext-fixedcolumns-1.13.6/css/fixedColumns.dataTables.min.css" rel="stylesheet">
<script src="site_libs/dt-ext-fixedcolumns-1.13.6/js/dataTables.fixedColumns.min.js"></script>
<link href="site_libs/dt-ext-fixedheader-1.13.6/css/fixedHeader.dataTables.min.css" rel="stylesheet">
<script src="site_libs/dt-ext-fixedheader-1.13.6/js/dataTables.fixedHeader.min.js"></script>
<link href="site_libs/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="site_libs/crosstalk-1.2.1/js/crosstalk.min.js"></script>
<style>
#quarto-sidebar .sidebar-item-text {
  font-weight: bold;
}
#quarto-sidebar .chapter-number {
  font-weight: normal
}
#quarto-sidebar .chapter-title {
  font-weight: normal
}
#.nav-tabs .nav-link.active {
  background-color: rgba(239, 236, 255, 0.5);
}
</style>


</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Development of taxonomical biomarkers for Ulva genus</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./markers-selection.html"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Gene markers selection</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Abstract</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cp-and-mt-dataset-overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Ulva organellar genomes retrieval and preparation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cp-and-mt-phylogeny-reconstruction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Species tree reconstruction from organellar genomes</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./markers-selection.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Gene markers selection</span></span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#on-this-page" id="toc-on-this-page" class="nav-link active" data-scroll-target="#on-this-page"><span class="header-section-number">3.1</span> On this page</a></li>
  <li><a href="#organellar-gene-selection" id="toc-organellar-gene-selection" class="nav-link" data-scroll-target="#organellar-gene-selection"><span class="header-section-number">3.2</span> Organellar gene selection</a>
  <ul>
  <li><a href="#gene-distances-to-species-three" id="toc-gene-distances-to-species-three" class="nav-link" data-scroll-target="#gene-distances-to-species-three"><span class="header-section-number">3.2.1</span> gene distances to species three</a></li>
  </ul></li>
  <li><a href="#markers-selection" id="toc-markers-selection" class="nav-link" data-scroll-target="#markers-selection"><span class="header-section-number">3.3</span> Markers selection</a>
  <ul>
  <li><a href="#cp-only" id="toc-cp-only" class="nav-link" data-scroll-target="#cp-only"><span class="header-section-number">3.3.1</span> cp only</a></li>
  <li><a href="#mt-only" id="toc-mt-only" class="nav-link" data-scroll-target="#mt-only"><span class="header-section-number">3.3.2</span> mt only</a></li>
  <li><a href="#cp-mt" id="toc-cp-mt" class="nav-link" data-scroll-target="#cp-mt"><span class="header-section-number">3.3.3</span> cp + mt</a></li>
  </ul></li>
  <li><a href="#sec-lesson-learnt-markers-selection" id="toc-sec-lesson-learnt-markers-selection" class="nav-link" data-scroll-target="#sec-lesson-learnt-markers-selection"><span class="header-section-number">3.4</span> Lessons Learnt</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Gene markers selection</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div class="cell">
<style type="text/css">

pre {
  max-height: 300px;
}

div.sourceCode {
  overflow: scroll;
}

.bash-code {
  font-style: italic;
}

</style>
</div>
<section id="on-this-page" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="on-this-page"><span class="header-section-number">3.1</span> On this page</h2>
<p>Biological insights and take-home messages are at the bottom of the page at Lesson Learnt: <a href="#sec-lesson-learnt-markers-selection" class="quarto-xref"><span>Section 3.4</span></a>.</p>
<ul>
<li>Here we focus on samples from Kidney Cancers patients and their overall relationships.</li>
<li>We then explore the association between each Kidney Cancer subtype and the major clinical covariates (e.g.: sex, disease status, cancer stage, clinical outcome).</li>
</ul>
</section>
<section id="organellar-gene-selection" class="level2" data-number="3.2">
<h2 data-number="3.2" class="anchored" data-anchor-id="organellar-gene-selection"><span class="header-section-number">3.2</span> Organellar gene selection</h2>
<p>select genes based on their distance to the reconstructed CP+MT phylogenetic three</p>
<p>we select only 10 genes (6 cp and 4 mt) because doing the combination with all the 99 genes wilñl take forever</p>
<section id="gene-distances-to-species-three" class="level3" data-number="3.2.1">
<h3 data-number="3.2.1" class="anchored" data-anchor-id="gene-distances-to-species-three"><span class="header-section-number">3.2.1</span> gene distances to species three</h3>
<p>We take cp+mt ML tree as the real species tree for Ulva. Let’s now calculate the distance between each gene-tree and the species tree that we have reconstructed. Then, we can chose the top 5-10 genes that are closer to the reconstructed species tree to see what is their minimal combination able to reconstruct the gene tree.</p>
<div class="cell" data-result="hide">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="co"># get cp ML gene trees</span></span>
<span id="cb1-2"><a href="#cb1-2"></a>cp_ML_genes <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb1-3"><a href="#cb1-3"></a>filenames <span class="ot">=</span> <span class="fu">list.files</span>(<span class="st">"./03_cp_singleGene_ML/"</span>, <span class="at">pattern =</span> <span class="st">"*.contree"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-4"><a href="#cb1-4"></a>cp_ML_genes <span class="ot">=</span> <span class="fu">lapply</span>(filenames, ape<span class="sc">::</span>read.tree)</span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="fu">names</span>(cp_ML_genes) <span class="ot">=</span> strigr<span class="sc">::</span><span class="fu">str_remove_all</span>(filenames, <span class="st">"./03_cp_singleGene_ML//"</span>)</span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="fu">names</span>(cp_ML_genes) <span class="ot">=</span> strigr<span class="sc">::</span><span class="fu">str_remove_all</span>(filenames, <span class="st">".aln.contree"</span>)</span>
<span id="cb1-7"><a href="#cb1-7"></a>cp_ML_genes[[<span class="st">"cp_mt_ML"</span>]] <span class="ot">=</span> phylogenetic_trees[[<span class="st">"cp_mt_ML"</span>]][[<span class="st">"tree"</span>]]</span>
<span id="cb1-8"><a href="#cb1-8"></a></span>
<span id="cb1-9"><a href="#cb1-9"></a><span class="co"># get mt ML gene trees</span></span>
<span id="cb1-10"><a href="#cb1-10"></a>mt_ML_genes <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb1-11"><a href="#cb1-11"></a>filenames <span class="ot">=</span> <span class="fu">list.files</span>(<span class="st">"./03_mt_singleGene_ML/"</span>, <span class="at">pattern =</span> <span class="st">"*.contree"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-12"><a href="#cb1-12"></a>mt_ML_genes <span class="ot">=</span> <span class="fu">lapply</span>(filenames, ape<span class="sc">::</span>read.tree)</span>
<span id="cb1-13"><a href="#cb1-13"></a><span class="fu">names</span>(mt_ML_genes) <span class="ot">=</span> strigr<span class="sc">::</span><span class="fu">str_remove_all</span>(filenames, <span class="st">"./03_mt_singleGene_ML//"</span>)</span>
<span id="cb1-14"><a href="#cb1-14"></a><span class="fu">names</span>(mt_ML_genes) <span class="ot">=</span> strigr<span class="sc">::</span><span class="fu">str_remove_all</span>(filenames, <span class="st">".aln.contree"</span>)</span>
<span id="cb1-15"><a href="#cb1-15"></a>mt_ML_genes[[<span class="st">"cp_mt_ML"</span>]] <span class="ot">=</span> phylogenetic_trees[[<span class="st">"cp_mt_ML"</span>]][[<span class="st">"tree"</span>]]</span>
<span id="cb1-16"><a href="#cb1-16"></a></span>
<span id="cb1-17"><a href="#cb1-17"></a><span class="co"># create distance lists</span></span>
<span id="cb1-18"><a href="#cb1-18"></a>genes_sets <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_ML"</span> <span class="ot">=</span> <span class="fu">c</span>(cp_ML_genes, mt_ML_genes[<span class="fu">which</span>(<span class="fu">names</span>(mt_ML_genes) <span class="sc">!=</span> <span class="st">"cp_mt_ML"</span>)]))</span>
<span id="cb1-19"><a href="#cb1-19"></a></span>
<span id="cb1-20"><a href="#cb1-20"></a>empty_matrix <span class="ot">=</span> <span class="fu">matrix</span>(<span class="at">nrow =</span> <span class="fu">length</span>(genes_sets[[<span class="st">"cp_mt_ML"</span>]]), <span class="at">ncol =</span> <span class="fu">length</span>(genes_sets[[<span class="st">"cp_mt_ML"</span>]]))</span>
<span id="cb1-21"><a href="#cb1-21"></a><span class="fu">colnames</span>(empty_matrix) <span class="ot">=</span> <span class="fu">names</span>(genes_sets[[<span class="st">"cp_mt_ML"</span>]])</span>
<span id="cb1-22"><a href="#cb1-22"></a><span class="fu">rownames</span>(empty_matrix) <span class="ot">=</span> <span class="fu">names</span>(genes_sets[[<span class="st">"cp_mt_ML"</span>]])</span>
<span id="cb1-23"><a href="#cb1-23"></a>cp_mt_genes_distances <span class="ot">=</span> <span class="fu">list</span>(</span>
<span id="cb1-24"><a href="#cb1-24"></a>  <span class="st">"common_nodes_corr"</span> <span class="ot">=</span> empty_matrix,</span>
<span id="cb1-25"><a href="#cb1-25"></a>  <span class="st">"RF"</span> <span class="ot">=</span> empty_matrix,</span>
<span id="cb1-26"><a href="#cb1-26"></a>  <span class="st">"KF"</span> <span class="ot">=</span> empty_matrix,</span>
<span id="cb1-27"><a href="#cb1-27"></a>  <span class="st">"MAST_%"</span> <span class="ot">=</span> empty_matrix</span>
<span id="cb1-28"><a href="#cb1-28"></a>)</span>
<span id="cb1-29"><a href="#cb1-29"></a></span>
<span id="cb1-30"><a href="#cb1-30"></a><span class="co"># populate distances lists</span></span>
<span id="cb1-31"><a href="#cb1-31"></a>distance_sets <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_genes_distances"</span> <span class="ot">=</span> cp_mt_genes_distances)</span>
<span id="cb1-32"><a href="#cb1-32"></a></span>
<span id="cb1-33"><a href="#cb1-33"></a><span class="co"># get pairwise distances</span></span>
<span id="cb1-34"><a href="#cb1-34"></a><span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(distance_sets)){</span>
<span id="cb1-35"><a href="#cb1-35"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(genes_sets[[j]])){</span>
<span id="cb1-36"><a href="#cb1-36"></a>    <span class="cf">for</span>(k <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(genes_sets[[j]])){</span>
<span id="cb1-37"><a href="#cb1-37"></a>      </span>
<span id="cb1-38"><a href="#cb1-38"></a>      <span class="co"># error handling</span></span>
<span id="cb1-39"><a href="#cb1-39"></a>      <span class="co"># MAST call errors for very unresulved trees (e.g.: psaM)</span></span>
<span id="cb1-40"><a href="#cb1-40"></a>      <span class="co">#   Error in root.phylo(x, bipart_x, resolve.root = TRUE) : </span></span>
<span id="cb1-41"><a href="#cb1-41"></a>      <span class="co">#   the specified outgroup is not monophyletic</span></span>
<span id="cb1-42"><a href="#cb1-42"></a>      <span class="co"># catch error and move forward</span></span>
<span id="cb1-43"><a href="#cb1-43"></a>      skip_to_next <span class="ot">=</span> <span class="cn">FALSE</span></span>
<span id="cb1-44"><a href="#cb1-44"></a>      </span>
<span id="cb1-45"><a href="#cb1-45"></a>      <span class="co"># get tree</span></span>
<span id="cb1-46"><a href="#cb1-46"></a>      tree1_tmp <span class="ot">=</span> genes_sets[[j]][[i]]</span>
<span id="cb1-47"><a href="#cb1-47"></a>      tree2_tmp <span class="ot">=</span> genes_sets[[j]][[k]]</span>
<span id="cb1-48"><a href="#cb1-48"></a>      </span>
<span id="cb1-49"><a href="#cb1-49"></a>      <span class="co"># list of common species</span></span>
<span id="cb1-50"><a href="#cb1-50"></a>      species_list <span class="ot">=</span> tree1_tmp<span class="sc">$</span>tip.label[<span class="fu">which</span>(tree1_tmp<span class="sc">$</span>tip.label <span class="sc">%in%</span> tree2_tmp<span class="sc">$</span>tip.label)]</span>
<span id="cb1-51"><a href="#cb1-51"></a>      </span>
<span id="cb1-52"><a href="#cb1-52"></a>      <span class="co"># get overlapping species</span></span>
<span id="cb1-53"><a href="#cb1-53"></a>      tree1 <span class="ot">=</span> ape<span class="sc">::</span><span class="fu">keep.tip</span>(tree1_tmp, species_list)</span>
<span id="cb1-54"><a href="#cb1-54"></a>      tree2 <span class="ot">=</span> ape<span class="sc">::</span><span class="fu">keep.tip</span>(tree2_tmp, species_list)</span>
<span id="cb1-55"><a href="#cb1-55"></a>      </span>
<span id="cb1-56"><a href="#cb1-56"></a>      <span class="co"># get distances</span></span>
<span id="cb1-57"><a href="#cb1-57"></a>      distances <span class="ot">=</span> phangorn<span class="sc">::</span><span class="fu">treedist</span>(tree1, tree2)</span>
<span id="cb1-58"><a href="#cb1-58"></a>      distance_sets[[j]][[<span class="st">"RF"</span>]][i, k] <span class="ot">=</span> distances[[<span class="dv">1</span>]]</span>
<span id="cb1-59"><a href="#cb1-59"></a>      distance_sets[[j]][[<span class="st">"KF"</span>]][i, k] <span class="ot">=</span> distances[[<span class="dv">2</span>]]</span>
<span id="cb1-60"><a href="#cb1-60"></a>      <span class="co">#dist_DB_pairwise[["path_diff"]][i, k] = distances[[3]]</span></span>
<span id="cb1-61"><a href="#cb1-61"></a>      distance_sets[[j]][[<span class="st">"MAST_%"</span>]][i, k] <span class="ot">=</span> <span class="fu">tryCatch</span>(</span>
<span id="cb1-62"><a href="#cb1-62"></a>        <span class="fu">length</span>(phangorn<span class="sc">::</span><span class="fu">mast</span>(tree1, tree2, <span class="at">tree =</span> <span class="cn">FALSE</span>)) <span class="sc">/</span> <span class="fu">length</span>(species_list),</span>
<span id="cb1-63"><a href="#cb1-63"></a>        <span class="at">error =</span> <span class="cf">function</span>(e) { skip_to_next <span class="ot">&lt;&lt;-</span> <span class="cn">TRUE</span> }</span>
<span id="cb1-64"><a href="#cb1-64"></a>      )</span>
<span id="cb1-65"><a href="#cb1-65"></a>      </span>
<span id="cb1-66"><a href="#cb1-66"></a>      <span class="co"># error handling</span></span>
<span id="cb1-67"><a href="#cb1-67"></a>      <span class="cf">if</span>(skip_to_next){distance_sets[[j]][[<span class="st">"MAST_%"</span>]][i, k] <span class="ot">=</span> <span class="cn">NA</span>}</span>
<span id="cb1-68"><a href="#cb1-68"></a>      </span>
<span id="cb1-69"><a href="#cb1-69"></a>    }</span>
<span id="cb1-70"><a href="#cb1-70"></a>  }</span>
<span id="cb1-71"><a href="#cb1-71"></a>}</span>
<span id="cb1-72"><a href="#cb1-72"></a></span>
<span id="cb1-73"><a href="#cb1-73"></a></span>
<span id="cb1-74"><a href="#cb1-74"></a><span class="do">### get pairwise common_nodes_corr distances</span></span>
<span id="cb1-75"><a href="#cb1-75"></a></span>
<span id="cb1-76"><a href="#cb1-76"></a><span class="co"># create CPU cluster</span></span>
<span id="cb1-77"><a href="#cb1-77"></a>cl <span class="ot">=</span> parallel<span class="sc">::</span><span class="fu">makeCluster</span>(<span class="dv">8</span>, <span class="at">type =</span> <span class="st">"SOCK"</span>)</span>
<span id="cb1-78"><a href="#cb1-78"></a><span class="fu">registerDoSNOW</span>(cl)</span>
<span id="cb1-79"><a href="#cb1-79"></a></span>
<span id="cb1-80"><a href="#cb1-80"></a><span class="co"># get chronos in parallel</span></span>
<span id="cb1-81"><a href="#cb1-81"></a>chronos_list <span class="ot">=</span> <span class="fu">foreach</span>(<span class="at">k =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(genes_sets[[<span class="st">"cp_mt_ML"</span>]]), <span class="at">.combine =</span> <span class="st">"c"</span>) <span class="sc">%dopar%</span> {</span>
<span id="cb1-82"><a href="#cb1-82"></a>  <span class="fu">library</span>(<span class="st">"ape"</span>)</span>
<span id="cb1-83"><a href="#cb1-83"></a>  <span class="co"># get tree</span></span>
<span id="cb1-84"><a href="#cb1-84"></a>  <span class="fu">tryCatch</span>(<span class="fu">chronos</span>(genes_sets[[<span class="st">"cp_mt_ML"</span>]][[k]]), <span class="at">error =</span> <span class="cf">function</span>(e) { <span class="cn">NA</span> })</span>
<span id="cb1-85"><a href="#cb1-85"></a>}  </span>
<span id="cb1-86"><a href="#cb1-86"></a>parallel<span class="sc">::</span><span class="fu">stopCluster</span>(cl)</span>
<span id="cb1-87"><a href="#cb1-87"></a><span class="fu">names</span>(chronos_list) <span class="ot">=</span> <span class="fu">names</span>(genes_sets[[<span class="st">"cp_mt_ML"</span>]])</span>
<span id="cb1-88"><a href="#cb1-88"></a></span>
<span id="cb1-89"><a href="#cb1-89"></a><span class="co"># get pairwise common_nodes_corr distances</span></span>
<span id="cb1-90"><a href="#cb1-90"></a><span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(distance_sets)){</span>
<span id="cb1-91"><a href="#cb1-91"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(genes_sets[[j]])){</span>
<span id="cb1-92"><a href="#cb1-92"></a>    <span class="cf">for</span>(k <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(genes_sets[[j]])){</span>
<span id="cb1-93"><a href="#cb1-93"></a>      </span>
<span id="cb1-94"><a href="#cb1-94"></a>      <span class="co"># get trees</span></span>
<span id="cb1-95"><a href="#cb1-95"></a>      tree1_tmp <span class="ot">=</span> chronos_list[[<span class="fu">names</span>(genes_sets[[j]])[i]]]</span>
<span id="cb1-96"><a href="#cb1-96"></a>      tree2_tmp <span class="ot">=</span> chronos_list[[<span class="fu">names</span>(genes_sets[[j]])[k]]]</span>
<span id="cb1-97"><a href="#cb1-97"></a>      </span>
<span id="cb1-98"><a href="#cb1-98"></a>      <span class="co"># check trees and get distance</span></span>
<span id="cb1-99"><a href="#cb1-99"></a>      <span class="cf">if</span>(<span class="fu">all</span>(<span class="sc">!</span><span class="fu">is.na</span>(tree1_tmp)) <span class="sc">&amp;</span> <span class="fu">all</span>(<span class="sc">!</span><span class="fu">is.na</span>(tree2_tmp))){</span>
<span id="cb1-100"><a href="#cb1-100"></a>        distance_sets[[j]][[<span class="st">"common_nodes_corr"</span>]][i, k] <span class="ot">=</span> <span class="fu">cor.dendlist</span>(</span>
<span id="cb1-101"><a href="#cb1-101"></a>          <span class="fu">dendlist</span>(<span class="fu">as.dendrogram</span>(tree1_tmp), <span class="fu">as.dendrogram</span>(tree2_tmp)),</span>
<span id="cb1-102"><a href="#cb1-102"></a>          <span class="at">method =</span> <span class="st">"common_nodes"</span>)[<span class="dv">2</span>]</span>
<span id="cb1-103"><a href="#cb1-103"></a>      } <span class="cf">else</span> {</span>
<span id="cb1-104"><a href="#cb1-104"></a>        distance_sets[[j]][[<span class="st">"common_nodes_corr"</span>]][i, k] <span class="ot">=</span> <span class="cn">NA</span></span>
<span id="cb1-105"><a href="#cb1-105"></a>      }</span>
<span id="cb1-106"><a href="#cb1-106"></a>    }</span>
<span id="cb1-107"><a href="#cb1-107"></a>  }</span>
<span id="cb1-108"><a href="#cb1-108"></a>}</span>
<span id="cb1-109"><a href="#cb1-109"></a></span>
<span id="cb1-110"><a href="#cb1-110"></a></span>
<span id="cb1-111"><a href="#cb1-111"></a><span class="co"># rename matrices for cp/mt overlapping rpl/rps genes (for plotting reasons)</span></span>
<span id="cb1-112"><a href="#cb1-112"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(distance_sets[[<span class="st">"cp_mt_genes_distances"</span>]])){</span>
<span id="cb1-113"><a href="#cb1-113"></a>  </span>
<span id="cb1-114"><a href="#cb1-114"></a>  <span class="fu">rownames</span>(distance_sets[[<span class="st">"cp_mt_genes_distances"</span>]][[i]]) <span class="ot">=</span> <span class="fu">c</span>(</span>
<span id="cb1-115"><a href="#cb1-115"></a>    <span class="st">"accD"</span>, <span class="st">"atpA"</span>, <span class="st">"atpB"</span>, <span class="st">"atpE"</span>, <span class="st">"atpF"</span>, <span class="st">"atpH"</span>, <span class="st">"atpI"</span>, <span class="st">"ccsA"</span>, <span class="st">"cemA"</span>, <span class="st">"chlI"</span>, <span class="st">"clpP"</span>,</span>
<span id="cb1-116"><a href="#cb1-116"></a>    <span class="st">"infA"</span>, <span class="st">"petA"</span>, <span class="st">"petB"</span>, <span class="st">"petD"</span>, <span class="st">"petG"</span>, <span class="st">"petL"</span>, <span class="st">"psaA"</span>, <span class="st">"psaB"</span>, <span class="st">"psaC"</span>, <span class="st">"psaI"</span>, <span class="st">"psaJ"</span>,</span>
<span id="cb1-117"><a href="#cb1-117"></a>    <span class="st">"psaM"</span>, <span class="st">"psbA"</span>, <span class="st">"psbB"</span>, <span class="st">"psbC"</span>, <span class="st">"psbD"</span>, <span class="st">"psbE"</span>, <span class="st">"psbF"</span>, <span class="st">"psbH"</span>, <span class="st">"psbI"</span>, <span class="st">"psbJ"</span>, <span class="st">"psbK"</span>,</span>
<span id="cb1-118"><a href="#cb1-118"></a>    <span class="st">"psbL"</span>, <span class="st">"psbM"</span>, <span class="st">"psbN"</span>, <span class="st">"psbT"</span>, <span class="st">"psbZ"</span>, <span class="st">"rbcL"</span>, <span class="st">"rpl12"</span>, <span class="st">"rpl14"</span>, <span class="st">"rpl16"</span>, <span class="st">"rpl19"</span>, <span class="st">"rpl2"</span>,</span>
<span id="cb1-119"><a href="#cb1-119"></a>    <span class="st">"rpl20"</span>, <span class="st">"rpl23"</span>, <span class="st">"rpl32"</span>, <span class="st">"rpl36"</span>, <span class="st">"rpl5"</span>, <span class="st">"rpoA"</span>, <span class="st">"rpoB"</span>, <span class="st">"rpoC1"</span>, <span class="st">"rpoC2"</span>, <span class="st">"rps11"</span>,</span>
<span id="cb1-120"><a href="#cb1-120"></a>    <span class="st">"rps12"</span>, <span class="st">"rps14"</span>, <span class="st">"rps18"</span>, <span class="st">"rps19"</span>, <span class="st">"rps2"</span>, <span class="st">"rps3"</span>,  <span class="st">"rps4"</span>, <span class="st">"rps7"</span>, <span class="st">"rps8"</span>, <span class="st">"rps9"</span>,</span>
<span id="cb1-121"><a href="#cb1-121"></a>    <span class="st">"tufA"</span>, <span class="st">"ycf1"</span>, <span class="st">"ycf12"</span>, <span class="st">"ycf20"</span>, <span class="st">"ycf3"</span>, <span class="st">"ycf4"</span>, <span class="st">"cp_mt_ML"</span>, <span class="st">"atp1"</span>, <span class="st">"atp4"</span>, <span class="st">"atp6"</span>,</span>
<span id="cb1-122"><a href="#cb1-122"></a>    <span class="st">"atp8"</span>, <span class="st">"atp9"</span>, <span class="st">"cob"</span>, <span class="st">"cox1"</span>, <span class="st">"cox2"</span>, <span class="st">"cox3"</span>, <span class="st">"nad1"</span>, <span class="st">"nad2"</span>, <span class="st">"nad3"</span>, <span class="st">"nad4"</span>, <span class="st">"nad4L"</span>,</span>
<span id="cb1-123"><a href="#cb1-123"></a>    <span class="st">"nad5"</span>, <span class="st">"nad6"</span>, <span class="st">"nad7"</span>, <span class="st">"rpl14 "</span>, <span class="st">"rpl16 "</span>, <span class="st">"rpl5 "</span>, <span class="st">"rps10"</span>, <span class="st">"rps11 "</span>, <span class="st">"rps12 "</span>, <span class="st">"rps13"</span>,</span>
<span id="cb1-124"><a href="#cb1-124"></a>    <span class="st">"rps14 "</span>, <span class="st">"rps19 "</span>, <span class="st">"rps2 "</span>, <span class="st">"rps3 "</span>, <span class="st">"rps4 "</span></span>
<span id="cb1-125"><a href="#cb1-125"></a>  )</span>
<span id="cb1-126"><a href="#cb1-126"></a>  </span>
<span id="cb1-127"><a href="#cb1-127"></a>  <span class="fu">colnames</span>(distance_sets[[<span class="st">"cp_mt_genes_distances"</span>]][[i]]) <span class="ot">=</span> <span class="fu">c</span>(</span>
<span id="cb1-128"><a href="#cb1-128"></a>    <span class="st">"accD"</span>, <span class="st">"atpA"</span>, <span class="st">"atpB"</span>, <span class="st">"atpE"</span>, <span class="st">"atpF"</span>, <span class="st">"atpH"</span>, <span class="st">"atpI"</span>, <span class="st">"ccsA"</span>, <span class="st">"cemA"</span>, <span class="st">"chlI"</span>, <span class="st">"clpP"</span>,</span>
<span id="cb1-129"><a href="#cb1-129"></a>    <span class="st">"infA"</span>, <span class="st">"petA"</span>, <span class="st">"petB"</span>, <span class="st">"petD"</span>, <span class="st">"petG"</span>, <span class="st">"petL"</span>, <span class="st">"psaA"</span>, <span class="st">"psaB"</span>, <span class="st">"psaC"</span>, <span class="st">"psaI"</span>, <span class="st">"psaJ"</span>,</span>
<span id="cb1-130"><a href="#cb1-130"></a>    <span class="st">"psaM"</span>, <span class="st">"psbA"</span>, <span class="st">"psbB"</span>, <span class="st">"psbC"</span>, <span class="st">"psbD"</span>, <span class="st">"psbE"</span>, <span class="st">"psbF"</span>, <span class="st">"psbH"</span>, <span class="st">"psbI"</span>, <span class="st">"psbJ"</span>, <span class="st">"psbK"</span>,</span>
<span id="cb1-131"><a href="#cb1-131"></a>    <span class="st">"psbL"</span>, <span class="st">"psbM"</span>, <span class="st">"psbN"</span>, <span class="st">"psbT"</span>, <span class="st">"psbZ"</span>, <span class="st">"rbcL"</span>, <span class="st">"rpl12"</span>, <span class="st">"rpl14"</span>, <span class="st">"rpl16"</span>, <span class="st">"rpl19"</span>, <span class="st">"rpl2"</span>,</span>
<span id="cb1-132"><a href="#cb1-132"></a>    <span class="st">"rpl20"</span>, <span class="st">"rpl23"</span>, <span class="st">"rpl32"</span>, <span class="st">"rpl36"</span>, <span class="st">"rpl5"</span>, <span class="st">"rpoA"</span>, <span class="st">"rpoB"</span>, <span class="st">"rpoC1"</span>, <span class="st">"rpoC2"</span>, <span class="st">"rps11"</span>,</span>
<span id="cb1-133"><a href="#cb1-133"></a>    <span class="st">"rps12"</span>, <span class="st">"rps14"</span>, <span class="st">"rps18"</span>, <span class="st">"rps19"</span>, <span class="st">"rps2"</span>, <span class="st">"rps3"</span>,  <span class="st">"rps4"</span>, <span class="st">"rps7"</span>, <span class="st">"rps8"</span>, <span class="st">"rps9"</span>,</span>
<span id="cb1-134"><a href="#cb1-134"></a>    <span class="st">"tufA"</span>, <span class="st">"ycf1"</span>, <span class="st">"ycf12"</span>, <span class="st">"ycf20"</span>, <span class="st">"ycf3"</span>, <span class="st">"ycf4"</span>, <span class="st">"cp_mt_ML"</span>, <span class="st">"atp1"</span>, <span class="st">"atp4"</span>, <span class="st">"atp6"</span>,</span>
<span id="cb1-135"><a href="#cb1-135"></a>    <span class="st">"atp8"</span>, <span class="st">"atp9"</span>, <span class="st">"cob"</span>, <span class="st">"cox1"</span>, <span class="st">"cox2"</span>, <span class="st">"cox3"</span>, <span class="st">"nad1"</span>, <span class="st">"nad2"</span>, <span class="st">"nad3"</span>, <span class="st">"nad4"</span>, <span class="st">"nad4L"</span>,</span>
<span id="cb1-136"><a href="#cb1-136"></a>    <span class="st">"nad5"</span>, <span class="st">"nad6"</span>, <span class="st">"nad7"</span>, <span class="st">"rpl14 "</span>, <span class="st">"rpl16 "</span>, <span class="st">"rpl5 "</span>, <span class="st">"rps10"</span>, <span class="st">"rps11 "</span>, <span class="st">"rps12 "</span>, <span class="st">"rps13"</span>,</span>
<span id="cb1-137"><a href="#cb1-137"></a>    <span class="st">"rps14 "</span>, <span class="st">"rps19 "</span>, <span class="st">"rps2 "</span>, <span class="st">"rps3 "</span>, <span class="st">"rps4 "</span></span>
<span id="cb1-138"><a href="#cb1-138"></a>  )</span>
<span id="cb1-139"><a href="#cb1-139"></a>  </span>
<span id="cb1-140"><a href="#cb1-140"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s now plot the generated distances!</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true">Correlation of the common nodes</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false">Robinson–Foulds distance</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-3" role="tab" aria-controls="tabset-1-3" aria-selected="false">Kuhner-Felsenstein distance</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-4" role="tab" aria-controls="tabset-1-4" aria-selected="false">Maximum Agreement SubTree (MAST %)</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<div class="cell" data-result="hide">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="co">#!/usr/bin/env R</span></span>
<span id="cb2-2"><a href="#cb2-2"></a></span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="co"># prepare gene lists</span></span>
<span id="cb2-4"><a href="#cb2-4"></a>cp_genes_list <span class="ot">=</span> <span class="fu">c</span>(</span>
<span id="cb2-5"><a href="#cb2-5"></a>  <span class="st">"accD"</span>, <span class="st">"atpA"</span>, <span class="st">"atpB"</span>, <span class="st">"atpE"</span>, <span class="st">"atpF"</span>, <span class="st">"atpH"</span>, <span class="st">"atpI"</span>, <span class="st">"ccsA"</span>, <span class="st">"cemA"</span>, <span class="st">"chlI"</span>, <span class="st">"clpP"</span>,</span>
<span id="cb2-6"><a href="#cb2-6"></a>  <span class="st">"infA"</span>, <span class="st">"petA"</span>, <span class="st">"petB"</span>, <span class="st">"petD"</span>, <span class="st">"petG"</span>, <span class="st">"petL"</span>, <span class="st">"psaA"</span>, <span class="st">"psaB"</span>, <span class="st">"psaC"</span>, <span class="st">"psaI"</span>, <span class="st">"psaJ"</span>,</span>
<span id="cb2-7"><a href="#cb2-7"></a>  <span class="st">"psaM"</span>, <span class="st">"psbA"</span>, <span class="st">"psbB"</span>, <span class="st">"psbC"</span>, <span class="st">"psbD"</span>, <span class="st">"psbE"</span>, <span class="st">"psbF"</span>, <span class="st">"psbH"</span>, <span class="st">"psbI"</span>, <span class="st">"psbJ"</span>, <span class="st">"psbK"</span>,</span>
<span id="cb2-8"><a href="#cb2-8"></a>  <span class="st">"psbL"</span>, <span class="st">"psbM"</span>, <span class="st">"psbN"</span>, <span class="st">"psbT"</span>, <span class="st">"psbZ"</span>, <span class="st">"rbcL"</span>, <span class="st">"rpl12"</span>, <span class="st">"rpl14"</span>, <span class="st">"rpl16"</span>, <span class="st">"rpl19"</span>, <span class="st">"rpl2"</span>,</span>
<span id="cb2-9"><a href="#cb2-9"></a>  <span class="st">"rpl20"</span>, <span class="st">"rpl23"</span>, <span class="st">"rpl32"</span>, <span class="st">"rpl36"</span>, <span class="st">"rpl5"</span>, <span class="st">"rpoA"</span>, <span class="st">"rpoB"</span>, <span class="st">"rpoC1"</span>, <span class="st">"rpoC2"</span>, <span class="st">"rps11"</span>,</span>
<span id="cb2-10"><a href="#cb2-10"></a>  <span class="st">"rps12"</span>, <span class="st">"rps14"</span>, <span class="st">"rps18"</span>, <span class="st">"rps19"</span>, <span class="st">"rps2"</span>, <span class="st">"rps3"</span>,  <span class="st">"rps4"</span>, <span class="st">"rps7"</span>, <span class="st">"rps8"</span>, <span class="st">"rps9"</span>,</span>
<span id="cb2-11"><a href="#cb2-11"></a>  <span class="st">"tufA"</span>, <span class="st">"ycf1"</span>, <span class="st">"ycf12"</span>, <span class="st">"ycf20"</span>, <span class="st">"ycf3"</span>, <span class="st">"ycf4"</span></span>
<span id="cb2-12"><a href="#cb2-12"></a>)</span>
<span id="cb2-13"><a href="#cb2-13"></a></span>
<span id="cb2-14"><a href="#cb2-14"></a>mt_genes_list <span class="ot">=</span> <span class="fu">c</span>(</span>
<span id="cb2-15"><a href="#cb2-15"></a>  <span class="st">"atp1"</span>, <span class="st">"atp4"</span>, <span class="st">"atp6"</span>, <span class="st">"atp8"</span>, <span class="st">"atp9"</span>, <span class="st">"cob"</span>, <span class="st">"cox1"</span>, <span class="st">"cox2"</span>, <span class="st">"cox3"</span>, <span class="st">"nad1"</span>, <span class="st">"nad2"</span>,</span>
<span id="cb2-16"><a href="#cb2-16"></a>  <span class="st">"nad3"</span>, <span class="st">"nad4"</span>, <span class="st">"nad4L"</span>, <span class="st">"nad5"</span>, <span class="st">"nad6"</span>, <span class="st">"nad7"</span>, <span class="st">"rpl14 "</span>, <span class="st">"rpl16 "</span>, <span class="st">"rpl5 "</span>, <span class="st">"rps10"</span>,</span>
<span id="cb2-17"><a href="#cb2-17"></a>  <span class="st">"rps11 "</span>, <span class="st">"rps12 "</span>, <span class="st">"rps13"</span>, <span class="st">"rps14 "</span>, <span class="st">"rps19 "</span>, <span class="st">"rps2 "</span>, <span class="st">"rps3 "</span>, <span class="st">"rps4 "</span></span>
<span id="cb2-18"><a href="#cb2-18"></a>)</span>
<span id="cb2-19"><a href="#cb2-19"></a></span>
<span id="cb2-20"><a href="#cb2-20"></a><span class="co"># remove possible rows/columns with only NAs</span></span>
<span id="cb2-21"><a href="#cb2-21"></a>tmp_matrix <span class="ot">=</span> distance_sets[[<span class="st">"cp_mt_genes_distances"</span>]][[<span class="st">"common_nodes_corr"</span>]]</span>
<span id="cb2-22"><a href="#cb2-22"></a>tmp_matrix <span class="ot">=</span> tmp_matrix[, <span class="fu">colSums</span>(<span class="fu">is.na</span>(tmp_matrix)) <span class="sc">&lt;</span> <span class="fu">nrow</span>(tmp_matrix)]</span>
<span id="cb2-23"><a href="#cb2-23"></a>tmp_matrix <span class="ot">=</span> tmp_matrix[<span class="fu">rowSums</span>(<span class="fu">is.na</span>(tmp_matrix)) <span class="sc">&lt;</span> <span class="fu">ncol</span>(tmp_matrix), ]</span>
<span id="cb2-24"><a href="#cb2-24"></a></span>
<span id="cb2-25"><a href="#cb2-25"></a><span class="co"># prep annotation</span></span>
<span id="cb2-26"><a href="#cb2-26"></a>organell <span class="ot">=</span> <span class="fu">ifelse</span>(<span class="fu">rownames</span>(tmp_matrix) <span class="sc">%in%</span> cp_genes_list, <span class="st">"cp"</span>,</span>
<span id="cb2-27"><a href="#cb2-27"></a>                  <span class="fu">ifelse</span>(<span class="fu">rownames</span>(tmp_matrix) <span class="sc">%in%</span> mt_genes_list, <span class="st">"mt"</span>, <span class="st">"cp_mt"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb2-28"><a href="#cb2-28"></a>              <span class="fu">as.data.frame</span>()</span>
<span id="cb2-29"><a href="#cb2-29"></a><span class="fu">rownames</span>(organell) <span class="ot">=</span> <span class="fu">rownames</span>(tmp_matrix)</span>
<span id="cb2-30"><a href="#cb2-30"></a><span class="fu">colnames</span>(organell) <span class="ot">=</span> <span class="st">"organell"</span></span>
<span id="cb2-31"><a href="#cb2-31"></a>organell<span class="sc">$</span>color <span class="ot">=</span> <span class="fu">ifelse</span>(organell<span class="sc">$</span>organell <span class="sc">==</span> <span class="st">"cp"</span>, <span class="st">"#469d89"</span>, <span class="fu">ifelse</span>(organell<span class="sc">$</span>organell <span class="sc">==</span> <span class="st">"mt"</span>, <span class="st">"steelblue"</span>, <span class="st">"red"</span>))</span>
<span id="cb2-32"><a href="#cb2-32"></a></span>
<span id="cb2-33"><a href="#cb2-33"></a><span class="co"># plot raw distance</span></span>
<span id="cb2-34"><a href="#cb2-34"></a>ComplexHeatmap<span class="sc">::</span><span class="fu">Heatmap</span>(</span>
<span id="cb2-35"><a href="#cb2-35"></a>  tmp_matrix,</span>
<span id="cb2-36"><a href="#cb2-36"></a>  <span class="at">col =</span> <span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="fu">min</span>(tmp_matrix[<span class="sc">!</span><span class="fu">is.na</span>(tmp_matrix)]), <span class="fu">max</span>(tmp_matrix[<span class="sc">!</span><span class="fu">is.na</span>(tmp_matrix)])),</span>
<span id="cb2-37"><a href="#cb2-37"></a>                   <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"steelblue"</span>)),</span>
<span id="cb2-38"><a href="#cb2-38"></a>  <span class="at">row_names_gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> organell<span class="sc">$</span>color),</span>
<span id="cb2-39"><a href="#cb2-39"></a>  <span class="at">column_names_gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> organell<span class="sc">$</span>color),</span>
<span id="cb2-40"><a href="#cb2-40"></a>  <span class="at">top_annotation =</span> <span class="fu">HeatmapAnnotation</span>(</span>
<span id="cb2-41"><a href="#cb2-41"></a>    <span class="at">organell =</span> <span class="fu">as.matrix</span>(organell<span class="sc">$</span>organell),</span>
<span id="cb2-42"><a href="#cb2-42"></a>    <span class="at">show_annotation_name =</span> <span class="cn">FALSE</span>,</span>
<span id="cb2-43"><a href="#cb2-43"></a>    <span class="at">show_legend =</span> <span class="cn">FALSE</span>,</span>
<span id="cb2-44"><a href="#cb2-44"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="at">organell =</span> <span class="fu">c</span>(<span class="st">"cp_mt"</span> <span class="ot">=</span> <span class="st">"red"</span>, <span class="st">"cp"</span> <span class="ot">=</span> <span class="st">"#469d89"</span>, <span class="st">"mt"</span> <span class="ot">=</span> <span class="st">"steelblue"</span>))),</span>
<span id="cb2-45"><a href="#cb2-45"></a>  <span class="at">left_annotation =</span> <span class="fu">rowAnnotation</span>(</span>
<span id="cb2-46"><a href="#cb2-46"></a>    <span class="at">organell =</span> <span class="fu">as.matrix</span>(organell<span class="sc">$</span>organell),</span>
<span id="cb2-47"><a href="#cb2-47"></a>    <span class="at">show_annotation_name =</span> <span class="cn">FALSE</span>,</span>
<span id="cb2-48"><a href="#cb2-48"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="at">organell =</span> <span class="fu">c</span>(<span class="st">"cp_mt"</span> <span class="ot">=</span> <span class="st">"red"</span>, <span class="st">"cp"</span> <span class="ot">=</span> <span class="st">"#469d89"</span>, <span class="st">"mt"</span> <span class="ot">=</span> <span class="st">"steelblue"</span>))),</span>
<span id="cb2-49"><a href="#cb2-49"></a>  <span class="at">name =</span> <span class="st">"common</span><span class="sc">\n</span><span class="st">nodes</span><span class="sc">\n</span><span class="st">corr"</span></span>
<span id="cb2-50"><a href="#cb2-50"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="markers-selection_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption>Figure 1: cp + mt species tree reconstruction</figcaption>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<div class="cell" data-result="hide">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="co">#!/usr/bin/env R</span></span>
<span id="cb3-2"><a href="#cb3-2"></a></span>
<span id="cb3-3"><a href="#cb3-3"></a></span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="co"># remove possible rows/columns with only NAs</span></span>
<span id="cb3-5"><a href="#cb3-5"></a>tmp_matrix <span class="ot">=</span> distance_sets[[<span class="st">"cp_mt_genes_distances"</span>]][[<span class="st">"RF"</span>]]</span>
<span id="cb3-6"><a href="#cb3-6"></a>tmp_matrix <span class="ot">=</span> tmp_matrix[, <span class="fu">colSums</span>(<span class="fu">is.na</span>(tmp_matrix)) <span class="sc">&lt;</span> <span class="fu">nrow</span>(tmp_matrix)]</span>
<span id="cb3-7"><a href="#cb3-7"></a>tmp_matrix <span class="ot">=</span> tmp_matrix[<span class="fu">rowSums</span>(<span class="fu">is.na</span>(tmp_matrix)) <span class="sc">&lt;</span> <span class="fu">ncol</span>(tmp_matrix), ]</span>
<span id="cb3-8"><a href="#cb3-8"></a></span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="co"># prep annotation</span></span>
<span id="cb3-10"><a href="#cb3-10"></a>organell <span class="ot">=</span> <span class="fu">ifelse</span>(<span class="fu">rownames</span>(tmp_matrix) <span class="sc">%in%</span> cp_genes_list, <span class="st">"cp"</span>,</span>
<span id="cb3-11"><a href="#cb3-11"></a>                  <span class="fu">ifelse</span>(<span class="fu">rownames</span>(tmp_matrix) <span class="sc">%in%</span> mt_genes_list, <span class="st">"mt"</span>, <span class="st">"cp_mt"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb3-12"><a href="#cb3-12"></a>              <span class="fu">as.data.frame</span>()</span>
<span id="cb3-13"><a href="#cb3-13"></a><span class="fu">rownames</span>(organell) <span class="ot">=</span> <span class="fu">rownames</span>(tmp_matrix)</span>
<span id="cb3-14"><a href="#cb3-14"></a><span class="fu">colnames</span>(organell) <span class="ot">=</span> <span class="st">"organell"</span></span>
<span id="cb3-15"><a href="#cb3-15"></a>organell<span class="sc">$</span>color <span class="ot">=</span> <span class="fu">ifelse</span>(organell<span class="sc">$</span>organell <span class="sc">==</span> <span class="st">"cp"</span>, <span class="st">"#469d89"</span>, <span class="fu">ifelse</span>(organell<span class="sc">$</span>organell <span class="sc">==</span> <span class="st">"mt"</span>, <span class="st">"steelblue"</span>, <span class="st">"red"</span>))</span>
<span id="cb3-16"><a href="#cb3-16"></a></span>
<span id="cb3-17"><a href="#cb3-17"></a><span class="co"># plot raw distance</span></span>
<span id="cb3-18"><a href="#cb3-18"></a>ComplexHeatmap<span class="sc">::</span><span class="fu">Heatmap</span>(</span>
<span id="cb3-19"><a href="#cb3-19"></a>  tmp_matrix,</span>
<span id="cb3-20"><a href="#cb3-20"></a>  <span class="at">col =</span> <span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="fu">min</span>(tmp_matrix[<span class="sc">!</span><span class="fu">is.na</span>(tmp_matrix)]), <span class="fu">max</span>(tmp_matrix[<span class="sc">!</span><span class="fu">is.na</span>(tmp_matrix)])),</span>
<span id="cb3-21"><a href="#cb3-21"></a>                   <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"#D55E00"</span>)),</span>
<span id="cb3-22"><a href="#cb3-22"></a>  <span class="at">row_names_gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> organell<span class="sc">$</span>color),</span>
<span id="cb3-23"><a href="#cb3-23"></a>  <span class="at">column_names_gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> organell<span class="sc">$</span>color),</span>
<span id="cb3-24"><a href="#cb3-24"></a>  <span class="at">top_annotation =</span> <span class="fu">HeatmapAnnotation</span>(</span>
<span id="cb3-25"><a href="#cb3-25"></a>    <span class="at">organell =</span> <span class="fu">as.matrix</span>(organell<span class="sc">$</span>organell),</span>
<span id="cb3-26"><a href="#cb3-26"></a>    <span class="at">show_annotation_name =</span> <span class="cn">FALSE</span>,</span>
<span id="cb3-27"><a href="#cb3-27"></a>    <span class="at">show_legend =</span> <span class="cn">FALSE</span>,</span>
<span id="cb3-28"><a href="#cb3-28"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="at">organell =</span> <span class="fu">c</span>(<span class="st">"cp_mt"</span> <span class="ot">=</span> <span class="st">"red"</span>, <span class="st">"cp"</span> <span class="ot">=</span> <span class="st">"#469d89"</span>, <span class="st">"mt"</span> <span class="ot">=</span> <span class="st">"steelblue"</span>))),</span>
<span id="cb3-29"><a href="#cb3-29"></a>  <span class="at">left_annotation =</span> <span class="fu">rowAnnotation</span>(</span>
<span id="cb3-30"><a href="#cb3-30"></a>    <span class="at">organell =</span> <span class="fu">as.matrix</span>(organell<span class="sc">$</span>organell),</span>
<span id="cb3-31"><a href="#cb3-31"></a>    <span class="at">show_annotation_name =</span> <span class="cn">FALSE</span>,</span>
<span id="cb3-32"><a href="#cb3-32"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="at">organell =</span> <span class="fu">c</span>(<span class="st">"cp_mt"</span> <span class="ot">=</span> <span class="st">"red"</span>, <span class="st">"cp"</span> <span class="ot">=</span> <span class="st">"#469d89"</span>, <span class="st">"mt"</span> <span class="ot">=</span> <span class="st">"steelblue"</span>))),</span>
<span id="cb3-33"><a href="#cb3-33"></a>  <span class="at">name =</span> <span class="st">"Robinson–Foulds"</span></span>
<span id="cb3-34"><a href="#cb3-34"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="markers-selection_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption>Figure 1: cp + mt species tree reconstruction</figcaption>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-1-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-3-tab">
<div class="cell" data-result="hide">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="co">#!/usr/bin/env R</span></span>
<span id="cb4-2"><a href="#cb4-2"></a></span>
<span id="cb4-3"><a href="#cb4-3"></a></span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="co"># remove possible rows/columns with only NAs</span></span>
<span id="cb4-5"><a href="#cb4-5"></a>tmp_matrix <span class="ot">=</span> distance_sets[[<span class="st">"cp_mt_genes_distances"</span>]][[<span class="st">"KF"</span>]]</span>
<span id="cb4-6"><a href="#cb4-6"></a>tmp_matrix <span class="ot">=</span> tmp_matrix[, <span class="fu">colSums</span>(<span class="fu">is.na</span>(tmp_matrix)) <span class="sc">&lt;</span> <span class="fu">nrow</span>(tmp_matrix)]</span>
<span id="cb4-7"><a href="#cb4-7"></a>tmp_matrix <span class="ot">=</span> tmp_matrix[<span class="fu">rowSums</span>(<span class="fu">is.na</span>(tmp_matrix)) <span class="sc">&lt;</span> <span class="fu">ncol</span>(tmp_matrix), ]</span>
<span id="cb4-8"><a href="#cb4-8"></a></span>
<span id="cb4-9"><a href="#cb4-9"></a><span class="co"># prep annotation</span></span>
<span id="cb4-10"><a href="#cb4-10"></a>organell <span class="ot">=</span> <span class="fu">ifelse</span>(<span class="fu">rownames</span>(tmp_matrix) <span class="sc">%in%</span> cp_genes_list, <span class="st">"cp"</span>,</span>
<span id="cb4-11"><a href="#cb4-11"></a>                  <span class="fu">ifelse</span>(<span class="fu">rownames</span>(tmp_matrix) <span class="sc">%in%</span> mt_genes_list, <span class="st">"mt"</span>, <span class="st">"cp_mt"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb4-12"><a href="#cb4-12"></a>              <span class="fu">as.data.frame</span>()</span>
<span id="cb4-13"><a href="#cb4-13"></a><span class="fu">rownames</span>(organell) <span class="ot">=</span> <span class="fu">rownames</span>(tmp_matrix)</span>
<span id="cb4-14"><a href="#cb4-14"></a><span class="fu">colnames</span>(organell) <span class="ot">=</span> <span class="st">"organell"</span></span>
<span id="cb4-15"><a href="#cb4-15"></a>organell<span class="sc">$</span>color <span class="ot">=</span> <span class="fu">ifelse</span>(organell<span class="sc">$</span>organell <span class="sc">==</span> <span class="st">"cp"</span>, <span class="st">"#469d89"</span>, <span class="fu">ifelse</span>(organell<span class="sc">$</span>organell <span class="sc">==</span> <span class="st">"mt"</span>, <span class="st">"steelblue"</span>, <span class="st">"red"</span>))</span>
<span id="cb4-16"><a href="#cb4-16"></a></span>
<span id="cb4-17"><a href="#cb4-17"></a><span class="co"># plot raw distance</span></span>
<span id="cb4-18"><a href="#cb4-18"></a>ComplexHeatmap<span class="sc">::</span><span class="fu">Heatmap</span>(</span>
<span id="cb4-19"><a href="#cb4-19"></a>  tmp_matrix,</span>
<span id="cb4-20"><a href="#cb4-20"></a>  <span class="at">col =</span> <span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="fu">min</span>(tmp_matrix[<span class="sc">!</span><span class="fu">is.na</span>(tmp_matrix)]), <span class="fu">max</span>(tmp_matrix[<span class="sc">!</span><span class="fu">is.na</span>(tmp_matrix)])),</span>
<span id="cb4-21"><a href="#cb4-21"></a>                   <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"#D55E00"</span>)),</span>
<span id="cb4-22"><a href="#cb4-22"></a>  <span class="at">row_names_gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> organell<span class="sc">$</span>color),</span>
<span id="cb4-23"><a href="#cb4-23"></a>  <span class="at">column_names_gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> organell<span class="sc">$</span>color),</span>
<span id="cb4-24"><a href="#cb4-24"></a>  <span class="at">top_annotation =</span> <span class="fu">HeatmapAnnotation</span>(</span>
<span id="cb4-25"><a href="#cb4-25"></a>    <span class="at">organell =</span> <span class="fu">as.matrix</span>(organell<span class="sc">$</span>organell),</span>
<span id="cb4-26"><a href="#cb4-26"></a>    <span class="at">show_annotation_name =</span> <span class="cn">FALSE</span>,</span>
<span id="cb4-27"><a href="#cb4-27"></a>    <span class="at">show_legend =</span> <span class="cn">FALSE</span>,</span>
<span id="cb4-28"><a href="#cb4-28"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="at">organell =</span> <span class="fu">c</span>(<span class="st">"cp_mt"</span> <span class="ot">=</span> <span class="st">"red"</span>, <span class="st">"cp"</span> <span class="ot">=</span> <span class="st">"#469d89"</span>, <span class="st">"mt"</span> <span class="ot">=</span> <span class="st">"steelblue"</span>))),</span>
<span id="cb4-29"><a href="#cb4-29"></a>  <span class="at">left_annotation =</span> <span class="fu">rowAnnotation</span>(</span>
<span id="cb4-30"><a href="#cb4-30"></a>    <span class="at">organell =</span> <span class="fu">as.matrix</span>(organell<span class="sc">$</span>organell),</span>
<span id="cb4-31"><a href="#cb4-31"></a>    <span class="at">show_annotation_name =</span> <span class="cn">FALSE</span>,</span>
<span id="cb4-32"><a href="#cb4-32"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="at">organell =</span> <span class="fu">c</span>(<span class="st">"cp_mt"</span> <span class="ot">=</span> <span class="st">"red"</span>, <span class="st">"cp"</span> <span class="ot">=</span> <span class="st">"#469d89"</span>, <span class="st">"mt"</span> <span class="ot">=</span> <span class="st">"steelblue"</span>))),</span>
<span id="cb4-33"><a href="#cb4-33"></a>  <span class="at">name =</span> <span class="st">"Kuhner-Felsenstein</span><span class="sc">\n</span><span class="st">distance"</span></span>
<span id="cb4-34"><a href="#cb4-34"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="markers-selection_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption>Figure 1: cp + mt species tree reconstruction</figcaption>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-1-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-4-tab">
<div class="cell" data-result="hide">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="co">#!/usr/bin/env R</span></span>
<span id="cb5-2"><a href="#cb5-2"></a></span>
<span id="cb5-3"><a href="#cb5-3"></a></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="co"># remove possible rows/columns with only NAs</span></span>
<span id="cb5-5"><a href="#cb5-5"></a>tmp_matrix <span class="ot">=</span> distance_sets[[<span class="st">"cp_mt_genes_distances"</span>]][[<span class="st">"MAST_%"</span>]]</span>
<span id="cb5-6"><a href="#cb5-6"></a>tmp_matrix <span class="ot">=</span> tmp_matrix[, <span class="fu">colSums</span>(<span class="fu">is.na</span>(tmp_matrix)) <span class="sc">&lt;</span> <span class="fu">nrow</span>(tmp_matrix)]</span>
<span id="cb5-7"><a href="#cb5-7"></a>tmp_matrix <span class="ot">=</span> tmp_matrix[<span class="fu">rowSums</span>(<span class="fu">is.na</span>(tmp_matrix)) <span class="sc">&lt;</span> <span class="fu">ncol</span>(tmp_matrix), ]</span>
<span id="cb5-8"><a href="#cb5-8"></a></span>
<span id="cb5-9"><a href="#cb5-9"></a><span class="co"># prep annotation</span></span>
<span id="cb5-10"><a href="#cb5-10"></a>organell <span class="ot">=</span> <span class="fu">ifelse</span>(<span class="fu">rownames</span>(tmp_matrix) <span class="sc">%in%</span> cp_genes_list, <span class="st">"cp"</span>,</span>
<span id="cb5-11"><a href="#cb5-11"></a>                  <span class="fu">ifelse</span>(<span class="fu">rownames</span>(tmp_matrix) <span class="sc">%in%</span> mt_genes_list, <span class="st">"mt"</span>, <span class="st">"cp_mt"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb5-12"><a href="#cb5-12"></a>              <span class="fu">as.data.frame</span>()</span>
<span id="cb5-13"><a href="#cb5-13"></a><span class="fu">rownames</span>(organell) <span class="ot">=</span> <span class="fu">rownames</span>(tmp_matrix)</span>
<span id="cb5-14"><a href="#cb5-14"></a><span class="fu">colnames</span>(organell) <span class="ot">=</span> <span class="st">"organell"</span></span>
<span id="cb5-15"><a href="#cb5-15"></a>organell<span class="sc">$</span>color <span class="ot">=</span> <span class="fu">ifelse</span>(organell<span class="sc">$</span>organell <span class="sc">==</span> <span class="st">"cp"</span>, <span class="st">"#469d89"</span>, <span class="fu">ifelse</span>(organell<span class="sc">$</span>organell <span class="sc">==</span> <span class="st">"mt"</span>, <span class="st">"steelblue"</span>, <span class="st">"red"</span>))</span>
<span id="cb5-16"><a href="#cb5-16"></a></span>
<span id="cb5-17"><a href="#cb5-17"></a><span class="co"># plot raw distance</span></span>
<span id="cb5-18"><a href="#cb5-18"></a>ComplexHeatmap<span class="sc">::</span><span class="fu">Heatmap</span>(</span>
<span id="cb5-19"><a href="#cb5-19"></a>  tmp_matrix,</span>
<span id="cb5-20"><a href="#cb5-20"></a>  <span class="at">col =</span> <span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="fu">min</span>(tmp_matrix[<span class="sc">!</span><span class="fu">is.na</span>(tmp_matrix)]), <span class="fu">max</span>(tmp_matrix[<span class="sc">!</span><span class="fu">is.na</span>(tmp_matrix)])),</span>
<span id="cb5-21"><a href="#cb5-21"></a>                   <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"steelblue"</span>)),</span>
<span id="cb5-22"><a href="#cb5-22"></a>  <span class="at">row_names_gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> organell<span class="sc">$</span>color),</span>
<span id="cb5-23"><a href="#cb5-23"></a>  <span class="at">column_names_gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> organell<span class="sc">$</span>color),</span>
<span id="cb5-24"><a href="#cb5-24"></a>  <span class="at">top_annotation =</span> <span class="fu">HeatmapAnnotation</span>(</span>
<span id="cb5-25"><a href="#cb5-25"></a>    <span class="at">organell =</span> <span class="fu">as.matrix</span>(organell<span class="sc">$</span>organell),</span>
<span id="cb5-26"><a href="#cb5-26"></a>    <span class="at">show_annotation_name =</span> <span class="cn">FALSE</span>,</span>
<span id="cb5-27"><a href="#cb5-27"></a>    <span class="at">show_legend =</span> <span class="cn">FALSE</span>,</span>
<span id="cb5-28"><a href="#cb5-28"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="at">organell =</span> <span class="fu">c</span>(<span class="st">"cp_mt"</span> <span class="ot">=</span> <span class="st">"red"</span>, <span class="st">"cp"</span> <span class="ot">=</span> <span class="st">"#469d89"</span>, <span class="st">"mt"</span> <span class="ot">=</span> <span class="st">"steelblue"</span>))),</span>
<span id="cb5-29"><a href="#cb5-29"></a>  <span class="at">left_annotation =</span> <span class="fu">rowAnnotation</span>(</span>
<span id="cb5-30"><a href="#cb5-30"></a>    <span class="at">organell =</span> <span class="fu">as.matrix</span>(organell<span class="sc">$</span>organell),</span>
<span id="cb5-31"><a href="#cb5-31"></a>    <span class="at">show_annotation_name =</span> <span class="cn">FALSE</span>,</span>
<span id="cb5-32"><a href="#cb5-32"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="at">organell =</span> <span class="fu">c</span>(<span class="st">"cp_mt"</span> <span class="ot">=</span> <span class="st">"red"</span>, <span class="st">"cp"</span> <span class="ot">=</span> <span class="st">"#469d89"</span>, <span class="st">"mt"</span> <span class="ot">=</span> <span class="st">"steelblue"</span>))),</span>
<span id="cb5-33"><a href="#cb5-33"></a>  <span class="at">name =</span> <span class="st">"MAST_%"</span></span>
<span id="cb5-34"><a href="#cb5-34"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="markers-selection_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption>Figure 1: cp + mt species tree reconstruction</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<p>The top 10 single genes that produce similar ML tree topology to the Ulva species tree reconstructed from supermatrix analysis. The genes are ordered based on the decreasing values of the corresponding index.</p>
<div class="cell">
<div class="cell-output-display">
<div class="datatables html-widget html-fill-item" id="htmlwidget-75ae5db211406dc59518" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-75ae5db211406dc59518">{"x":{"filter":"none","vertical":false,"class":"display","extensions":["FixedColumns","FixedHeader"],"caption":"<caption>Table 1: top 10 genes closest to the reconstructed Ulva species tree.<\/caption>","data":[["1","2","3","4","5","6","7","8","9","10"],["psaA","rpoC2","rpoA","rpoC1","psaB","psbB","rps12","ycf1","chlI","rps2"],["psbB","psaA","rps2","petB","rpoC1","psaB","psbD","rpl12","rpoC2","rpoA"],["rpoA","ycf20","rpoB","rpl32","rps8","accD","atpF","rps4","rps9","rpl2"],["psaA","psaB","ycf1","chlI","rpoA","psbB","rpl19","tufA","rpoC1","accD"],["rps10","rps3","cox1","rps2","atp1","atp6","cox2","nad4","rps4","nad2"],["cox1","cox2","atp1","cox3","rps4","rps2","rps3","cob","rps19","rps10"],["rps3","rps10","rps2","nad5","nad7","cox1","rpl16","rpl14","rps14","rps4"],["cox2","atp6","cox3","nad5","rpl16","rps11","rps4","nad2","nad4","cox1"],["psaA","rpoC2","rpoA","rpoC1","psaB","psbB","rps12","ycf1","chlI","rps2"],["psbB","psaA","rps2","cox1","cox2","petB","rpoC1","psaB","psbD","rpl12"],["mt_rps3","mt_rps10","rpoA","mt_rps2","nad5","ycf20","rpoB","rpl32","rps8","accD"],["psaA","psaB","ycf1","cox2","chlI","rpoA","psbB","rpl19","tufA","atp6"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th rowspan=\"2\">Rank<\/th>\n      <th colspan=\"4\">chloroplast<\/th>\n      <th colspan=\"4\">mitochondria<\/th>\n      <th colspan=\"4\">chloroplast + mitochondria<\/th>\n    <\/tr>\n    <tr>\n      <th>RF<\/th>\n      <th>MAST %<\/th>\n      <th>KF<\/th>\n      <th>comm. node corr.<\/th>\n      <th>RF<\/th>\n      <th>MAST %<\/th>\n      <th>KF<\/th>\n      <th>comm. node corr.<\/th>\n      <th>RF<\/th>\n      <th>MAST %<\/th>\n      <th>KF<\/th>\n      <th>comm. node corr.<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":false,"pageLength":10,"columnDefs":[{"name":"rank","targets":0},{"name":"cp.RF","targets":1},{"name":"cp.MAST.","targets":2},{"name":"cp.KF","targets":3},{"name":"cp.corr","targets":4},{"name":"mt.RF","targets":5},{"name":"mt.MAST.","targets":6},{"name":"mt.KF","targets":7},{"name":"mt.corr","targets":8},{"name":"cp_mt.RF","targets":9},{"name":"cp_mt.MAST.","targets":10},{"name":"cp_mt.KF","targets":11},{"name":"cp_mt.corr","targets":12}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<p>Ide</p>
<p>Among the 70 cp genes chosen, the four indices RF, KF, MAST and the correlation between the common nodes metrics pointed out 24 unique cp genes that provide the most similar topology compared to the species tree constructed using 70 genes, namely accD, atpF, chlI, petB, psaA, psaB, psbB, psbD, rpl12, rpl19, rpl2, rpl32, rpoA, rpoB, rpoC1, rpoC2, rps12, rps2, rps4, rps8, rps9, tufA, ycf1, and ycf20 (Table 1). Among them, several genes such as psaA, psaB, and psbB are consistently identified by different matrices. Similarly, the same approach applied to the Ulva species trees reconstructed using mt suggested 19 unique mt genes, namely atp1, atp6, cob, cox1, cox2, cox3, nad2, nad4, nad5, nad7, rpl14, rpl16, rps10, rps11, rps14, rps19, rps2, rps3, and rps4, of which several genes such as atp6, cox1, cox2, rps3, and rps10 are suggested by different indices. Since the Ulva specimens used in this study do not always have both cp and mt genomes, the Ulva species trees reconstructed using cp and mt supermatrix have different set of tips, depending on the availability of the genomes. Prior analyses suggested that MAST% and common correlation nodes metrics are the best performers to assess phylogenetic tree distances when the tree do not have the same set of tips, we then selected (arbitrarily) 10 genes for the combinatorial approach, petB, psaA, psaB, psbB, psbD, rps2, atp6, cox1, cox2, and mt_rps3, based on the top 10 single gene ML trees with topology closest to cp_mt_ML. The first six genes are cp genes while the other four genes are mt genes.</p>
<p>Table 1: The top 10 single genes that produce similar ML tree topology to the Ulva species tree reconstructed from supermatrix analysis. The genes are ordered based on the decreasing values of the corresponding index.</p>
<p>cp cp_mt mt RF MAST% KF cor RF MAST% KF cor RF MAST% KF cor psaA psbB rpoA psaA psaA psbB mt_rps3 psaA rps10 cox1 rps3 cox2 rpoC2 psaA ycf20 psaB rpoC2 psaA mt_rps10 psaB rps3 cox2 rps10 atp6 rpoA rps2 rpoB ycf1 rpoA rps2 rpoA ycf1 cox1 atp1 rps2 cox3 rpoC1 petB rpl32 chlI rpoC1 cox1 mt_rps2 cox2 rps2 cox3 nad5 nad5 psaB rpoC1 rps8 rpoA psaB cox2 nad5 chlI atp1 rps4 nad7 rpl16 psbB psaB accD psbB psbB petB ycf20 rpoA atp6 rps2 cox1 rps11 rps12 psbD atpF rpl19 rps12 rpoC1 rpoB psbB cox2 rps3 rpl16 rps4 ycf1 rpl12 rps4 tufA ycf1 psaB rpl32 rpl19 nad4 cob rpl14 nad2 chlI rpoC2 rps9 rpoC1 chlI psbD rps8 tufA rps4 rps19 rps14 nad4 rps2 rpoA rpl2 accD rps2 rpl12 accD atp6 nad2 rps10 rps4 cox1</p>
<p>Ide</p>
</section>
</section>
<section id="markers-selection" class="level2" data-number="3.3">
<h2 data-number="3.3" class="anchored" data-anchor-id="markers-selection"><span class="header-section-number">3.3</span> Markers selection</h2>
<p>We use a combinatorial approach to select the minimum list of genes to approxima te the CP+MT species three</p>
<p>cp_genes = c(“petB”, “psaA”, “psaB”, “psbB”, “psbD”, “rps2”) mt_genes = c(“atp6”, “cox1”, “cox2”, “rps3”)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1"></a></span>
<span id="cb6-2"><a href="#cb6-2"></a></span>
<span id="cb6-3"><a href="#cb6-3"></a></span>
<span id="cb6-4"><a href="#cb6-4"></a></span>
<span id="cb6-5"><a href="#cb6-5"></a>        <span class="ex">python3.10</span> combinatorial_ML_phylogeny.py <span class="at">--infolder</span> <span class="st">"ANALYSES_AND_RESULTS/</span></span>
<span id="cb6-6"><a href="#cb6-6"></a><span class="st">09_combinatorial_phylogeny/00_cp_genes/,ANALYSES_AND_RESULTS/09_combinatorial_phylogeny/00_mt_genes/"</span> <span class="at">--names</span> <span class="st">"CP,MT"</span> <span class="at">--iqtree</span> /home/Andrea/Desktop/ANALYSES_AND_RESULTS/bin/iqtree-2.2.0-Linux/bin/iqtree2 <span class="at">--n_genes</span> 10 <span class="at">--outdir</span> /home/Andrea/Desktop/CP_MT/ANALYSES_AND_RESULTS/09_combinatorial_phylogeny/</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="cp-only" class="level3" data-number="3.3.1">
<h3 data-number="3.3.1" class="anchored" data-anchor-id="cp-only"><span class="header-section-number">3.3.1</span> cp only</h3>
<p>ss</p>
</section>
<section id="mt-only" class="level3" data-number="3.3.2">
<h3 data-number="3.3.2" class="anchored" data-anchor-id="mt-only"><span class="header-section-number">3.3.2</span> mt only</h3>
<p>ss</p>
</section>
<section id="cp-mt" class="level3" data-number="3.3.3">
<h3 data-number="3.3.3" class="anchored" data-anchor-id="cp-mt"><span class="header-section-number">3.3.3</span> cp + mt</h3>
<p>ss</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="co"># declare gene lists</span></span>
<span id="cb7-2"><a href="#cb7-2"></a>cp_genes <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"petB"</span>, <span class="st">"psaA"</span>, <span class="st">"psaB"</span>, <span class="st">"psbB"</span>, <span class="st">"psbD"</span>, <span class="st">"rps2"</span>)</span>
<span id="cb7-3"><a href="#cb7-3"></a>mt_genes <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"atp6"</span>, <span class="st">"cox1"</span>, <span class="st">"cox2"</span>, <span class="st">"rps3"</span>)</span>
<span id="cb7-4"><a href="#cb7-4"></a></span>
<span id="cb7-5"><a href="#cb7-5"></a><span class="co"># import ref species tree</span></span>
<span id="cb7-6"><a href="#cb7-6"></a>cp_mt_ML_dendro <span class="ot">=</span> ape<span class="sc">::</span><span class="fu">read.tree</span>(<span class="at">file =</span> <span class="st">"06_cp_mt_concat_ML/cp_mt_allgenes_concat.contree"</span>)</span>
<span id="cb7-7"><a href="#cb7-7"></a></span>
<span id="cb7-8"><a href="#cb7-8"></a><span class="co"># create empty list of trees</span></span>
<span id="cb7-9"><a href="#cb7-9"></a>combinatorial_trees <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_conc_2_genes"</span> <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_ML"</span> <span class="ot">=</span> cp_mt_ML_dendro),</span>
<span id="cb7-10"><a href="#cb7-10"></a>                           <span class="st">"cp_mt_conc_3_genes"</span> <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_ML"</span> <span class="ot">=</span> cp_mt_ML_dendro),</span>
<span id="cb7-11"><a href="#cb7-11"></a>                           <span class="st">"cp_mt_conc_4_genes"</span> <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_ML"</span> <span class="ot">=</span> cp_mt_ML_dendro),</span>
<span id="cb7-12"><a href="#cb7-12"></a>                           <span class="st">"cp_mt_conc_5_genes"</span> <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_ML"</span> <span class="ot">=</span> cp_mt_ML_dendro),</span>
<span id="cb7-13"><a href="#cb7-13"></a>                           <span class="st">"cp_mt_conc_6_genes"</span> <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_ML"</span> <span class="ot">=</span> cp_mt_ML_dendro),</span>
<span id="cb7-14"><a href="#cb7-14"></a>                           <span class="st">"cp_mt_conc_7_genes"</span> <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_ML"</span> <span class="ot">=</span> cp_mt_ML_dendro),</span>
<span id="cb7-15"><a href="#cb7-15"></a>                           <span class="st">"cp_mt_conc_8_genes"</span> <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_ML"</span> <span class="ot">=</span> cp_mt_ML_dendro),</span>
<span id="cb7-16"><a href="#cb7-16"></a>                           <span class="st">"cp_mt_conc_9_genes"</span> <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_ML"</span> <span class="ot">=</span> cp_mt_ML_dendro),</span>
<span id="cb7-17"><a href="#cb7-17"></a>                           <span class="st">"cp_mt_conc_10_genes"</span> <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_ML"</span> <span class="ot">=</span> cp_mt_ML_dendro))</span>
<span id="cb7-18"><a href="#cb7-18"></a></span>
<span id="cb7-19"><a href="#cb7-19"></a><span class="co"># populate combinatorial trees</span></span>
<span id="cb7-20"><a href="#cb7-20"></a><span class="cf">for</span>(k <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>){</span>
<span id="cb7-21"><a href="#cb7-21"></a>  </span>
<span id="cb7-22"><a href="#cb7-22"></a>  <span class="co"># get file list</span></span>
<span id="cb7-23"><a href="#cb7-23"></a>  tree_list_all <span class="ot">=</span> <span class="fu">list.files</span>(<span class="at">path =</span> <span class="fu">paste</span>(mainDir, <span class="st">"/09_combinatorial_phylogeny/01_cp_mt_combinatorial/cp_mt_conc_"</span>, k, <span class="st">"_genes"</span>, <span class="at">sep =</span> <span class="st">""</span>),</span>
<span id="cb7-24"><a href="#cb7-24"></a>                             <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">.contree$"</span>)</span>
<span id="cb7-25"><a href="#cb7-25"></a>  <span class="co"># keep only trees with both cp and mt markers</span></span>
<span id="cb7-26"><a href="#cb7-26"></a>  tree_list_fltr <span class="ot">=</span> <span class="fu">c</span>()</span>
<span id="cb7-27"><a href="#cb7-27"></a>  <span class="cf">for</span>(tree <span class="cf">in</span> tree_list_all){</span>
<span id="cb7-28"><a href="#cb7-28"></a>    cp <span class="ot">=</span> <span class="cn">FALSE</span></span>
<span id="cb7-29"><a href="#cb7-29"></a>    mt <span class="ot">=</span> <span class="cn">FALSE</span></span>
<span id="cb7-30"><a href="#cb7-30"></a>    <span class="co"># check if cp is present</span></span>
<span id="cb7-31"><a href="#cb7-31"></a>    <span class="cf">for</span>(gene <span class="cf">in</span> cp_genes){</span>
<span id="cb7-32"><a href="#cb7-32"></a>      <span class="cf">if</span>(<span class="fu">grepl</span>(gene, tree, <span class="at">fixed =</span> <span class="cn">TRUE</span>)){</span>
<span id="cb7-33"><a href="#cb7-33"></a>        cp <span class="ot">=</span> <span class="cn">TRUE</span></span>
<span id="cb7-34"><a href="#cb7-34"></a>      }</span>
<span id="cb7-35"><a href="#cb7-35"></a>    }</span>
<span id="cb7-36"><a href="#cb7-36"></a>    <span class="co"># check if mt is present</span></span>
<span id="cb7-37"><a href="#cb7-37"></a>    <span class="cf">for</span>(gene <span class="cf">in</span> mt_genes){</span>
<span id="cb7-38"><a href="#cb7-38"></a>      <span class="cf">if</span>(<span class="fu">grepl</span>(gene, tree, <span class="at">fixed =</span> <span class="cn">TRUE</span>)){</span>
<span id="cb7-39"><a href="#cb7-39"></a>        mt <span class="ot">=</span> <span class="cn">TRUE</span></span>
<span id="cb7-40"><a href="#cb7-40"></a>      }</span>
<span id="cb7-41"><a href="#cb7-41"></a>    }</span>
<span id="cb7-42"><a href="#cb7-42"></a>    <span class="cf">if</span>(cp <span class="sc">==</span> <span class="cn">TRUE</span> <span class="sc">&amp;</span> mt <span class="sc">==</span> <span class="cn">TRUE</span>){</span>
<span id="cb7-43"><a href="#cb7-43"></a>      tree_list_fltr <span class="ot">=</span> <span class="fu">c</span>(tree_list_fltr, tree)</span>
<span id="cb7-44"><a href="#cb7-44"></a>    }</span>
<span id="cb7-45"><a href="#cb7-45"></a>  }</span>
<span id="cb7-46"><a href="#cb7-46"></a>  </span>
<span id="cb7-47"><a href="#cb7-47"></a>  <span class="co"># import trees</span></span>
<span id="cb7-48"><a href="#cb7-48"></a>  <span class="cf">for</span>(tree <span class="cf">in</span> tree_list_fltr){</span>
<span id="cb7-49"><a href="#cb7-49"></a>    combinatorial_trees[[<span class="fu">paste</span>(<span class="st">"cp_mt_conc_"</span>, k, <span class="st">"_genes"</span>, <span class="at">sep =</span> <span class="st">""</span>)]][[<span class="fu">length</span>(combinatorial_trees[[<span class="fu">paste</span>(<span class="st">"cp_mt_conc_"</span>, k, <span class="st">"_genes"</span>, <span class="at">sep =</span> <span class="st">""</span>)]]) <span class="sc">+</span> <span class="dv">1</span>]] <span class="ot">=</span> ape<span class="sc">::</span><span class="fu">read.tree</span>(<span class="at">file =</span> <span class="fu">paste</span>(mainDir, <span class="st">"/09_combinatorial_phylogeny/01_cp_mt_combinatorial/cp_mt_conc_"</span>, k, <span class="st">"_genes/"</span>, tree, <span class="at">sep =</span> <span class="st">""</span>))</span>
<span id="cb7-50"><a href="#cb7-50"></a>    <span class="fu">names</span>(combinatorial_trees[[<span class="fu">paste</span>(<span class="st">"cp_mt_conc_"</span>, k, <span class="st">"_genes"</span>, <span class="at">sep =</span> <span class="st">""</span>)]])[[<span class="fu">length</span>(combinatorial_trees[[<span class="fu">paste</span>(<span class="st">"cp_mt_conc_"</span>, k, <span class="st">"_genes"</span>, <span class="at">sep =</span> <span class="st">""</span>)]])]] <span class="ot">=</span> stringr<span class="sc">::</span><span class="fu">str_remove</span>(tree, <span class="st">".aln.contree"</span>)</span>
<span id="cb7-51"><a href="#cb7-51"></a>  }</span>
<span id="cb7-52"><a href="#cb7-52"></a>  </span>
<span id="cb7-53"><a href="#cb7-53"></a>  <span class="co"># clean</span></span>
<span id="cb7-54"><a href="#cb7-54"></a>  <span class="fu">rm</span>(cp, gene, k, mt, tree, tree_list_all, tree_list_fltr)</span>
<span id="cb7-55"><a href="#cb7-55"></a>}</span>
<span id="cb7-56"><a href="#cb7-56"></a></span>
<span id="cb7-57"><a href="#cb7-57"></a></span>
<span id="cb7-58"><a href="#cb7-58"></a></span>
<span id="cb7-59"><a href="#cb7-59"></a><span class="co">#------------------------------------------------------------------------------#</span></span>
<span id="cb7-60"><a href="#cb7-60"></a><span class="do">## 2.2 - Get distance matrices                                              ####</span></span>
<span id="cb7-61"><a href="#cb7-61"></a></span>
<span id="cb7-62"><a href="#cb7-62"></a><span class="co"># create distance lists</span></span>
<span id="cb7-63"><a href="#cb7-63"></a>combinatorial_distances <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_conc_2_genes"</span> <span class="ot">=</span> <span class="cn">NULL</span>,</span>
<span id="cb7-64"><a href="#cb7-64"></a>                               <span class="st">"cp_mt_conc_3_genes"</span> <span class="ot">=</span> <span class="cn">NULL</span>,</span>
<span id="cb7-65"><a href="#cb7-65"></a>                               <span class="st">"cp_mt_conc_4_genes"</span> <span class="ot">=</span> <span class="cn">NULL</span>,</span>
<span id="cb7-66"><a href="#cb7-66"></a>                               <span class="st">"cp_mt_conc_5_genes"</span> <span class="ot">=</span> <span class="cn">NULL</span>,</span>
<span id="cb7-67"><a href="#cb7-67"></a>                               <span class="st">"cp_mt_conc_6_genes"</span> <span class="ot">=</span> <span class="cn">NULL</span>,</span>
<span id="cb7-68"><a href="#cb7-68"></a>                               <span class="st">"cp_mt_conc_7_genes"</span> <span class="ot">=</span> <span class="cn">NULL</span>,</span>
<span id="cb7-69"><a href="#cb7-69"></a>                               <span class="st">"cp_mt_conc_8_genes"</span> <span class="ot">=</span> <span class="cn">NULL</span>,</span>
<span id="cb7-70"><a href="#cb7-70"></a>                               <span class="st">"cp_mt_conc_9_genes"</span> <span class="ot">=</span> <span class="cn">NULL</span>,</span>
<span id="cb7-71"><a href="#cb7-71"></a>                               <span class="st">"cp_mt_conc_10_genes"</span> <span class="ot">=</span> <span class="cn">NULL</span>)</span>
<span id="cb7-72"><a href="#cb7-72"></a></span>
<span id="cb7-73"><a href="#cb7-73"></a><span class="co"># populate distance lists</span></span>
<span id="cb7-74"><a href="#cb7-74"></a>combinatorial_list <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">25</span>, <span class="dv">97</span>, <span class="dv">195</span>, <span class="dv">247</span>, <span class="dv">210</span>, <span class="dv">121</span>, <span class="dv">46</span>, <span class="dv">11</span>, <span class="dv">2</span>)</span>
<span id="cb7-75"><a href="#cb7-75"></a><span class="cf">for</span>(k <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(combinatorial_distances)){</span>
<span id="cb7-76"><a href="#cb7-76"></a>  </span>
<span id="cb7-77"><a href="#cb7-77"></a>  <span class="co"># get size of the results matrix</span></span>
<span id="cb7-78"><a href="#cb7-78"></a>  empty_matrix <span class="ot">=</span> <span class="fu">matrix</span>(<span class="at">nrow =</span> combinatorial_list[[k]],</span>
<span id="cb7-79"><a href="#cb7-79"></a>                        <span class="at">ncol =</span> combinatorial_list[[k]])</span>
<span id="cb7-80"><a href="#cb7-80"></a>  <span class="fu">colnames</span>(empty_matrix) <span class="ot">=</span> <span class="fu">names</span>(combinatorial_trees[[k]])</span>
<span id="cb7-81"><a href="#cb7-81"></a>  <span class="fu">rownames</span>(empty_matrix) <span class="ot">=</span> <span class="fu">names</span>(combinatorial_trees[[k]])</span>
<span id="cb7-82"><a href="#cb7-82"></a>  </span>
<span id="cb7-83"><a href="#cb7-83"></a>  <span class="co"># populate</span></span>
<span id="cb7-84"><a href="#cb7-84"></a>  combinatorial_distances[[k]] <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"common_nodes_corr"</span> <span class="ot">=</span> empty_matrix,</span>
<span id="cb7-85"><a href="#cb7-85"></a>                                      <span class="st">"RF"</span> <span class="ot">=</span> empty_matrix,</span>
<span id="cb7-86"><a href="#cb7-86"></a>                                      <span class="st">"KF"</span> <span class="ot">=</span> empty_matrix,</span>
<span id="cb7-87"><a href="#cb7-87"></a>                                      <span class="st">"MAST_%"</span> <span class="ot">=</span> empty_matrix)</span>
<span id="cb7-88"><a href="#cb7-88"></a>  </span>
<span id="cb7-89"><a href="#cb7-89"></a>  <span class="co"># clean</span></span>
<span id="cb7-90"><a href="#cb7-90"></a>  <span class="fu">rm</span>(empty_matrix)</span>
<span id="cb7-91"><a href="#cb7-91"></a>}</span>
<span id="cb7-92"><a href="#cb7-92"></a></span>
<span id="cb7-93"><a href="#cb7-93"></a><span class="co"># get pairwise distances</span></span>
<span id="cb7-94"><a href="#cb7-94"></a><span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(combinatorial_trees)){</span>
<span id="cb7-95"><a href="#cb7-95"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(combinatorial_trees[[j]])){</span>
<span id="cb7-96"><a href="#cb7-96"></a>    <span class="cf">for</span>(k <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(combinatorial_trees[[j]])){</span>
<span id="cb7-97"><a href="#cb7-97"></a>      </span>
<span id="cb7-98"><a href="#cb7-98"></a>      <span class="co"># get tree</span></span>
<span id="cb7-99"><a href="#cb7-99"></a>      tree1_tmp <span class="ot">=</span> combinatorial_trees[[j]][[i]]</span>
<span id="cb7-100"><a href="#cb7-100"></a>      tree2_tmp <span class="ot">=</span> combinatorial_trees[[j]][[k]]</span>
<span id="cb7-101"><a href="#cb7-101"></a></span>
<span id="cb7-102"><a href="#cb7-102"></a>      <span class="co"># list of common species</span></span>
<span id="cb7-103"><a href="#cb7-103"></a>      species_list <span class="ot">=</span> tree1_tmp<span class="sc">$</span>tip.label[<span class="fu">which</span>(tree1_tmp<span class="sc">$</span>tip.label <span class="sc">%in%</span> tree2_tmp<span class="sc">$</span>tip.label)]</span>
<span id="cb7-104"><a href="#cb7-104"></a></span>
<span id="cb7-105"><a href="#cb7-105"></a>      <span class="co"># get overlapping species</span></span>
<span id="cb7-106"><a href="#cb7-106"></a>      tree1 <span class="ot">=</span> ape<span class="sc">::</span><span class="fu">keep.tip</span>(tree1_tmp, species_list)</span>
<span id="cb7-107"><a href="#cb7-107"></a>      tree2 <span class="ot">=</span> ape<span class="sc">::</span><span class="fu">keep.tip</span>(tree2_tmp, species_list)</span>
<span id="cb7-108"><a href="#cb7-108"></a>      </span>
<span id="cb7-109"><a href="#cb7-109"></a>      <span class="co"># get distances</span></span>
<span id="cb7-110"><a href="#cb7-110"></a>      distances <span class="ot">=</span> phangorn<span class="sc">::</span><span class="fu">treedist</span>(tree1, tree2)</span>
<span id="cb7-111"><a href="#cb7-111"></a>      combinatorial_distances[[j]][[<span class="st">"RF"</span>]][i, k] <span class="ot">=</span> distances[[<span class="dv">1</span>]]</span>
<span id="cb7-112"><a href="#cb7-112"></a>      combinatorial_distances[[j]][[<span class="st">"KF"</span>]][i, k] <span class="ot">=</span> distances[[<span class="dv">2</span>]]</span>
<span id="cb7-113"><a href="#cb7-113"></a>      combinatorial_distances[[j]][[<span class="st">"MAST_%"</span>]][i, k] <span class="ot">=</span> <span class="fu">length</span>(phangorn<span class="sc">::</span><span class="fu">mast</span>(tree1, tree2, <span class="at">tree =</span> <span class="cn">FALSE</span>)) <span class="sc">/</span> <span class="fu">length</span>(species_list)</span>
<span id="cb7-114"><a href="#cb7-114"></a>      </span>
<span id="cb7-115"><a href="#cb7-115"></a>      <span class="co"># clean</span></span>
<span id="cb7-116"><a href="#cb7-116"></a>      <span class="fu">rm</span>(tree1, tree2, tree1_tmp, tree2_tmp, distances)</span>
<span id="cb7-117"><a href="#cb7-117"></a>    }</span>
<span id="cb7-118"><a href="#cb7-118"></a>  }</span>
<span id="cb7-119"><a href="#cb7-119"></a>}</span>
<span id="cb7-120"><a href="#cb7-120"></a></span>
<span id="cb7-121"><a href="#cb7-121"></a></span>
<span id="cb7-122"><a href="#cb7-122"></a><span class="do">### get pairwise common_nodes_corr distances</span></span>
<span id="cb7-123"><a href="#cb7-123"></a></span>
<span id="cb7-124"><a href="#cb7-124"></a><span class="co"># iterate all combinatorial trees</span></span>
<span id="cb7-125"><a href="#cb7-125"></a><span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(combinatorial_trees)){</span>
<span id="cb7-126"><a href="#cb7-126"></a>  </span>
<span id="cb7-127"><a href="#cb7-127"></a>  <span class="co"># create CPU cluster</span></span>
<span id="cb7-128"><a href="#cb7-128"></a>  cl <span class="ot">=</span> parallel<span class="sc">::</span><span class="fu">makeCluster</span>(<span class="dv">8</span>, <span class="at">type =</span> <span class="st">"SOCK"</span>)</span>
<span id="cb7-129"><a href="#cb7-129"></a>  doSNOW<span class="sc">::</span><span class="fu">registerDoSNOW</span>(cl)</span>
<span id="cb7-130"><a href="#cb7-130"></a>  </span>
<span id="cb7-131"><a href="#cb7-131"></a>  <span class="co"># get chronos in parallel</span></span>
<span id="cb7-132"><a href="#cb7-132"></a>  chronos_list <span class="ot">=</span> <span class="fu">foreach</span>(<span class="at">k =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(combinatorial_trees[[j]]), <span class="at">.combine =</span> <span class="st">"c"</span>) <span class="sc">%dopar%</span> {</span>
<span id="cb7-133"><a href="#cb7-133"></a>    <span class="fu">library</span>(<span class="st">"ape"</span>)</span>
<span id="cb7-134"><a href="#cb7-134"></a>    <span class="co"># get tree</span></span>
<span id="cb7-135"><a href="#cb7-135"></a>    <span class="fu">tryCatch</span>(<span class="fu">chronos</span>(combinatorial_trees[[j]][[k]]), <span class="at">error =</span> <span class="cf">function</span>(e) { <span class="cn">NA</span> })</span>
<span id="cb7-136"><a href="#cb7-136"></a>  }  </span>
<span id="cb7-137"><a href="#cb7-137"></a>  parallel<span class="sc">::</span><span class="fu">stopCluster</span>(cl)</span>
<span id="cb7-138"><a href="#cb7-138"></a>  <span class="fu">names</span>(chronos_list) <span class="ot">=</span> <span class="fu">names</span>(combinatorial_trees[[j]])</span>
<span id="cb7-139"><a href="#cb7-139"></a>  </span>
<span id="cb7-140"><a href="#cb7-140"></a>  <span class="co"># get pairwise common_nodes_corr distances</span></span>
<span id="cb7-141"><a href="#cb7-141"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(combinatorial_trees[[j]])){</span>
<span id="cb7-142"><a href="#cb7-142"></a>    <span class="cf">for</span>(k <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(combinatorial_trees[[j]])){</span>
<span id="cb7-143"><a href="#cb7-143"></a>      </span>
<span id="cb7-144"><a href="#cb7-144"></a>      <span class="fu">print</span>(<span class="fu">paste</span>(j, i, k, <span class="at">sep =</span> <span class="st">","</span>))</span>
<span id="cb7-145"><a href="#cb7-145"></a>      </span>
<span id="cb7-146"><a href="#cb7-146"></a>      <span class="co"># get trees</span></span>
<span id="cb7-147"><a href="#cb7-147"></a>      tree1_tmp <span class="ot">=</span> chronos_list[[<span class="fu">names</span>(combinatorial_trees[[j]])[i]]]</span>
<span id="cb7-148"><a href="#cb7-148"></a>      tree2_tmp <span class="ot">=</span> chronos_list[[<span class="fu">names</span>(combinatorial_trees[[j]])[k]]]</span>
<span id="cb7-149"><a href="#cb7-149"></a>      </span>
<span id="cb7-150"><a href="#cb7-150"></a>      <span class="co"># check trees and get distance</span></span>
<span id="cb7-151"><a href="#cb7-151"></a>      <span class="cf">if</span>(<span class="fu">all</span>(<span class="sc">!</span><span class="fu">is.na</span>(tree1_tmp)) <span class="sc">&amp;</span> <span class="fu">all</span>(<span class="sc">!</span><span class="fu">is.na</span>(tree2_tmp))){</span>
<span id="cb7-152"><a href="#cb7-152"></a>        combinatorial_distances[[j]][[<span class="st">"common_nodes_corr"</span>]][i, k] <span class="ot">=</span> <span class="fu">tryCatch</span>(</span>
<span id="cb7-153"><a href="#cb7-153"></a>          <span class="fu">cor.dendlist</span>(<span class="fu">dendlist</span>(<span class="fu">as.dendrogram</span>(ape<span class="sc">::</span><span class="fu">root</span>(tree1_tmp, <span class="at">outgroup =</span> <span class="fu">c</span>(<span class="st">"Oviri"</span>, <span class="st">"Pakin"</span>), <span class="at">resolve.root =</span> <span class="cn">TRUE</span>)),</span>
<span id="cb7-154"><a href="#cb7-154"></a>                                <span class="fu">as.dendrogram</span>(ape<span class="sc">::</span><span class="fu">root</span>(tree2_tmp, <span class="at">outgroup =</span> <span class="fu">c</span>(<span class="st">"Oviri"</span>, <span class="st">"Pakin"</span>), <span class="at">resolve.root =</span> <span class="cn">TRUE</span>))),</span>
<span id="cb7-155"><a href="#cb7-155"></a>                       <span class="at">method =</span> <span class="st">"common_nodes"</span>)[<span class="dv">2</span>],</span>
<span id="cb7-156"><a href="#cb7-156"></a>          <span class="at">error =</span> <span class="cf">function</span>(e) { <span class="cn">NA</span> })</span>
<span id="cb7-157"><a href="#cb7-157"></a>      } <span class="cf">else</span> {</span>
<span id="cb7-158"><a href="#cb7-158"></a>        combinatorial_distances[[j]][[<span class="st">"common_nodes_corr"</span>]][i, k] <span class="ot">=</span> <span class="cn">NA</span></span>
<span id="cb7-159"><a href="#cb7-159"></a>      }</span>
<span id="cb7-160"><a href="#cb7-160"></a>    }</span>
<span id="cb7-161"><a href="#cb7-161"></a>  }</span>
<span id="cb7-162"><a href="#cb7-162"></a>}</span>
<span id="cb7-163"><a href="#cb7-163"></a></span>
<span id="cb7-164"><a href="#cb7-164"></a></span>
<span id="cb7-165"><a href="#cb7-165"></a><span class="do">## export the distance matrices</span></span>
<span id="cb7-166"><a href="#cb7-166"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(combinatorial_distances)){</span>
<span id="cb7-167"><a href="#cb7-167"></a>  <span class="fu">write.table</span>(combinatorial_distances[[i]][[<span class="st">"common_nodes_corr"</span>]],</span>
<span id="cb7-168"><a href="#cb7-168"></a>              <span class="at">file =</span> <span class="fu">paste</span>(<span class="st">"UlvaOmics."</span>, <span class="fu">format</span>(<span class="fu">Sys.Date</span>(), <span class="at">format =</span> <span class="st">"%Y%m%d"</span>), <span class="st">".05_combinatorial_stats.cp_mt_conc_"</span>, i <span class="sc">+</span> <span class="dv">1</span>, <span class="st">"_genes.common_nodes_corr.txt"</span>,  <span class="at">sep =</span> <span class="st">""</span>),</span>
<span id="cb7-169"><a href="#cb7-169"></a>              <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>, <span class="at">col.names =</span> <span class="cn">TRUE</span>, <span class="at">row.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-170"><a href="#cb7-170"></a>  <span class="fu">write.table</span>(combinatorial_distances[[i]][[<span class="st">"RF"</span>]],</span>
<span id="cb7-171"><a href="#cb7-171"></a>              <span class="at">file =</span> <span class="fu">paste</span>(<span class="st">"UlvaOmics."</span>, <span class="fu">format</span>(<span class="fu">Sys.Date</span>(), <span class="at">format =</span> <span class="st">"%Y%m%d"</span>), <span class="st">".05_combinatorial_stats.cp_mt_conc_"</span>, i <span class="sc">+</span> <span class="dv">1</span>, <span class="st">"_genes.RF.txt"</span>,  <span class="at">sep =</span> <span class="st">""</span>),</span>
<span id="cb7-172"><a href="#cb7-172"></a>              <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>, <span class="at">col.names =</span> <span class="cn">TRUE</span>, <span class="at">row.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-173"><a href="#cb7-173"></a>  <span class="fu">write.table</span>(combinatorial_distances[[i]][[<span class="st">"KF"</span>]],</span>
<span id="cb7-174"><a href="#cb7-174"></a>              <span class="at">file =</span> <span class="fu">paste</span>(<span class="st">"UlvaOmics."</span>, <span class="fu">format</span>(<span class="fu">Sys.Date</span>(), <span class="at">format =</span> <span class="st">"%Y%m%d"</span>), <span class="st">".05_combinatorial_stats.cp_mt_conc_"</span>, i <span class="sc">+</span> <span class="dv">1</span>, <span class="st">"_genes.KF.txt"</span>,  <span class="at">sep =</span> <span class="st">""</span>),</span>
<span id="cb7-175"><a href="#cb7-175"></a>              <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>, <span class="at">col.names =</span> <span class="cn">TRUE</span>, <span class="at">row.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-176"><a href="#cb7-176"></a>  <span class="fu">write.table</span>(combinatorial_distances[[i]][[<span class="st">"MAST_%"</span>]],</span>
<span id="cb7-177"><a href="#cb7-177"></a>              <span class="at">file =</span> <span class="fu">paste</span>(<span class="st">"UlvaOmics."</span>, <span class="fu">format</span>(<span class="fu">Sys.Date</span>(), <span class="at">format =</span> <span class="st">"%Y%m%d"</span>), <span class="st">".05_combinatorial_stats.cp_mt_conc_"</span>, i <span class="sc">+</span> <span class="dv">1</span>, <span class="st">"_genes.MAST_%.txt"</span>,  <span class="at">sep =</span> <span class="st">""</span>),</span>
<span id="cb7-178"><a href="#cb7-178"></a>              <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>, <span class="at">col.names =</span> <span class="cn">TRUE</span>, <span class="at">row.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-179"><a href="#cb7-179"></a>}</span>
<span id="cb7-180"><a href="#cb7-180"></a></span>
<span id="cb7-181"><a href="#cb7-181"></a></span>
<span id="cb7-182"><a href="#cb7-182"></a></span>
<span id="cb7-183"><a href="#cb7-183"></a><span class="co">#------------------------------------------------------------------------------#</span></span>
<span id="cb7-184"><a href="#cb7-184"></a><span class="do">## 2.3 - Plot distance matrices                                             ####</span></span>
<span id="cb7-185"><a href="#cb7-185"></a></span>
<span id="cb7-186"><a href="#cb7-186"></a></span>
<span id="cb7-187"><a href="#cb7-187"></a><span class="co"># </span></span>
<span id="cb7-188"><a href="#cb7-188"></a><span class="co"># for(k in 1:length(combinatorial_distances)){</span></span>
<span id="cb7-189"><a href="#cb7-189"></a><span class="co">#   </span></span>
<span id="cb7-190"><a href="#cb7-190"></a><span class="co">#   # remove possible rows/columns with only NAs</span></span>
<span id="cb7-191"><a href="#cb7-191"></a><span class="co">#   tmp_matrix = distance_sets[[i]][[k]]</span></span>
<span id="cb7-192"><a href="#cb7-192"></a><span class="co">#   tmp_matrix = tmp_matrix[, colSums(is.na(tmp_matrix)) &lt; nrow(tmp_matrix)]</span></span>
<span id="cb7-193"><a href="#cb7-193"></a><span class="co">#   tmp_matrix = tmp_matrix[rowSums(is.na(tmp_matrix)) &lt; ncol(tmp_matrix), ]</span></span>
<span id="cb7-194"><a href="#cb7-194"></a><span class="co">#   </span></span>
<span id="cb7-195"><a href="#cb7-195"></a><span class="co">#   # get names</span></span>
<span id="cb7-196"><a href="#cb7-196"></a><span class="co">#   plot_name = ifelse(names(distance_sets[[i]])[[k]] == "common_nodes_corr", "common\nnodes\ncorr", names(distance_sets[[i]])[[k]])</span></span>
<span id="cb7-197"><a href="#cb7-197"></a><span class="co">#   </span></span>
<span id="cb7-198"><a href="#cb7-198"></a><span class="co">#   # prep annotation</span></span>
<span id="cb7-199"><a href="#cb7-199"></a><span class="co">#   organell = as.data.frame(ifelse(rownames(tmp_matrix) %in% cp_genes_list, "cp",</span></span>
<span id="cb7-200"><a href="#cb7-200"></a><span class="co">#                                   ifelse(rownames(tmp_matrix) %in% mt_genes_list, "mt", "cp_mt")))</span></span>
<span id="cb7-201"><a href="#cb7-201"></a><span class="co">#   rownames(organell) = rownames(tmp_matrix)</span></span>
<span id="cb7-202"><a href="#cb7-202"></a><span class="co">#   colnames(organell) = "organell"</span></span>
<span id="cb7-203"><a href="#cb7-203"></a><span class="co">#   organell$color = ifelse(organell$organell == "cp", "#469d89", ifelse(organell$organell == "mt", "steelblue", "red"))</span></span>
<span id="cb7-204"><a href="#cb7-204"></a><span class="co">#   </span></span>
<span id="cb7-205"><a href="#cb7-205"></a><span class="co">#   # plot raw distance</span></span>
<span id="cb7-206"><a href="#cb7-206"></a><span class="co">#   distances_plots[[i]][[k]] = grid.grabExpr(draw(ComplexHeatmap::Heatmap(tmp_matrix,</span></span>
<span id="cb7-207"><a href="#cb7-207"></a><span class="co">#                                                                          col = colorRamp2(c(min(tmp_matrix[!is.na(tmp_matrix)]),</span></span>
<span id="cb7-208"><a href="#cb7-208"></a><span class="co">#                                                                                             max(tmp_matrix[!is.na(tmp_matrix)])),</span></span>
<span id="cb7-209"><a href="#cb7-209"></a><span class="co">#                                                                                           c("white", "#D55E00")),</span></span>
<span id="cb7-210"><a href="#cb7-210"></a><span class="co">#                                                                          row_names_gp = gpar(col = organell$color),</span></span>
<span id="cb7-211"><a href="#cb7-211"></a><span class="co">#                                                                          column_names_gp = gpar(col = organell$color),</span></span>
<span id="cb7-212"><a href="#cb7-212"></a><span class="co">#                                                                          top_annotation = HeatmapAnnotation(organell = as.matrix(organell$organell),</span></span>
<span id="cb7-213"><a href="#cb7-213"></a><span class="co">#                                                                                                             show_annotation_name = FALSE,</span></span>
<span id="cb7-214"><a href="#cb7-214"></a><span class="co">#                                                                                                             show_legend = FALSE,</span></span>
<span id="cb7-215"><a href="#cb7-215"></a><span class="co">#                                                                                                             col = list(organell = c("cp_mt" = "red", "cp" = "#469d89", "mt" = "steelblue"))),</span></span>
<span id="cb7-216"><a href="#cb7-216"></a><span class="co">#                                                                          left_annotation = rowAnnotation(organell = as.matrix(organell$organell),</span></span>
<span id="cb7-217"><a href="#cb7-217"></a><span class="co">#                                                                                                          show_annotation_name = FALSE,</span></span>
<span id="cb7-218"><a href="#cb7-218"></a><span class="co">#                                                                                                          col = list(organell = c("cp_mt" = "red", "cp" = "#469d89", "mt" = "steelblue"))),</span></span>
<span id="cb7-219"><a href="#cb7-219"></a><span class="co">#                                                                          name = plot_name)))</span></span>
<span id="cb7-220"><a href="#cb7-220"></a><span class="co">#   grid.newpage(recording = TRUE)</span></span>
<span id="cb7-221"><a href="#cb7-221"></a><span class="co">#   grid.draw(distances_plots[[i]][[k]])</span></span>
<span id="cb7-222"><a href="#cb7-222"></a><span class="co"># </span></span>
<span id="cb7-223"><a href="#cb7-223"></a><span class="co"># </span></span>
<span id="cb7-224"><a href="#cb7-224"></a><span class="co"># }</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To resolve Ulva species phylogenetic relationships, supermatrix and coalescence-based phylogenetic approaches were applied to the chloroplast and mitochondrial datasets. To appreciate the different phylogenetic signals present in the two organellar datasets, the chloroplast and mitochondrial datasets were first analyzed independently. Afterwards, a supermatrix containing both chloroplast and mitochondrial genes was composed in order to infer the Ulva species phylogenetic tree. For the supermatrix Maximum-likelihood (ML) reconstruction, a gene-wise partitioned analysis was performed (Chernomor et al 2016). To estimate the best substitution model of each partition, ML trees were built with IQtree v.2.2.0 (Minh et al 2020) for each single-copy gene, inferring the best model and rate of heterogeneity across sites. All ML analyses were run with 1,000 ultra-fast bootstrap and SH-aLRT branch test replicates. Gene trees were used also for the coalescent-based analyses, using ASTRAL v5.7.8 (Zhang et al 2018), using the multilocus bootstrap support (MLBS) approach. 100 replicates were run (-r) starting from the 1,000 bootstrap trees generated for each gene during the ML analyses, allowing gene and site resampling (–gene-resampling flag). Four metrics were used to assess the concordance of the Ulva species tree reconstructed from supermatrix and coalescence-based analyses obtained from the analyses of chloroplast genes, mitochondrial genes and chloroplast and mitochondrial genes: Robinson-Foulds (Robinson &amp; Foulds 1981), Kuhner-Felsenstein (Kuhner &amp; Felsenstein 1994), Maximum Agreement Subtree (Finden &amp; Gordon 1985) calculated with phangorn package (Schliep 2011), and correlation between the common nodes calculated with corr.dendlist function from dendexted package (Galili 2015). The RF distance measures the number of unique splits between two trees by comparing their bipartitions. It ranges from 0 (identical trees) to a higher value indicating greater dissimilarity with more unique splits (Robinson &amp; Foulds 1981). The KF index quantifies the average genetic differences per site between populations, considering genetic variation within and between populations to provide a quantitative measure of genetic distance (Kuhner &amp; Felsenstein 1994). The MAST (%) identifies the subset of nodes and branches that are common to both trees, disregarding any additional or missing branches. It aims to find the largest possible subtree that can be extracted from both trees while preserving their structural similarities, representing the maximum level of agreement in their branching patterns (Finden &amp; Gordon 1985). The correlation between the common nodes is calculated as the pairwise cophenetic correlation coefficient between the distance matrix and the cophenetic matrix of the common nodes (Galili 2015). Both phangorn and dendextend packages are available in R statistical programming language (R Core Team 2022). Organellar gene ranking To identify the minimum number of genes (markers) necessary to reconstruct the Ulva species tree reconstructed with the 99 chloroplast and mitochondrial markers, Robinson-Foulds, Kuhner-Felsenstein, Maximum Agreement Subtree and correlation between the common nodes metrics were calculated between each of the 99 chloroplast and mitochondrial markers and the Ulva species tree reconstructed from supermatrix analysis. The same approach applied to the Ulva species trees reconstructed using chloroplast and mitochondrial supermatrix and coalesce-based analyses suggested that Maximum Agreement Subtree and common correlation nodes metrics are the best performers to assess phylogenetic tree distances when the tree do not have the same set of tips. The 10 markers (6 chloroplasts and 4 mitochondrial) with the best MAST% and common correlation nodes scores were arbitrarily chosen as candidates for the subsequent combinatorial phylogenetic analyses. Combinatorial phylogenetic analysis For the combinatorial phylogenetic analysis, a custom python script was used to exhaustively calculate the permutations between the chosen 10 marker genes, sampling between 2 and 10 markers. This resulted in the 945 unique combinations of at least a chloroplast and a mitochondrial marker (24 combinations of 2 marker genes, 96 of 3 markers, 194 of 4 markers, 246 of 5 markers, 209 of 6 markers, 120 of 7 markers, 45 of 8 markers, 10 of 9 markers and finally 1 containing all 10 markers genes). For each combination of markers, the script would then independently reconstruct the corresponding phylogenetic tree by invoking IQtree, inferring the best model and rate of heterogeneity across sites, with 1,000 ultra-fast bootstrap and SH-aLRT branch test replicates. Four metrics (RF, KF, MAST% and common correlation nodes scores) between each of the reconstructed 945 tree and the Ulva species tree reconstructed with the 99 chloroplast and mitochondrial markers were calculated in an iterative manner to identify the minimum set of chloroplast and mitochondrial markers necessary to reconstruct a ML phylogenetic tree concordant with the Ulva species tree reconstructed in this study. These markers are suggested to be used as a novel combination of markers to molecularly characterize Ulva species, and universal primers will be designed for each of the markers.</p>
<p>r each combination.</p>
<p>genes representative best genes markers MAST% # of equivalent combinations Common node correlation # of equivalent combinations 2 psaA_cox1 0.824 1 0.901 1 psaA_cox2 3 rps2_psaA_cox1 0.882 6 0.931 1 psbB_cox1_cox2 4 psbD_psaA_cox1_cox2 0.941 1 0.950 1 5 psbD_rps2_psaA_cox1_cox2 0.941 6 0.960 1 psaB_psbB_rps2_cox1_cox2 6 psbB_psbD_rps2_psaA_cox1_cox2 0.941 10 0.960 1 petB_psaB_psbB_rps2_cox1_cox2 7 psaB_psbB_psbD_rps2_psaA_cox1_cox2 0.941 8 0.960 3 petB_psbD_rps2_psaA_atp6_cox1_cox2 8 petB_psbB_psbD_rps2_psaA_atp6_cox1_cox2 0.941 2 0.960 1 9 petB_psaB_psbB_psbD_rps2_psaA_atp6_cox1_rps3 0.922 2 0.950 2 10 petB_psaB_psbB_psbD_rps2_psaA_atp6_cox1_cox2_rps3 0.882 1 0.931 1</p>
</section>
</section>
<section id="sec-lesson-learnt-markers-selection" class="level2" data-number="3.4">
<h2 data-number="3.4" class="anchored" data-anchor-id="sec-lesson-learnt-markers-selection"><span class="header-section-number">3.4</span> Lessons Learnt</h2>
<p>So far, we have learnt:</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./cp-and-mt-phylogeny-reconstruction.html" class="pagination-link" aria-label="Species tree reconstruction from organellar genomes">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Species tree reconstruction from organellar genomes</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      <ul class="footer-items list-unstyled">
    <li class="nav-item">
 Copyright © 2025 Andrea Del Cortona.
  </li>  
</ul>
    </div>   
    <div class="nav-footer-center">
      <ul class="footer-items list-unstyled">
    <li class="nav-item">
    <a class="nav-link" href="https://quarto.org/">
<p>This page is built with ❤️ and <em>Quarto</em>.</p>
</a>
  </li>  
</ul>
    </div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/19ADC99">
      <i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/andrea-del-cortona-86a75474/">
      <i class="bi bi-linkedin" role="img" aria-label="Lindekin">
</i> 
    </a>
  </li>  
</ul>
    </div>
  </div>
</footer>




</body></html>