<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>3&nbsp; Gene markers selection – Development of taxonomical biomarkers for *Ulva* genus</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./cp-and-mt-phylogeny-reconstruction.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<link href="site_libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="site_libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="site_libs/datatables-binding-0.33/datatables.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="site_libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="site_libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="site_libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script>
<link href="site_libs/dt-ext-fixedcolumns-1.13.6/css/fixedColumns.dataTables.min.css" rel="stylesheet">
<script src="site_libs/dt-ext-fixedcolumns-1.13.6/js/dataTables.fixedColumns.min.js"></script>
<link href="site_libs/dt-ext-fixedheader-1.13.6/css/fixedHeader.dataTables.min.css" rel="stylesheet">
<script src="site_libs/dt-ext-fixedheader-1.13.6/js/dataTables.fixedHeader.min.js"></script>
<link href="site_libs/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="site_libs/crosstalk-1.2.1/js/crosstalk.min.js"></script>
<style>
#quarto-sidebar .sidebar-item-text {
  font-weight: bold;
}
#quarto-sidebar .chapter-number {
  font-weight: normal
}
#quarto-sidebar .chapter-title {
  font-weight: normal
}
#.nav-tabs .nav-link.active {
  background-color: rgba(239, 236, 255, 0.5);
}
</style>


</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Development of taxonomical biomarkers for <em>Ulva</em> genus</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./markers-selection.html"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Gene markers selection</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Abstract</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cp-and-mt-dataset-overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title"><em>Ulva</em> organellar genomes retrieval and preparation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./cp-and-mt-phylogeny-reconstruction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Species tree reconstruction from organellar genomes</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./markers-selection.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Gene markers selection</span></span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#on-this-page" id="toc-on-this-page" class="nav-link active" data-scroll-target="#on-this-page"><span class="header-section-number">3.1</span> On this page</a></li>
  <li><a href="#organellar-gene-selection" id="toc-organellar-gene-selection" class="nav-link" data-scroll-target="#organellar-gene-selection"><span class="header-section-number">3.2</span> Organellar gene selection</a>
  <ul>
  <li><a href="#gene-distances-to-species-tree" id="toc-gene-distances-to-species-tree" class="nav-link" data-scroll-target="#gene-distances-to-species-tree"><span class="header-section-number">3.2.1</span> Gene distances to species tree</a></li>
  </ul></li>
  <li><a href="#markers-selection" id="toc-markers-selection" class="nav-link" data-scroll-target="#markers-selection"><span class="header-section-number">3.3</span> Markers selection</a></li>
  <li><a href="#sec-lesson-learnt-markers-selection" id="toc-sec-lesson-learnt-markers-selection" class="nav-link" data-scroll-target="#sec-lesson-learnt-markers-selection"><span class="header-section-number">3.4</span> Lessons Learnt</a></li>
  <li><a href="#session-information" id="toc-session-information" class="nav-link" data-scroll-target="#session-information"><span class="header-section-number">3.5</span> Session Information</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Gene markers selection</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div class="cell">
<style type="text/css">

pre {
  max-height: 300px;
}

div.sourceCode {
  overflow: scroll;
}

.bash-code {
  font-style: italic;
}

</style>
</div>
<section id="on-this-page" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="on-this-page"><span class="header-section-number">3.1</span> On this page</h2>
<p>Biological insights and take-home messages are at the bottom of the page at Lesson Learnt: <a href="#sec-lesson-learnt-markers-selection" class="quarto-xref"><span>Section 3.4</span></a>.</p>
<ul>
<li>Here we compute the distance between the single-gene ML trees and the reconstructed <em>Ulva</em> species tree;</li>
<li>we select the top 10 chloroplast and mitochondrial genes which single-gene tree is closer to the <em>Ulva</em> species tree;</li>
<li>we deploy a combinatorial approach to compute all the possible 2:10 gene combinations and the resulting ML phylogenetic tree;</li>
<li>finally, we determine the minimum set of genes that can confidently reconstruct the <em>Ulva</em> species tree.</li>
</ul>
</section>
<section id="organellar-gene-selection" class="level2" data-number="3.2">
<h2 data-number="3.2" class="anchored" data-anchor-id="organellar-gene-selection"><span class="header-section-number">3.2</span> Organellar gene selection</h2>
<p>In order to select the top 10 genes for the combinatorial approach, we first need to calculate the distance metrics between the single-gene ML tree and the reconstructed <em>Ulva</em> species tree. We have decided to constrain the search to the top 10 genes, since the space of gene combinations that are possible increase exponentially the more genes we add to the combinatorial analysis.</p>
<section id="gene-distances-to-species-tree" class="level3" data-number="3.2.1">
<h3 data-number="3.2.1" class="anchored" data-anchor-id="gene-distances-to-species-tree"><span class="header-section-number">3.2.1</span> Gene distances to species tree</h3>
<p>For each of the 99 chloroplast and mitochondrial genes we will calculate the four distance metrics that we have used so far: common nodes correlation, RF distance, KF distance and MAST %.</p>
<div class="cell" data-result="hide">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="co"># get cp ML gene trees</span></span>
<span id="cb1-2"><a href="#cb1-2"></a>cp_ML_genes <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb1-3"><a href="#cb1-3"></a>filenames <span class="ot">=</span> <span class="fu">list.files</span>(<span class="st">"./03_cp_singleGene_ML/"</span>, <span class="at">pattern =</span> <span class="st">"*.contree"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-4"><a href="#cb1-4"></a>cp_ML_genes <span class="ot">=</span> <span class="fu">lapply</span>(filenames, ape<span class="sc">::</span>read.tree)</span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="fu">names</span>(cp_ML_genes) <span class="ot">=</span> strigr<span class="sc">::</span><span class="fu">str_remove_all</span>(filenames, <span class="st">"./03_cp_singleGene_ML//"</span>)</span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="fu">names</span>(cp_ML_genes) <span class="ot">=</span> strigr<span class="sc">::</span><span class="fu">str_remove_all</span>(filenames, <span class="st">".aln.contree"</span>)</span>
<span id="cb1-7"><a href="#cb1-7"></a>cp_ML_genes[[<span class="st">"cp_mt_ML"</span>]] <span class="ot">=</span> phylogenetic_trees[[<span class="st">"cp_mt_ML"</span>]][[<span class="st">"tree"</span>]]</span>
<span id="cb1-8"><a href="#cb1-8"></a></span>
<span id="cb1-9"><a href="#cb1-9"></a><span class="co"># get mt ML gene trees</span></span>
<span id="cb1-10"><a href="#cb1-10"></a>mt_ML_genes <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb1-11"><a href="#cb1-11"></a>filenames <span class="ot">=</span> <span class="fu">list.files</span>(<span class="st">"./03_mt_singleGene_ML/"</span>, <span class="at">pattern =</span> <span class="st">"*.contree"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-12"><a href="#cb1-12"></a>mt_ML_genes <span class="ot">=</span> <span class="fu">lapply</span>(filenames, ape<span class="sc">::</span>read.tree)</span>
<span id="cb1-13"><a href="#cb1-13"></a><span class="fu">names</span>(mt_ML_genes) <span class="ot">=</span> strigr<span class="sc">::</span><span class="fu">str_remove_all</span>(filenames, <span class="st">"./03_mt_singleGene_ML//"</span>)</span>
<span id="cb1-14"><a href="#cb1-14"></a><span class="fu">names</span>(mt_ML_genes) <span class="ot">=</span> strigr<span class="sc">::</span><span class="fu">str_remove_all</span>(filenames, <span class="st">".aln.contree"</span>)</span>
<span id="cb1-15"><a href="#cb1-15"></a>mt_ML_genes[[<span class="st">"cp_mt_ML"</span>]] <span class="ot">=</span> phylogenetic_trees[[<span class="st">"cp_mt_ML"</span>]][[<span class="st">"tree"</span>]]</span>
<span id="cb1-16"><a href="#cb1-16"></a></span>
<span id="cb1-17"><a href="#cb1-17"></a><span class="co"># create distance lists</span></span>
<span id="cb1-18"><a href="#cb1-18"></a>genes_sets <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_ML"</span> <span class="ot">=</span> <span class="fu">c</span>(cp_ML_genes, mt_ML_genes[<span class="fu">which</span>(<span class="fu">names</span>(mt_ML_genes) <span class="sc">!=</span> <span class="st">"cp_mt_ML"</span>)]))</span>
<span id="cb1-19"><a href="#cb1-19"></a></span>
<span id="cb1-20"><a href="#cb1-20"></a>empty_matrix <span class="ot">=</span> <span class="fu">matrix</span>(<span class="at">nrow =</span> <span class="fu">length</span>(genes_sets[[<span class="st">"cp_mt_ML"</span>]]), <span class="at">ncol =</span> <span class="fu">length</span>(genes_sets[[<span class="st">"cp_mt_ML"</span>]]))</span>
<span id="cb1-21"><a href="#cb1-21"></a><span class="fu">colnames</span>(empty_matrix) <span class="ot">=</span> <span class="fu">names</span>(genes_sets[[<span class="st">"cp_mt_ML"</span>]])</span>
<span id="cb1-22"><a href="#cb1-22"></a><span class="fu">rownames</span>(empty_matrix) <span class="ot">=</span> <span class="fu">names</span>(genes_sets[[<span class="st">"cp_mt_ML"</span>]])</span>
<span id="cb1-23"><a href="#cb1-23"></a>cp_mt_genes_distances <span class="ot">=</span> <span class="fu">list</span>(</span>
<span id="cb1-24"><a href="#cb1-24"></a>  <span class="st">"common_nodes_corr"</span> <span class="ot">=</span> empty_matrix,</span>
<span id="cb1-25"><a href="#cb1-25"></a>  <span class="st">"RF"</span> <span class="ot">=</span> empty_matrix,</span>
<span id="cb1-26"><a href="#cb1-26"></a>  <span class="st">"KF"</span> <span class="ot">=</span> empty_matrix,</span>
<span id="cb1-27"><a href="#cb1-27"></a>  <span class="st">"MAST_%"</span> <span class="ot">=</span> empty_matrix</span>
<span id="cb1-28"><a href="#cb1-28"></a>)</span>
<span id="cb1-29"><a href="#cb1-29"></a></span>
<span id="cb1-30"><a href="#cb1-30"></a><span class="co"># populate distances lists</span></span>
<span id="cb1-31"><a href="#cb1-31"></a>distance_sets <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_genes_distances"</span> <span class="ot">=</span> cp_mt_genes_distances)</span>
<span id="cb1-32"><a href="#cb1-32"></a></span>
<span id="cb1-33"><a href="#cb1-33"></a><span class="co"># get pairwise distances</span></span>
<span id="cb1-34"><a href="#cb1-34"></a><span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(distance_sets)){</span>
<span id="cb1-35"><a href="#cb1-35"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(genes_sets[[j]])){</span>
<span id="cb1-36"><a href="#cb1-36"></a>    <span class="cf">for</span>(k <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(genes_sets[[j]])){</span>
<span id="cb1-37"><a href="#cb1-37"></a>      </span>
<span id="cb1-38"><a href="#cb1-38"></a>      <span class="co"># error handling</span></span>
<span id="cb1-39"><a href="#cb1-39"></a>      <span class="co"># MAST call errors for very unresulved trees (e.g.: psaM)</span></span>
<span id="cb1-40"><a href="#cb1-40"></a>      <span class="co">#   Error in root.phylo(x, bipart_x, resolve.root = TRUE) : </span></span>
<span id="cb1-41"><a href="#cb1-41"></a>      <span class="co">#   the specified outgroup is not monophyletic</span></span>
<span id="cb1-42"><a href="#cb1-42"></a>      <span class="co"># catch error and move forward</span></span>
<span id="cb1-43"><a href="#cb1-43"></a>      skip_to_next <span class="ot">=</span> <span class="cn">FALSE</span></span>
<span id="cb1-44"><a href="#cb1-44"></a>      </span>
<span id="cb1-45"><a href="#cb1-45"></a>      <span class="co"># get tree</span></span>
<span id="cb1-46"><a href="#cb1-46"></a>      tree1_tmp <span class="ot">=</span> genes_sets[[j]][[i]]</span>
<span id="cb1-47"><a href="#cb1-47"></a>      tree2_tmp <span class="ot">=</span> genes_sets[[j]][[k]]</span>
<span id="cb1-48"><a href="#cb1-48"></a>      </span>
<span id="cb1-49"><a href="#cb1-49"></a>      <span class="co"># list of common species</span></span>
<span id="cb1-50"><a href="#cb1-50"></a>      species_list <span class="ot">=</span> tree1_tmp<span class="sc">$</span>tip.label[<span class="fu">which</span>(tree1_tmp<span class="sc">$</span>tip.label <span class="sc">%in%</span> tree2_tmp<span class="sc">$</span>tip.label)]</span>
<span id="cb1-51"><a href="#cb1-51"></a>      </span>
<span id="cb1-52"><a href="#cb1-52"></a>      <span class="co"># get overlapping species</span></span>
<span id="cb1-53"><a href="#cb1-53"></a>      tree1 <span class="ot">=</span> ape<span class="sc">::</span><span class="fu">keep.tip</span>(tree1_tmp, species_list)</span>
<span id="cb1-54"><a href="#cb1-54"></a>      tree2 <span class="ot">=</span> ape<span class="sc">::</span><span class="fu">keep.tip</span>(tree2_tmp, species_list)</span>
<span id="cb1-55"><a href="#cb1-55"></a>      </span>
<span id="cb1-56"><a href="#cb1-56"></a>      <span class="co"># get distances</span></span>
<span id="cb1-57"><a href="#cb1-57"></a>      distances <span class="ot">=</span> phangorn<span class="sc">::</span><span class="fu">treedist</span>(tree1, tree2)</span>
<span id="cb1-58"><a href="#cb1-58"></a>      distance_sets[[j]][[<span class="st">"RF"</span>]][i, k] <span class="ot">=</span> distances[[<span class="dv">1</span>]]</span>
<span id="cb1-59"><a href="#cb1-59"></a>      distance_sets[[j]][[<span class="st">"KF"</span>]][i, k] <span class="ot">=</span> distances[[<span class="dv">2</span>]]</span>
<span id="cb1-60"><a href="#cb1-60"></a>      <span class="co">#dist_DB_pairwise[["path_diff"]][i, k] = distances[[3]]</span></span>
<span id="cb1-61"><a href="#cb1-61"></a>      distance_sets[[j]][[<span class="st">"MAST_%"</span>]][i, k] <span class="ot">=</span> <span class="fu">tryCatch</span>(</span>
<span id="cb1-62"><a href="#cb1-62"></a>        <span class="fu">length</span>(phangorn<span class="sc">::</span><span class="fu">mast</span>(tree1, tree2, <span class="at">tree =</span> <span class="cn">FALSE</span>)) <span class="sc">/</span> <span class="fu">length</span>(species_list),</span>
<span id="cb1-63"><a href="#cb1-63"></a>        <span class="at">error =</span> <span class="cf">function</span>(e) { skip_to_next <span class="ot">&lt;&lt;-</span> <span class="cn">TRUE</span> }</span>
<span id="cb1-64"><a href="#cb1-64"></a>      )</span>
<span id="cb1-65"><a href="#cb1-65"></a>      </span>
<span id="cb1-66"><a href="#cb1-66"></a>      <span class="co"># error handling</span></span>
<span id="cb1-67"><a href="#cb1-67"></a>      <span class="cf">if</span>(skip_to_next){distance_sets[[j]][[<span class="st">"MAST_%"</span>]][i, k] <span class="ot">=</span> <span class="cn">NA</span>}</span>
<span id="cb1-68"><a href="#cb1-68"></a>      </span>
<span id="cb1-69"><a href="#cb1-69"></a>    }</span>
<span id="cb1-70"><a href="#cb1-70"></a>  }</span>
<span id="cb1-71"><a href="#cb1-71"></a>}</span>
<span id="cb1-72"><a href="#cb1-72"></a></span>
<span id="cb1-73"><a href="#cb1-73"></a></span>
<span id="cb1-74"><a href="#cb1-74"></a><span class="do">### get pairwise common_nodes_corr distances</span></span>
<span id="cb1-75"><a href="#cb1-75"></a></span>
<span id="cb1-76"><a href="#cb1-76"></a><span class="co"># create CPU cluster</span></span>
<span id="cb1-77"><a href="#cb1-77"></a>cl <span class="ot">=</span> parallel<span class="sc">::</span><span class="fu">makeCluster</span>(<span class="dv">8</span>, <span class="at">type =</span> <span class="st">"SOCK"</span>)</span>
<span id="cb1-78"><a href="#cb1-78"></a><span class="fu">registerDoSNOW</span>(cl)</span>
<span id="cb1-79"><a href="#cb1-79"></a></span>
<span id="cb1-80"><a href="#cb1-80"></a><span class="co"># get chronos in parallel</span></span>
<span id="cb1-81"><a href="#cb1-81"></a>chronos_list <span class="ot">=</span> <span class="fu">foreach</span>(<span class="at">k =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(genes_sets[[<span class="st">"cp_mt_ML"</span>]]), <span class="at">.combine =</span> <span class="st">"c"</span>) <span class="sc">%dopar%</span> {</span>
<span id="cb1-82"><a href="#cb1-82"></a>  <span class="fu">library</span>(<span class="st">"ape"</span>)</span>
<span id="cb1-83"><a href="#cb1-83"></a>  <span class="co"># get tree</span></span>
<span id="cb1-84"><a href="#cb1-84"></a>  <span class="fu">tryCatch</span>(<span class="fu">chronos</span>(genes_sets[[<span class="st">"cp_mt_ML"</span>]][[k]]), <span class="at">error =</span> <span class="cf">function</span>(e) { <span class="cn">NA</span> })</span>
<span id="cb1-85"><a href="#cb1-85"></a>}  </span>
<span id="cb1-86"><a href="#cb1-86"></a>parallel<span class="sc">::</span><span class="fu">stopCluster</span>(cl)</span>
<span id="cb1-87"><a href="#cb1-87"></a><span class="fu">names</span>(chronos_list) <span class="ot">=</span> <span class="fu">names</span>(genes_sets[[<span class="st">"cp_mt_ML"</span>]])</span>
<span id="cb1-88"><a href="#cb1-88"></a></span>
<span id="cb1-89"><a href="#cb1-89"></a><span class="co"># get pairwise common_nodes_corr distances</span></span>
<span id="cb1-90"><a href="#cb1-90"></a><span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(distance_sets)){</span>
<span id="cb1-91"><a href="#cb1-91"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(genes_sets[[j]])){</span>
<span id="cb1-92"><a href="#cb1-92"></a>    <span class="cf">for</span>(k <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(genes_sets[[j]])){</span>
<span id="cb1-93"><a href="#cb1-93"></a>      </span>
<span id="cb1-94"><a href="#cb1-94"></a>      <span class="co"># get trees</span></span>
<span id="cb1-95"><a href="#cb1-95"></a>      tree1_tmp <span class="ot">=</span> chronos_list[[<span class="fu">names</span>(genes_sets[[j]])[i]]]</span>
<span id="cb1-96"><a href="#cb1-96"></a>      tree2_tmp <span class="ot">=</span> chronos_list[[<span class="fu">names</span>(genes_sets[[j]])[k]]]</span>
<span id="cb1-97"><a href="#cb1-97"></a>      </span>
<span id="cb1-98"><a href="#cb1-98"></a>      <span class="co"># check trees and get distance</span></span>
<span id="cb1-99"><a href="#cb1-99"></a>      <span class="cf">if</span>(<span class="fu">all</span>(<span class="sc">!</span><span class="fu">is.na</span>(tree1_tmp)) <span class="sc">&amp;</span> <span class="fu">all</span>(<span class="sc">!</span><span class="fu">is.na</span>(tree2_tmp))){</span>
<span id="cb1-100"><a href="#cb1-100"></a>        distance_sets[[j]][[<span class="st">"common_nodes_corr"</span>]][i, k] <span class="ot">=</span> <span class="fu">cor.dendlist</span>(</span>
<span id="cb1-101"><a href="#cb1-101"></a>          <span class="fu">dendlist</span>(<span class="fu">as.dendrogram</span>(tree1_tmp), <span class="fu">as.dendrogram</span>(tree2_tmp)),</span>
<span id="cb1-102"><a href="#cb1-102"></a>          <span class="at">method =</span> <span class="st">"common_nodes"</span>)[<span class="dv">2</span>]</span>
<span id="cb1-103"><a href="#cb1-103"></a>      } <span class="cf">else</span> {</span>
<span id="cb1-104"><a href="#cb1-104"></a>        distance_sets[[j]][[<span class="st">"common_nodes_corr"</span>]][i, k] <span class="ot">=</span> <span class="cn">NA</span></span>
<span id="cb1-105"><a href="#cb1-105"></a>      }</span>
<span id="cb1-106"><a href="#cb1-106"></a>    }</span>
<span id="cb1-107"><a href="#cb1-107"></a>  }</span>
<span id="cb1-108"><a href="#cb1-108"></a>}</span>
<span id="cb1-109"><a href="#cb1-109"></a></span>
<span id="cb1-110"><a href="#cb1-110"></a></span>
<span id="cb1-111"><a href="#cb1-111"></a><span class="co"># rename matrices for cp/mt overlapping rpl/rps genes (for plotting reasons)</span></span>
<span id="cb1-112"><a href="#cb1-112"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(distance_sets[[<span class="st">"cp_mt_genes_distances"</span>]])){</span>
<span id="cb1-113"><a href="#cb1-113"></a>  </span>
<span id="cb1-114"><a href="#cb1-114"></a>  <span class="fu">rownames</span>(distance_sets[[<span class="st">"cp_mt_genes_distances"</span>]][[i]]) <span class="ot">=</span> <span class="fu">c</span>(</span>
<span id="cb1-115"><a href="#cb1-115"></a>    <span class="st">"accD"</span>, <span class="st">"atpA"</span>, <span class="st">"atpB"</span>, <span class="st">"atpE"</span>, <span class="st">"atpF"</span>, <span class="st">"atpH"</span>, <span class="st">"atpI"</span>, <span class="st">"ccsA"</span>, <span class="st">"cemA"</span>, <span class="st">"chlI"</span>, <span class="st">"clpP"</span>,</span>
<span id="cb1-116"><a href="#cb1-116"></a>    <span class="st">"infA"</span>, <span class="st">"petA"</span>, <span class="st">"petB"</span>, <span class="st">"petD"</span>, <span class="st">"petG"</span>, <span class="st">"petL"</span>, <span class="st">"psaA"</span>, <span class="st">"psaB"</span>, <span class="st">"psaC"</span>, <span class="st">"psaI"</span>, <span class="st">"psaJ"</span>,</span>
<span id="cb1-117"><a href="#cb1-117"></a>    <span class="st">"psaM"</span>, <span class="st">"psbA"</span>, <span class="st">"psbB"</span>, <span class="st">"psbC"</span>, <span class="st">"psbD"</span>, <span class="st">"psbE"</span>, <span class="st">"psbF"</span>, <span class="st">"psbH"</span>, <span class="st">"psbI"</span>, <span class="st">"psbJ"</span>, <span class="st">"psbK"</span>,</span>
<span id="cb1-118"><a href="#cb1-118"></a>    <span class="st">"psbL"</span>, <span class="st">"psbM"</span>, <span class="st">"psbN"</span>, <span class="st">"psbT"</span>, <span class="st">"psbZ"</span>, <span class="st">"rbcL"</span>, <span class="st">"rpl12"</span>, <span class="st">"rpl14"</span>, <span class="st">"rpl16"</span>, <span class="st">"rpl19"</span>, <span class="st">"rpl2"</span>,</span>
<span id="cb1-119"><a href="#cb1-119"></a>    <span class="st">"rpl20"</span>, <span class="st">"rpl23"</span>, <span class="st">"rpl32"</span>, <span class="st">"rpl36"</span>, <span class="st">"rpl5"</span>, <span class="st">"rpoA"</span>, <span class="st">"rpoB"</span>, <span class="st">"rpoC1"</span>, <span class="st">"rpoC2"</span>, <span class="st">"rps11"</span>,</span>
<span id="cb1-120"><a href="#cb1-120"></a>    <span class="st">"rps12"</span>, <span class="st">"rps14"</span>, <span class="st">"rps18"</span>, <span class="st">"rps19"</span>, <span class="st">"rps2"</span>, <span class="st">"rps3"</span>,  <span class="st">"rps4"</span>, <span class="st">"rps7"</span>, <span class="st">"rps8"</span>, <span class="st">"rps9"</span>,</span>
<span id="cb1-121"><a href="#cb1-121"></a>    <span class="st">"tufA"</span>, <span class="st">"ycf1"</span>, <span class="st">"ycf12"</span>, <span class="st">"ycf20"</span>, <span class="st">"ycf3"</span>, <span class="st">"ycf4"</span>, <span class="st">"cp_mt_ML"</span>, <span class="st">"atp1"</span>, <span class="st">"atp4"</span>, <span class="st">"atp6"</span>,</span>
<span id="cb1-122"><a href="#cb1-122"></a>    <span class="st">"atp8"</span>, <span class="st">"atp9"</span>, <span class="st">"cob"</span>, <span class="st">"cox1"</span>, <span class="st">"cox2"</span>, <span class="st">"cox3"</span>, <span class="st">"nad1"</span>, <span class="st">"nad2"</span>, <span class="st">"nad3"</span>, <span class="st">"nad4"</span>, <span class="st">"nad4L"</span>,</span>
<span id="cb1-123"><a href="#cb1-123"></a>    <span class="st">"nad5"</span>, <span class="st">"nad6"</span>, <span class="st">"nad7"</span>, <span class="st">"rpl14 "</span>, <span class="st">"rpl16 "</span>, <span class="st">"rpl5 "</span>, <span class="st">"rps10"</span>, <span class="st">"rps11 "</span>, <span class="st">"rps12 "</span>, <span class="st">"rps13"</span>,</span>
<span id="cb1-124"><a href="#cb1-124"></a>    <span class="st">"rps14 "</span>, <span class="st">"rps19 "</span>, <span class="st">"rps2 "</span>, <span class="st">"rps3 "</span>, <span class="st">"rps4 "</span></span>
<span id="cb1-125"><a href="#cb1-125"></a>  )</span>
<span id="cb1-126"><a href="#cb1-126"></a>  </span>
<span id="cb1-127"><a href="#cb1-127"></a>  <span class="fu">colnames</span>(distance_sets[[<span class="st">"cp_mt_genes_distances"</span>]][[i]]) <span class="ot">=</span> <span class="fu">c</span>(</span>
<span id="cb1-128"><a href="#cb1-128"></a>    <span class="st">"accD"</span>, <span class="st">"atpA"</span>, <span class="st">"atpB"</span>, <span class="st">"atpE"</span>, <span class="st">"atpF"</span>, <span class="st">"atpH"</span>, <span class="st">"atpI"</span>, <span class="st">"ccsA"</span>, <span class="st">"cemA"</span>, <span class="st">"chlI"</span>, <span class="st">"clpP"</span>,</span>
<span id="cb1-129"><a href="#cb1-129"></a>    <span class="st">"infA"</span>, <span class="st">"petA"</span>, <span class="st">"petB"</span>, <span class="st">"petD"</span>, <span class="st">"petG"</span>, <span class="st">"petL"</span>, <span class="st">"psaA"</span>, <span class="st">"psaB"</span>, <span class="st">"psaC"</span>, <span class="st">"psaI"</span>, <span class="st">"psaJ"</span>,</span>
<span id="cb1-130"><a href="#cb1-130"></a>    <span class="st">"psaM"</span>, <span class="st">"psbA"</span>, <span class="st">"psbB"</span>, <span class="st">"psbC"</span>, <span class="st">"psbD"</span>, <span class="st">"psbE"</span>, <span class="st">"psbF"</span>, <span class="st">"psbH"</span>, <span class="st">"psbI"</span>, <span class="st">"psbJ"</span>, <span class="st">"psbK"</span>,</span>
<span id="cb1-131"><a href="#cb1-131"></a>    <span class="st">"psbL"</span>, <span class="st">"psbM"</span>, <span class="st">"psbN"</span>, <span class="st">"psbT"</span>, <span class="st">"psbZ"</span>, <span class="st">"rbcL"</span>, <span class="st">"rpl12"</span>, <span class="st">"rpl14"</span>, <span class="st">"rpl16"</span>, <span class="st">"rpl19"</span>, <span class="st">"rpl2"</span>,</span>
<span id="cb1-132"><a href="#cb1-132"></a>    <span class="st">"rpl20"</span>, <span class="st">"rpl23"</span>, <span class="st">"rpl32"</span>, <span class="st">"rpl36"</span>, <span class="st">"rpl5"</span>, <span class="st">"rpoA"</span>, <span class="st">"rpoB"</span>, <span class="st">"rpoC1"</span>, <span class="st">"rpoC2"</span>, <span class="st">"rps11"</span>,</span>
<span id="cb1-133"><a href="#cb1-133"></a>    <span class="st">"rps12"</span>, <span class="st">"rps14"</span>, <span class="st">"rps18"</span>, <span class="st">"rps19"</span>, <span class="st">"rps2"</span>, <span class="st">"rps3"</span>,  <span class="st">"rps4"</span>, <span class="st">"rps7"</span>, <span class="st">"rps8"</span>, <span class="st">"rps9"</span>,</span>
<span id="cb1-134"><a href="#cb1-134"></a>    <span class="st">"tufA"</span>, <span class="st">"ycf1"</span>, <span class="st">"ycf12"</span>, <span class="st">"ycf20"</span>, <span class="st">"ycf3"</span>, <span class="st">"ycf4"</span>, <span class="st">"cp_mt_ML"</span>, <span class="st">"atp1"</span>, <span class="st">"atp4"</span>, <span class="st">"atp6"</span>,</span>
<span id="cb1-135"><a href="#cb1-135"></a>    <span class="st">"atp8"</span>, <span class="st">"atp9"</span>, <span class="st">"cob"</span>, <span class="st">"cox1"</span>, <span class="st">"cox2"</span>, <span class="st">"cox3"</span>, <span class="st">"nad1"</span>, <span class="st">"nad2"</span>, <span class="st">"nad3"</span>, <span class="st">"nad4"</span>, <span class="st">"nad4L"</span>,</span>
<span id="cb1-136"><a href="#cb1-136"></a>    <span class="st">"nad5"</span>, <span class="st">"nad6"</span>, <span class="st">"nad7"</span>, <span class="st">"rpl14 "</span>, <span class="st">"rpl16 "</span>, <span class="st">"rpl5 "</span>, <span class="st">"rps10"</span>, <span class="st">"rps11 "</span>, <span class="st">"rps12 "</span>, <span class="st">"rps13"</span>,</span>
<span id="cb1-137"><a href="#cb1-137"></a>    <span class="st">"rps14 "</span>, <span class="st">"rps19 "</span>, <span class="st">"rps2 "</span>, <span class="st">"rps3 "</span>, <span class="st">"rps4 "</span></span>
<span id="cb1-138"><a href="#cb1-138"></a>  )</span>
<span id="cb1-139"><a href="#cb1-139"></a>  </span>
<span id="cb1-140"><a href="#cb1-140"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Very well, let’s now plot the generated distances!</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true">Correlation of the common nodes</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false">Robinson–Foulds distance</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-3" role="tab" aria-controls="tabset-1-3" aria-selected="false">Kuhner-Felsenstein distance</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-4" role="tab" aria-controls="tabset-1-4" aria-selected="false">Maximum Agreement SubTree (MAST %)</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<div class="cell" data-result="hide">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="co"># prepare gene lists</span></span>
<span id="cb2-2"><a href="#cb2-2"></a>cp_genes_list <span class="ot">=</span> <span class="fu">c</span>(</span>
<span id="cb2-3"><a href="#cb2-3"></a>  <span class="st">"accD"</span>, <span class="st">"atpA"</span>, <span class="st">"atpB"</span>, <span class="st">"atpE"</span>, <span class="st">"atpF"</span>, <span class="st">"atpH"</span>, <span class="st">"atpI"</span>, <span class="st">"ccsA"</span>, <span class="st">"cemA"</span>, <span class="st">"chlI"</span>, <span class="st">"clpP"</span>,</span>
<span id="cb2-4"><a href="#cb2-4"></a>  <span class="st">"infA"</span>, <span class="st">"petA"</span>, <span class="st">"petB"</span>, <span class="st">"petD"</span>, <span class="st">"petG"</span>, <span class="st">"petL"</span>, <span class="st">"psaA"</span>, <span class="st">"psaB"</span>, <span class="st">"psaC"</span>, <span class="st">"psaI"</span>, <span class="st">"psaJ"</span>,</span>
<span id="cb2-5"><a href="#cb2-5"></a>  <span class="st">"psaM"</span>, <span class="st">"psbA"</span>, <span class="st">"psbB"</span>, <span class="st">"psbC"</span>, <span class="st">"psbD"</span>, <span class="st">"psbE"</span>, <span class="st">"psbF"</span>, <span class="st">"psbH"</span>, <span class="st">"psbI"</span>, <span class="st">"psbJ"</span>, <span class="st">"psbK"</span>,</span>
<span id="cb2-6"><a href="#cb2-6"></a>  <span class="st">"psbL"</span>, <span class="st">"psbM"</span>, <span class="st">"psbN"</span>, <span class="st">"psbT"</span>, <span class="st">"psbZ"</span>, <span class="st">"rbcL"</span>, <span class="st">"rpl12"</span>, <span class="st">"rpl14"</span>, <span class="st">"rpl16"</span>, <span class="st">"rpl19"</span>, <span class="st">"rpl2"</span>,</span>
<span id="cb2-7"><a href="#cb2-7"></a>  <span class="st">"rpl20"</span>, <span class="st">"rpl23"</span>, <span class="st">"rpl32"</span>, <span class="st">"rpl36"</span>, <span class="st">"rpl5"</span>, <span class="st">"rpoA"</span>, <span class="st">"rpoB"</span>, <span class="st">"rpoC1"</span>, <span class="st">"rpoC2"</span>, <span class="st">"rps11"</span>,</span>
<span id="cb2-8"><a href="#cb2-8"></a>  <span class="st">"rps12"</span>, <span class="st">"rps14"</span>, <span class="st">"rps18"</span>, <span class="st">"rps19"</span>, <span class="st">"rps2"</span>, <span class="st">"rps3"</span>,  <span class="st">"rps4"</span>, <span class="st">"rps7"</span>, <span class="st">"rps8"</span>, <span class="st">"rps9"</span>,</span>
<span id="cb2-9"><a href="#cb2-9"></a>  <span class="st">"tufA"</span>, <span class="st">"ycf1"</span>, <span class="st">"ycf12"</span>, <span class="st">"ycf20"</span>, <span class="st">"ycf3"</span>, <span class="st">"ycf4"</span></span>
<span id="cb2-10"><a href="#cb2-10"></a>)</span>
<span id="cb2-11"><a href="#cb2-11"></a></span>
<span id="cb2-12"><a href="#cb2-12"></a>mt_genes_list <span class="ot">=</span> <span class="fu">c</span>(</span>
<span id="cb2-13"><a href="#cb2-13"></a>  <span class="st">"atp1"</span>, <span class="st">"atp4"</span>, <span class="st">"atp6"</span>, <span class="st">"atp8"</span>, <span class="st">"atp9"</span>, <span class="st">"cob"</span>, <span class="st">"cox1"</span>, <span class="st">"cox2"</span>, <span class="st">"cox3"</span>, <span class="st">"nad1"</span>, <span class="st">"nad2"</span>,</span>
<span id="cb2-14"><a href="#cb2-14"></a>  <span class="st">"nad3"</span>, <span class="st">"nad4"</span>, <span class="st">"nad4L"</span>, <span class="st">"nad5"</span>, <span class="st">"nad6"</span>, <span class="st">"nad7"</span>, <span class="st">"rpl14 "</span>, <span class="st">"rpl16 "</span>, <span class="st">"rpl5 "</span>, <span class="st">"rps10"</span>,</span>
<span id="cb2-15"><a href="#cb2-15"></a>  <span class="st">"rps11 "</span>, <span class="st">"rps12 "</span>, <span class="st">"rps13"</span>, <span class="st">"rps14 "</span>, <span class="st">"rps19 "</span>, <span class="st">"rps2 "</span>, <span class="st">"rps3 "</span>, <span class="st">"rps4 "</span></span>
<span id="cb2-16"><a href="#cb2-16"></a>)</span>
<span id="cb2-17"><a href="#cb2-17"></a></span>
<span id="cb2-18"><a href="#cb2-18"></a><span class="co"># remove possible rows/columns with only NAs</span></span>
<span id="cb2-19"><a href="#cb2-19"></a>tmp_matrix <span class="ot">=</span> distance_sets[[<span class="st">"cp_mt_genes_distances"</span>]][[<span class="st">"common_nodes_corr"</span>]]</span>
<span id="cb2-20"><a href="#cb2-20"></a>tmp_matrix <span class="ot">=</span> tmp_matrix[, <span class="fu">colSums</span>(<span class="fu">is.na</span>(tmp_matrix)) <span class="sc">&lt;</span> <span class="fu">nrow</span>(tmp_matrix)]</span>
<span id="cb2-21"><a href="#cb2-21"></a>tmp_matrix <span class="ot">=</span> tmp_matrix[<span class="fu">rowSums</span>(<span class="fu">is.na</span>(tmp_matrix)) <span class="sc">&lt;</span> <span class="fu">ncol</span>(tmp_matrix), ]</span>
<span id="cb2-22"><a href="#cb2-22"></a></span>
<span id="cb2-23"><a href="#cb2-23"></a><span class="co"># prep annotation</span></span>
<span id="cb2-24"><a href="#cb2-24"></a>organell <span class="ot">=</span> <span class="fu">ifelse</span>(<span class="fu">rownames</span>(tmp_matrix) <span class="sc">%in%</span> cp_genes_list, <span class="st">"cp"</span>,</span>
<span id="cb2-25"><a href="#cb2-25"></a>                  <span class="fu">ifelse</span>(<span class="fu">rownames</span>(tmp_matrix) <span class="sc">%in%</span> mt_genes_list, <span class="st">"mt"</span>, <span class="st">"cp_mt"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb2-26"><a href="#cb2-26"></a>              <span class="fu">as.data.frame</span>()</span>
<span id="cb2-27"><a href="#cb2-27"></a><span class="fu">rownames</span>(organell) <span class="ot">=</span> <span class="fu">rownames</span>(tmp_matrix)</span>
<span id="cb2-28"><a href="#cb2-28"></a><span class="fu">colnames</span>(organell) <span class="ot">=</span> <span class="st">"organell"</span></span>
<span id="cb2-29"><a href="#cb2-29"></a>organell<span class="sc">$</span>color <span class="ot">=</span> <span class="fu">ifelse</span>(organell<span class="sc">$</span>organell <span class="sc">==</span> <span class="st">"cp"</span>, <span class="st">"#469d89"</span>, <span class="fu">ifelse</span>(organell<span class="sc">$</span>organell <span class="sc">==</span> <span class="st">"mt"</span>, <span class="st">"steelblue"</span>, <span class="st">"red"</span>))</span>
<span id="cb2-30"><a href="#cb2-30"></a></span>
<span id="cb2-31"><a href="#cb2-31"></a><span class="co"># plot raw distance</span></span>
<span id="cb2-32"><a href="#cb2-32"></a>ComplexHeatmap<span class="sc">::</span><span class="fu">Heatmap</span>(</span>
<span id="cb2-33"><a href="#cb2-33"></a>  tmp_matrix,</span>
<span id="cb2-34"><a href="#cb2-34"></a>  <span class="at">col =</span> <span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="fu">min</span>(tmp_matrix[<span class="sc">!</span><span class="fu">is.na</span>(tmp_matrix)]), <span class="fu">max</span>(tmp_matrix[<span class="sc">!</span><span class="fu">is.na</span>(tmp_matrix)])),</span>
<span id="cb2-35"><a href="#cb2-35"></a>                   <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"steelblue"</span>)),</span>
<span id="cb2-36"><a href="#cb2-36"></a>  <span class="at">row_names_gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> organell<span class="sc">$</span>color),</span>
<span id="cb2-37"><a href="#cb2-37"></a>  <span class="at">column_names_gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> organell<span class="sc">$</span>color),</span>
<span id="cb2-38"><a href="#cb2-38"></a>  <span class="at">top_annotation =</span> <span class="fu">HeatmapAnnotation</span>(</span>
<span id="cb2-39"><a href="#cb2-39"></a>    <span class="at">organell =</span> <span class="fu">as.matrix</span>(organell<span class="sc">$</span>organell),</span>
<span id="cb2-40"><a href="#cb2-40"></a>    <span class="at">show_annotation_name =</span> <span class="cn">FALSE</span>,</span>
<span id="cb2-41"><a href="#cb2-41"></a>    <span class="at">show_legend =</span> <span class="cn">FALSE</span>,</span>
<span id="cb2-42"><a href="#cb2-42"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="at">organell =</span> <span class="fu">c</span>(<span class="st">"cp_mt"</span> <span class="ot">=</span> <span class="st">"red"</span>, <span class="st">"cp"</span> <span class="ot">=</span> <span class="st">"#469d89"</span>, <span class="st">"mt"</span> <span class="ot">=</span> <span class="st">"steelblue"</span>))),</span>
<span id="cb2-43"><a href="#cb2-43"></a>  <span class="at">left_annotation =</span> <span class="fu">rowAnnotation</span>(</span>
<span id="cb2-44"><a href="#cb2-44"></a>    <span class="at">organell =</span> <span class="fu">as.matrix</span>(organell<span class="sc">$</span>organell),</span>
<span id="cb2-45"><a href="#cb2-45"></a>    <span class="at">show_annotation_name =</span> <span class="cn">FALSE</span>,</span>
<span id="cb2-46"><a href="#cb2-46"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="at">organell =</span> <span class="fu">c</span>(<span class="st">"cp_mt"</span> <span class="ot">=</span> <span class="st">"red"</span>, <span class="st">"cp"</span> <span class="ot">=</span> <span class="st">"#469d89"</span>, <span class="st">"mt"</span> <span class="ot">=</span> <span class="st">"steelblue"</span>))),</span>
<span id="cb2-47"><a href="#cb2-47"></a>  <span class="at">name =</span> <span class="st">"common</span><span class="sc">\n</span><span class="st">nodes</span><span class="sc">\n</span><span class="st">corr"</span></span>
<span id="cb2-48"><a href="#cb2-48"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="markers-selection_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption>Figure 1: Common node correlation between single-gene tree and reconstructed <em>Ulva</em> species tree.</figcaption>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<div class="cell" data-result="hide">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="co"># remove possible rows/columns with only NAs</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>tmp_matrix <span class="ot">=</span> distance_sets[[<span class="st">"cp_mt_genes_distances"</span>]][[<span class="st">"RF"</span>]]</span>
<span id="cb3-3"><a href="#cb3-3"></a>tmp_matrix <span class="ot">=</span> tmp_matrix[, <span class="fu">colSums</span>(<span class="fu">is.na</span>(tmp_matrix)) <span class="sc">&lt;</span> <span class="fu">nrow</span>(tmp_matrix)]</span>
<span id="cb3-4"><a href="#cb3-4"></a>tmp_matrix <span class="ot">=</span> tmp_matrix[<span class="fu">rowSums</span>(<span class="fu">is.na</span>(tmp_matrix)) <span class="sc">&lt;</span> <span class="fu">ncol</span>(tmp_matrix), ]</span>
<span id="cb3-5"><a href="#cb3-5"></a></span>
<span id="cb3-6"><a href="#cb3-6"></a><span class="co"># prep annotation</span></span>
<span id="cb3-7"><a href="#cb3-7"></a>organell <span class="ot">=</span> <span class="fu">ifelse</span>(<span class="fu">rownames</span>(tmp_matrix) <span class="sc">%in%</span> cp_genes_list, <span class="st">"cp"</span>,</span>
<span id="cb3-8"><a href="#cb3-8"></a>                  <span class="fu">ifelse</span>(<span class="fu">rownames</span>(tmp_matrix) <span class="sc">%in%</span> mt_genes_list, <span class="st">"mt"</span>, <span class="st">"cp_mt"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb3-9"><a href="#cb3-9"></a>              <span class="fu">as.data.frame</span>()</span>
<span id="cb3-10"><a href="#cb3-10"></a><span class="fu">rownames</span>(organell) <span class="ot">=</span> <span class="fu">rownames</span>(tmp_matrix)</span>
<span id="cb3-11"><a href="#cb3-11"></a><span class="fu">colnames</span>(organell) <span class="ot">=</span> <span class="st">"organell"</span></span>
<span id="cb3-12"><a href="#cb3-12"></a>organell<span class="sc">$</span>color <span class="ot">=</span> <span class="fu">ifelse</span>(organell<span class="sc">$</span>organell <span class="sc">==</span> <span class="st">"cp"</span>, <span class="st">"#469d89"</span>, <span class="fu">ifelse</span>(organell<span class="sc">$</span>organell <span class="sc">==</span> <span class="st">"mt"</span>, <span class="st">"steelblue"</span>, <span class="st">"red"</span>))</span>
<span id="cb3-13"><a href="#cb3-13"></a></span>
<span id="cb3-14"><a href="#cb3-14"></a><span class="co"># plot raw distance</span></span>
<span id="cb3-15"><a href="#cb3-15"></a>ComplexHeatmap<span class="sc">::</span><span class="fu">Heatmap</span>(</span>
<span id="cb3-16"><a href="#cb3-16"></a>  tmp_matrix,</span>
<span id="cb3-17"><a href="#cb3-17"></a>  <span class="at">col =</span> <span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="fu">min</span>(tmp_matrix[<span class="sc">!</span><span class="fu">is.na</span>(tmp_matrix)]), <span class="fu">max</span>(tmp_matrix[<span class="sc">!</span><span class="fu">is.na</span>(tmp_matrix)])),</span>
<span id="cb3-18"><a href="#cb3-18"></a>                   <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"#D55E00"</span>)),</span>
<span id="cb3-19"><a href="#cb3-19"></a>  <span class="at">row_names_gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> organell<span class="sc">$</span>color),</span>
<span id="cb3-20"><a href="#cb3-20"></a>  <span class="at">column_names_gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> organell<span class="sc">$</span>color),</span>
<span id="cb3-21"><a href="#cb3-21"></a>  <span class="at">top_annotation =</span> <span class="fu">HeatmapAnnotation</span>(</span>
<span id="cb3-22"><a href="#cb3-22"></a>    <span class="at">organell =</span> <span class="fu">as.matrix</span>(organell<span class="sc">$</span>organell),</span>
<span id="cb3-23"><a href="#cb3-23"></a>    <span class="at">show_annotation_name =</span> <span class="cn">FALSE</span>,</span>
<span id="cb3-24"><a href="#cb3-24"></a>    <span class="at">show_legend =</span> <span class="cn">FALSE</span>,</span>
<span id="cb3-25"><a href="#cb3-25"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="at">organell =</span> <span class="fu">c</span>(<span class="st">"cp_mt"</span> <span class="ot">=</span> <span class="st">"red"</span>, <span class="st">"cp"</span> <span class="ot">=</span> <span class="st">"#469d89"</span>, <span class="st">"mt"</span> <span class="ot">=</span> <span class="st">"steelblue"</span>))),</span>
<span id="cb3-26"><a href="#cb3-26"></a>  <span class="at">left_annotation =</span> <span class="fu">rowAnnotation</span>(</span>
<span id="cb3-27"><a href="#cb3-27"></a>    <span class="at">organell =</span> <span class="fu">as.matrix</span>(organell<span class="sc">$</span>organell),</span>
<span id="cb3-28"><a href="#cb3-28"></a>    <span class="at">show_annotation_name =</span> <span class="cn">FALSE</span>,</span>
<span id="cb3-29"><a href="#cb3-29"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="at">organell =</span> <span class="fu">c</span>(<span class="st">"cp_mt"</span> <span class="ot">=</span> <span class="st">"red"</span>, <span class="st">"cp"</span> <span class="ot">=</span> <span class="st">"#469d89"</span>, <span class="st">"mt"</span> <span class="ot">=</span> <span class="st">"steelblue"</span>))),</span>
<span id="cb3-30"><a href="#cb3-30"></a>  <span class="at">name =</span> <span class="st">"Robinson–Foulds"</span></span>
<span id="cb3-31"><a href="#cb3-31"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="markers-selection_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption>Figure 2: Robinson–Foulds distance between single-gene tree and reconstructed <em>Ulva</em> species tree.</figcaption>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-1-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-3-tab">
<div class="cell" data-result="hide">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="co"># remove possible rows/columns with only NAs</span></span>
<span id="cb4-2"><a href="#cb4-2"></a>tmp_matrix <span class="ot">=</span> distance_sets[[<span class="st">"cp_mt_genes_distances"</span>]][[<span class="st">"KF"</span>]]</span>
<span id="cb4-3"><a href="#cb4-3"></a>tmp_matrix <span class="ot">=</span> tmp_matrix[, <span class="fu">colSums</span>(<span class="fu">is.na</span>(tmp_matrix)) <span class="sc">&lt;</span> <span class="fu">nrow</span>(tmp_matrix)]</span>
<span id="cb4-4"><a href="#cb4-4"></a>tmp_matrix <span class="ot">=</span> tmp_matrix[<span class="fu">rowSums</span>(<span class="fu">is.na</span>(tmp_matrix)) <span class="sc">&lt;</span> <span class="fu">ncol</span>(tmp_matrix), ]</span>
<span id="cb4-5"><a href="#cb4-5"></a></span>
<span id="cb4-6"><a href="#cb4-6"></a><span class="co"># prep annotation</span></span>
<span id="cb4-7"><a href="#cb4-7"></a>organell <span class="ot">=</span> <span class="fu">ifelse</span>(<span class="fu">rownames</span>(tmp_matrix) <span class="sc">%in%</span> cp_genes_list, <span class="st">"cp"</span>,</span>
<span id="cb4-8"><a href="#cb4-8"></a>                  <span class="fu">ifelse</span>(<span class="fu">rownames</span>(tmp_matrix) <span class="sc">%in%</span> mt_genes_list, <span class="st">"mt"</span>, <span class="st">"cp_mt"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb4-9"><a href="#cb4-9"></a>              <span class="fu">as.data.frame</span>()</span>
<span id="cb4-10"><a href="#cb4-10"></a><span class="fu">rownames</span>(organell) <span class="ot">=</span> <span class="fu">rownames</span>(tmp_matrix)</span>
<span id="cb4-11"><a href="#cb4-11"></a><span class="fu">colnames</span>(organell) <span class="ot">=</span> <span class="st">"organell"</span></span>
<span id="cb4-12"><a href="#cb4-12"></a>organell<span class="sc">$</span>color <span class="ot">=</span> <span class="fu">ifelse</span>(organell<span class="sc">$</span>organell <span class="sc">==</span> <span class="st">"cp"</span>, <span class="st">"#469d89"</span>, <span class="fu">ifelse</span>(organell<span class="sc">$</span>organell <span class="sc">==</span> <span class="st">"mt"</span>, <span class="st">"steelblue"</span>, <span class="st">"red"</span>))</span>
<span id="cb4-13"><a href="#cb4-13"></a></span>
<span id="cb4-14"><a href="#cb4-14"></a><span class="co"># plot raw distance</span></span>
<span id="cb4-15"><a href="#cb4-15"></a>ComplexHeatmap<span class="sc">::</span><span class="fu">Heatmap</span>(</span>
<span id="cb4-16"><a href="#cb4-16"></a>  tmp_matrix,</span>
<span id="cb4-17"><a href="#cb4-17"></a>  <span class="at">col =</span> <span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="fu">min</span>(tmp_matrix[<span class="sc">!</span><span class="fu">is.na</span>(tmp_matrix)]), <span class="fu">max</span>(tmp_matrix[<span class="sc">!</span><span class="fu">is.na</span>(tmp_matrix)])),</span>
<span id="cb4-18"><a href="#cb4-18"></a>                   <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"#D55E00"</span>)),</span>
<span id="cb4-19"><a href="#cb4-19"></a>  <span class="at">row_names_gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> organell<span class="sc">$</span>color),</span>
<span id="cb4-20"><a href="#cb4-20"></a>  <span class="at">column_names_gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> organell<span class="sc">$</span>color),</span>
<span id="cb4-21"><a href="#cb4-21"></a>  <span class="at">top_annotation =</span> <span class="fu">HeatmapAnnotation</span>(</span>
<span id="cb4-22"><a href="#cb4-22"></a>    <span class="at">organell =</span> <span class="fu">as.matrix</span>(organell<span class="sc">$</span>organell),</span>
<span id="cb4-23"><a href="#cb4-23"></a>    <span class="at">show_annotation_name =</span> <span class="cn">FALSE</span>,</span>
<span id="cb4-24"><a href="#cb4-24"></a>    <span class="at">show_legend =</span> <span class="cn">FALSE</span>,</span>
<span id="cb4-25"><a href="#cb4-25"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="at">organell =</span> <span class="fu">c</span>(<span class="st">"cp_mt"</span> <span class="ot">=</span> <span class="st">"red"</span>, <span class="st">"cp"</span> <span class="ot">=</span> <span class="st">"#469d89"</span>, <span class="st">"mt"</span> <span class="ot">=</span> <span class="st">"steelblue"</span>))),</span>
<span id="cb4-26"><a href="#cb4-26"></a>  <span class="at">left_annotation =</span> <span class="fu">rowAnnotation</span>(</span>
<span id="cb4-27"><a href="#cb4-27"></a>    <span class="at">organell =</span> <span class="fu">as.matrix</span>(organell<span class="sc">$</span>organell),</span>
<span id="cb4-28"><a href="#cb4-28"></a>    <span class="at">show_annotation_name =</span> <span class="cn">FALSE</span>,</span>
<span id="cb4-29"><a href="#cb4-29"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="at">organell =</span> <span class="fu">c</span>(<span class="st">"cp_mt"</span> <span class="ot">=</span> <span class="st">"red"</span>, <span class="st">"cp"</span> <span class="ot">=</span> <span class="st">"#469d89"</span>, <span class="st">"mt"</span> <span class="ot">=</span> <span class="st">"steelblue"</span>))),</span>
<span id="cb4-30"><a href="#cb4-30"></a>  <span class="at">name =</span> <span class="st">"Kuhner-Felsenstein</span><span class="sc">\n</span><span class="st">distance"</span></span>
<span id="cb4-31"><a href="#cb4-31"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="markers-selection_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption>Figure 3: Kuhner-Felsenstein distance between single-gene tree and reconstructed <em>Ulva</em> species tree.</figcaption>
</figure>
</div>
</div>
</div>
</div>
<div id="tabset-1-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-4-tab">
<div class="cell" data-result="hide">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="co"># remove possible rows/columns with only NAs</span></span>
<span id="cb5-2"><a href="#cb5-2"></a>tmp_matrix <span class="ot">=</span> distance_sets[[<span class="st">"cp_mt_genes_distances"</span>]][[<span class="st">"MAST_%"</span>]]</span>
<span id="cb5-3"><a href="#cb5-3"></a>tmp_matrix <span class="ot">=</span> tmp_matrix[, <span class="fu">colSums</span>(<span class="fu">is.na</span>(tmp_matrix)) <span class="sc">&lt;</span> <span class="fu">nrow</span>(tmp_matrix)]</span>
<span id="cb5-4"><a href="#cb5-4"></a>tmp_matrix <span class="ot">=</span> tmp_matrix[<span class="fu">rowSums</span>(<span class="fu">is.na</span>(tmp_matrix)) <span class="sc">&lt;</span> <span class="fu">ncol</span>(tmp_matrix), ]</span>
<span id="cb5-5"><a href="#cb5-5"></a></span>
<span id="cb5-6"><a href="#cb5-6"></a><span class="co"># prep annotation</span></span>
<span id="cb5-7"><a href="#cb5-7"></a>organell <span class="ot">=</span> <span class="fu">ifelse</span>(<span class="fu">rownames</span>(tmp_matrix) <span class="sc">%in%</span> cp_genes_list, <span class="st">"cp"</span>,</span>
<span id="cb5-8"><a href="#cb5-8"></a>                  <span class="fu">ifelse</span>(<span class="fu">rownames</span>(tmp_matrix) <span class="sc">%in%</span> mt_genes_list, <span class="st">"mt"</span>, <span class="st">"cp_mt"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb5-9"><a href="#cb5-9"></a>              <span class="fu">as.data.frame</span>()</span>
<span id="cb5-10"><a href="#cb5-10"></a><span class="fu">rownames</span>(organell) <span class="ot">=</span> <span class="fu">rownames</span>(tmp_matrix)</span>
<span id="cb5-11"><a href="#cb5-11"></a><span class="fu">colnames</span>(organell) <span class="ot">=</span> <span class="st">"organell"</span></span>
<span id="cb5-12"><a href="#cb5-12"></a>organell<span class="sc">$</span>color <span class="ot">=</span> <span class="fu">ifelse</span>(organell<span class="sc">$</span>organell <span class="sc">==</span> <span class="st">"cp"</span>, <span class="st">"#469d89"</span>, <span class="fu">ifelse</span>(organell<span class="sc">$</span>organell <span class="sc">==</span> <span class="st">"mt"</span>, <span class="st">"steelblue"</span>, <span class="st">"red"</span>))</span>
<span id="cb5-13"><a href="#cb5-13"></a></span>
<span id="cb5-14"><a href="#cb5-14"></a><span class="co"># plot raw distance</span></span>
<span id="cb5-15"><a href="#cb5-15"></a>ComplexHeatmap<span class="sc">::</span><span class="fu">Heatmap</span>(</span>
<span id="cb5-16"><a href="#cb5-16"></a>  tmp_matrix,</span>
<span id="cb5-17"><a href="#cb5-17"></a>  <span class="at">col =</span> <span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="fu">min</span>(tmp_matrix[<span class="sc">!</span><span class="fu">is.na</span>(tmp_matrix)]), <span class="fu">max</span>(tmp_matrix[<span class="sc">!</span><span class="fu">is.na</span>(tmp_matrix)])),</span>
<span id="cb5-18"><a href="#cb5-18"></a>                   <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"steelblue"</span>)),</span>
<span id="cb5-19"><a href="#cb5-19"></a>  <span class="at">row_names_gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> organell<span class="sc">$</span>color),</span>
<span id="cb5-20"><a href="#cb5-20"></a>  <span class="at">column_names_gp =</span> <span class="fu">gpar</span>(<span class="at">col =</span> organell<span class="sc">$</span>color),</span>
<span id="cb5-21"><a href="#cb5-21"></a>  <span class="at">top_annotation =</span> <span class="fu">HeatmapAnnotation</span>(</span>
<span id="cb5-22"><a href="#cb5-22"></a>    <span class="at">organell =</span> <span class="fu">as.matrix</span>(organell<span class="sc">$</span>organell),</span>
<span id="cb5-23"><a href="#cb5-23"></a>    <span class="at">show_annotation_name =</span> <span class="cn">FALSE</span>,</span>
<span id="cb5-24"><a href="#cb5-24"></a>    <span class="at">show_legend =</span> <span class="cn">FALSE</span>,</span>
<span id="cb5-25"><a href="#cb5-25"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="at">organell =</span> <span class="fu">c</span>(<span class="st">"cp_mt"</span> <span class="ot">=</span> <span class="st">"red"</span>, <span class="st">"cp"</span> <span class="ot">=</span> <span class="st">"#469d89"</span>, <span class="st">"mt"</span> <span class="ot">=</span> <span class="st">"steelblue"</span>))),</span>
<span id="cb5-26"><a href="#cb5-26"></a>  <span class="at">left_annotation =</span> <span class="fu">rowAnnotation</span>(</span>
<span id="cb5-27"><a href="#cb5-27"></a>    <span class="at">organell =</span> <span class="fu">as.matrix</span>(organell<span class="sc">$</span>organell),</span>
<span id="cb5-28"><a href="#cb5-28"></a>    <span class="at">show_annotation_name =</span> <span class="cn">FALSE</span>,</span>
<span id="cb5-29"><a href="#cb5-29"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="at">organell =</span> <span class="fu">c</span>(<span class="st">"cp_mt"</span> <span class="ot">=</span> <span class="st">"red"</span>, <span class="st">"cp"</span> <span class="ot">=</span> <span class="st">"#469d89"</span>, <span class="st">"mt"</span> <span class="ot">=</span> <span class="st">"steelblue"</span>))),</span>
<span id="cb5-30"><a href="#cb5-30"></a>  <span class="at">name =</span> <span class="st">"MAST_%"</span></span>
<span id="cb5-31"><a href="#cb5-31"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="markers-selection_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption>Figure 4: Maximum Agreement SubTree % between single-gene tree and reconstructed <em>Ulva</em> species tree.</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<p>The four indices RF, KF, MAST and the correlation between the common nodes metrics pointed out 24 chloroplast genes that provide the most similar topology compared to the <em>Ulva</em> species tree, namely <em>accD</em>, <em>atpF</em>, <em>chlI</em>, <em>petB</em>, <em>psaA</em>, <em>psaB</em>, <em>psbB</em>, <em>psbD</em>, <em>rpl12</em>, <em>rpl19</em>, <em>rpl2</em>, <em>rpl32</em>, <em>rpoA</em>, <em>rpoB</em>, <em>rpoC1</em>, <em>rpoC2</em>, <em>rps12</em>, <em>rps2</em>, <em>rps4</em>, <em>rps8</em>, <em>rps9</em>, <em>tufA</em>, <em>ycf1</em>, and <em>ycf20</em>. Among them, several genes such as <em>psaA</em>, <em>psaB</em>, and <em>psbB</em> are consistently identified by different matrices.</p>
<p>For the mitochondrial, 19 genes provide the most similar topology compared to the <em>Ulva</em> species tree, namely <em>atp1</em>, <em>atp6</em>, <em>cob</em>, <em>cox1</em>, <em>cox2</em>, <em>cox3</em>, <em>nad2</em>, <em>nad4</em>, <em>nad5</em>, <em>nad7</em>, <em>rpl14</em>, <em>rpl16</em>, <em>rps10</em>, <em>rps11</em>, <em>rps14</em>, <em>rps19</em>, <em>rps2</em>, <em>rps3</em>, and <em>rps4</em>.</p>
<p>For each dataset and distance metric, we report below the top 10 single genes most similar to the <em>Ulva</em> species tree that we have reconstructed. The genes are ordered based on the decreasing values of the corresponding index.</p>
<div class="cell">
<div class="cell-output-display">
<div class="datatables html-widget html-fill-item" id="htmlwidget-c0f8198cbd7f7aa5b843" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-c0f8198cbd7f7aa5b843">{"x":{"filter":"none","vertical":false,"class":"display","extensions":["FixedColumns","FixedHeader"],"caption":"<caption>Table 1: top 10 genes closest to the reconstructed Ulva species tree.<\/caption>","data":[["1","2","3","4","5","6","7","8","9","10"],["psaA","rpoC2","rpoA","rpoC1","psaB","psbB","rps12","ycf1","chlI","rps2"],["psbB","psaA","rps2","petB","rpoC1","psaB","psbD","rpl12","rpoC2","rpoA"],["rpoA","ycf20","rpoB","rpl32","rps8","accD","atpF","rps4","rps9","rpl2"],["psaA","psaB","ycf1","chlI","rpoA","psbB","rpl19","tufA","rpoC1","accD"],["rps10","rps3","cox1","rps2","atp1","atp6","cox2","nad4","rps4","nad2"],["cox1","cox2","atp1","cox3","rps4","rps2","rps3","cob","rps19","rps10"],["rps3","rps10","rps2","nad5","nad7","cox1","rpl16","rpl14","rps14","rps4"],["cox2","atp6","cox3","nad5","rpl16","rps11","rps4","nad2","nad4","cox1"],["psaA","rpoC2","rpoA","rpoC1","psaB","psbB","rps12","ycf1","chlI","rps2"],["psbB","psaA","rps2","cox1","cox2","petB","rpoC1","psaB","psbD","rpl12"],["mt_rps3","mt_rps10","rpoA","mt_rps2","nad5","ycf20","rpoB","rpl32","rps8","accD"],["psaA","psaB","ycf1","cox2","chlI","rpoA","psbB","rpl19","tufA","atp6"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th rowspan=\"2\">Rank<\/th>\n      <th colspan=\"4\">chloroplast<\/th>\n      <th colspan=\"4\">mitochondria<\/th>\n      <th colspan=\"4\">chloroplast + mitochondria<\/th>\n    <\/tr>\n    <tr>\n      <th>RF<\/th>\n      <th>MAST %<\/th>\n      <th>KF<\/th>\n      <th>comm. node corr.<\/th>\n      <th>RF<\/th>\n      <th>MAST %<\/th>\n      <th>KF<\/th>\n      <th>comm. node corr.<\/th>\n      <th>RF<\/th>\n      <th>MAST %<\/th>\n      <th>KF<\/th>\n      <th>comm. node corr.<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":false,"pageLength":10,"columnDefs":[{"name":"rank","targets":0},{"name":"cp.RF","targets":1},{"name":"cp.MAST.","targets":2},{"name":"cp.KF","targets":3},{"name":"cp.corr","targets":4},{"name":"mt.RF","targets":5},{"name":"mt.MAST.","targets":6},{"name":"mt.KF","targets":7},{"name":"mt.corr","targets":8},{"name":"cp_mt.RF","targets":9},{"name":"cp_mt.MAST.","targets":10},{"name":"cp_mt.KF","targets":11},{"name":"cp_mt.corr","targets":12}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
</section>
</section>
<section id="markers-selection" class="level2" data-number="3.3">
<h2 data-number="3.3" class="anchored" data-anchor-id="markers-selection"><span class="header-section-number">3.3</span> Markers selection</h2>
<p>Prior analyses suggested that MAST% and common correlation nodes metrics are the best performers to assess phylogenetic tree distances when the tree do not have the same set of tips, we then selected (arbitrarily) 10 genes for the combinatorial approach, <em>petB</em>, <em>psaA</em>, <em>psaB</em>, <em>psbB</em>, <em>psbD</em>, <em>rps2</em> from the chloroplast and <em>atp6</em>, <em>cox1</em>, <em>cox2</em>, and <em>rps3</em> from the mitochondria.</p>
<p>With the combinatorial approach, we can now compute their 2:10 genes combination, concatenate their alignment and reconstruct the corresponding ML tree.</p>
<div class="cell" data-result="hide">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource bash bash-code number-lines code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1"></a><span class="co">#!/usr/bin/env bash</span></span>
<span id="cb6-2"><a href="#cb6-2"></a></span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="ex">python3.10</span> combinatorial_ML_phylogeny.py <span class="dt">\</span></span>
<span id="cb6-4"><a href="#cb6-4"></a>  <span class="at">--infolder</span> <span class="st">"./09_combinatorial_phylogeny/00_cp_genes/,./09_combinatorial_phylogeny/00_mt_genes/"</span> <span class="dt">\</span></span>
<span id="cb6-5"><a href="#cb6-5"></a>  <span class="at">--names</span> <span class="st">"CP,MT"</span> <span class="dt">\</span></span>
<span id="cb6-6"><a href="#cb6-6"></a>  <span class="at">--iqtree</span> ~/bin/iqtree-2.2.0-Linux/bin/iqtree2 <span class="dt">\</span></span>
<span id="cb6-7"><a href="#cb6-7"></a>  <span class="at">--n_genes</span> 10 <span class="dt">\</span></span>
<span id="cb6-8"><a href="#cb6-8"></a>  <span class="at">--outdir</span> ./09_combinatorial_phylogeny/</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Awesome!</p>
<p>we have generated all possible 2-10 genes combination out of our selection, and we have concatenated the corresponding nucleotidic sequences and we have reconstructed their phylogenetic trees.</p>
<p>The combinatorial approach generated 945 unique combinations of at least a chloroplast and a mitochondrial marker:</p>
<ul>
<li>24 combinations of 2 marker genes;</li>
<li>96 combinations of 3 markers;</li>
<li>194 combinations of 4 markers;</li>
<li>246 combinations of 5 markers;</li>
<li>209 combinations of 6 markers;</li>
<li>120 combinations of 7 markers;</li>
<li>45 combinations of 8 markers;</li>
<li>10 combinations of 9 markers;</li>
<li>finally 1 containing all 10 markers genes</li>
</ul>
<p>Let’s now see what is the minimum set of chloroplast and mitochondrial marker that approximate the reconstructed <em>Ulva</em> species tree.</p>
<p>First, we import all the generated combinatorial trees.</p>
<div class="cell" data-result="hide">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="co"># declare gene lists</span></span>
<span id="cb7-2"><a href="#cb7-2"></a>cp_genes <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"petB"</span>, <span class="st">"psaA"</span>, <span class="st">"psaB"</span>, <span class="st">"psbB"</span>, <span class="st">"psbD"</span>, <span class="st">"rps2"</span>)</span>
<span id="cb7-3"><a href="#cb7-3"></a>mt_genes <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"atp6"</span>, <span class="st">"cox1"</span>, <span class="st">"cox2"</span>, <span class="st">"rps3"</span>)</span>
<span id="cb7-4"><a href="#cb7-4"></a></span>
<span id="cb7-5"><a href="#cb7-5"></a><span class="co"># import ref species tree</span></span>
<span id="cb7-6"><a href="#cb7-6"></a>cp_mt_ML_dendro <span class="ot">=</span> ape<span class="sc">::</span><span class="fu">read.tree</span>(<span class="at">file =</span> <span class="st">"06_cp_mt_concat_ML/cp_mt_allgenes_concat.contree"</span>)</span>
<span id="cb7-7"><a href="#cb7-7"></a></span>
<span id="cb7-8"><a href="#cb7-8"></a><span class="co"># create empty list of trees</span></span>
<span id="cb7-9"><a href="#cb7-9"></a>combinatorial_trees <span class="ot">=</span> <span class="fu">list</span>(</span>
<span id="cb7-10"><a href="#cb7-10"></a>  <span class="st">"cp_mt_conc_2_genes"</span> <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_ML"</span> <span class="ot">=</span> cp_mt_ML_dendro),</span>
<span id="cb7-11"><a href="#cb7-11"></a>  <span class="st">"cp_mt_conc_3_genes"</span> <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_ML"</span> <span class="ot">=</span> cp_mt_ML_dendro),</span>
<span id="cb7-12"><a href="#cb7-12"></a>  <span class="st">"cp_mt_conc_4_genes"</span> <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_ML"</span> <span class="ot">=</span> cp_mt_ML_dendro),</span>
<span id="cb7-13"><a href="#cb7-13"></a>  <span class="st">"cp_mt_conc_5_genes"</span> <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_ML"</span> <span class="ot">=</span> cp_mt_ML_dendro),</span>
<span id="cb7-14"><a href="#cb7-14"></a>  <span class="st">"cp_mt_conc_6_genes"</span> <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_ML"</span> <span class="ot">=</span> cp_mt_ML_dendro),</span>
<span id="cb7-15"><a href="#cb7-15"></a>  <span class="st">"cp_mt_conc_7_genes"</span> <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_ML"</span> <span class="ot">=</span> cp_mt_ML_dendro),</span>
<span id="cb7-16"><a href="#cb7-16"></a>  <span class="st">"cp_mt_conc_8_genes"</span> <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_ML"</span> <span class="ot">=</span> cp_mt_ML_dendro),</span>
<span id="cb7-17"><a href="#cb7-17"></a>  <span class="st">"cp_mt_conc_9_genes"</span> <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_ML"</span> <span class="ot">=</span> cp_mt_ML_dendro),</span>
<span id="cb7-18"><a href="#cb7-18"></a>  <span class="st">"cp_mt_conc_10_genes"</span> <span class="ot">=</span> <span class="fu">list</span>(<span class="st">"cp_mt_ML"</span> <span class="ot">=</span> cp_mt_ML_dendro)</span>
<span id="cb7-19"><a href="#cb7-19"></a>)</span>
<span id="cb7-20"><a href="#cb7-20"></a></span>
<span id="cb7-21"><a href="#cb7-21"></a><span class="co"># populate combinatorial trees</span></span>
<span id="cb7-22"><a href="#cb7-22"></a><span class="cf">for</span>(k <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>){</span>
<span id="cb7-23"><a href="#cb7-23"></a>  </span>
<span id="cb7-24"><a href="#cb7-24"></a>  <span class="co"># get file list</span></span>
<span id="cb7-25"><a href="#cb7-25"></a>  tree_list_all <span class="ot">=</span> <span class="fu">list.files</span>(<span class="at">path =</span> <span class="fu">paste</span>(mainDir, <span class="st">"/09_combinatorial_phylogeny/01_cp_mt_combinatorial/cp_mt_conc_"</span>, k, <span class="st">"_genes"</span>, <span class="at">sep =</span> <span class="st">""</span>),</span>
<span id="cb7-26"><a href="#cb7-26"></a>                             <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">.contree$"</span>)</span>
<span id="cb7-27"><a href="#cb7-27"></a>  <span class="co"># keep only trees with both cp and mt markers</span></span>
<span id="cb7-28"><a href="#cb7-28"></a>  tree_list_fltr <span class="ot">=</span> <span class="fu">c</span>()</span>
<span id="cb7-29"><a href="#cb7-29"></a>  <span class="cf">for</span>(tree <span class="cf">in</span> tree_list_all){</span>
<span id="cb7-30"><a href="#cb7-30"></a>    cp <span class="ot">=</span> <span class="cn">FALSE</span></span>
<span id="cb7-31"><a href="#cb7-31"></a>    mt <span class="ot">=</span> <span class="cn">FALSE</span></span>
<span id="cb7-32"><a href="#cb7-32"></a>    <span class="co"># check if cp is present</span></span>
<span id="cb7-33"><a href="#cb7-33"></a>    <span class="cf">for</span>(gene <span class="cf">in</span> cp_genes){</span>
<span id="cb7-34"><a href="#cb7-34"></a>      <span class="cf">if</span>(<span class="fu">grepl</span>(gene, tree, <span class="at">fixed =</span> <span class="cn">TRUE</span>)){</span>
<span id="cb7-35"><a href="#cb7-35"></a>        cp <span class="ot">=</span> <span class="cn">TRUE</span></span>
<span id="cb7-36"><a href="#cb7-36"></a>      }</span>
<span id="cb7-37"><a href="#cb7-37"></a>    }</span>
<span id="cb7-38"><a href="#cb7-38"></a>    <span class="co"># check if mt is present</span></span>
<span id="cb7-39"><a href="#cb7-39"></a>    <span class="cf">for</span>(gene <span class="cf">in</span> mt_genes){</span>
<span id="cb7-40"><a href="#cb7-40"></a>      <span class="cf">if</span>(<span class="fu">grepl</span>(gene, tree, <span class="at">fixed =</span> <span class="cn">TRUE</span>)){</span>
<span id="cb7-41"><a href="#cb7-41"></a>        mt <span class="ot">=</span> <span class="cn">TRUE</span></span>
<span id="cb7-42"><a href="#cb7-42"></a>      }</span>
<span id="cb7-43"><a href="#cb7-43"></a>    }</span>
<span id="cb7-44"><a href="#cb7-44"></a>    <span class="cf">if</span>(cp <span class="sc">==</span> <span class="cn">TRUE</span> <span class="sc">&amp;</span> mt <span class="sc">==</span> <span class="cn">TRUE</span>){</span>
<span id="cb7-45"><a href="#cb7-45"></a>      tree_list_fltr <span class="ot">=</span> <span class="fu">c</span>(tree_list_fltr, tree)</span>
<span id="cb7-46"><a href="#cb7-46"></a>    }</span>
<span id="cb7-47"><a href="#cb7-47"></a>  }</span>
<span id="cb7-48"><a href="#cb7-48"></a>  </span>
<span id="cb7-49"><a href="#cb7-49"></a>  <span class="co"># import trees</span></span>
<span id="cb7-50"><a href="#cb7-50"></a>  <span class="cf">for</span>(tree <span class="cf">in</span> tree_list_fltr){</span>
<span id="cb7-51"><a href="#cb7-51"></a>    combinatorial_trees[[<span class="fu">paste</span>(<span class="st">"cp_mt_conc_"</span>, k, <span class="st">"_genes"</span>, <span class="at">sep =</span> <span class="st">""</span>)]][[<span class="fu">length</span>(combinatorial_trees[[<span class="fu">paste</span>(<span class="st">"cp_mt_conc_"</span>, k, <span class="st">"_genes"</span>, <span class="at">sep =</span> <span class="st">""</span>)]]) <span class="sc">+</span> <span class="dv">1</span>]] <span class="ot">=</span> ape<span class="sc">::</span><span class="fu">read.tree</span>(<span class="at">file =</span> <span class="fu">paste</span>(mainDir, <span class="st">"/09_combinatorial_phylogeny/01_cp_mt_combinatorial/cp_mt_conc_"</span>, k, <span class="st">"_genes/"</span>, tree, <span class="at">sep =</span> <span class="st">""</span>))</span>
<span id="cb7-52"><a href="#cb7-52"></a>    <span class="fu">names</span>(combinatorial_trees[[<span class="fu">paste</span>(<span class="st">"cp_mt_conc_"</span>, k, <span class="st">"_genes"</span>, <span class="at">sep =</span> <span class="st">""</span>)]])[[<span class="fu">length</span>(combinatorial_trees[[<span class="fu">paste</span>(<span class="st">"cp_mt_conc_"</span>, k, <span class="st">"_genes"</span>, <span class="at">sep =</span> <span class="st">""</span>)]])]] <span class="ot">=</span> stringr<span class="sc">::</span><span class="fu">str_remove</span>(tree, <span class="st">".aln.contree"</span>)</span>
<span id="cb7-53"><a href="#cb7-53"></a>  }</span>
<span id="cb7-54"><a href="#cb7-54"></a>  </span>
<span id="cb7-55"><a href="#cb7-55"></a>  <span class="co"># clean</span></span>
<span id="cb7-56"><a href="#cb7-56"></a>  <span class="fu">rm</span>(cp, gene, k, mt, tree, tree_list_all, tree_list_fltr)</span>
<span id="cb7-57"><a href="#cb7-57"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next step: we calculate the distances between the combinatorial trees and the reconstructed <em>Ulva</em> species tree.</p>
<div class="cell" data-result="hide">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="co"># create distance lists</span></span>
<span id="cb8-2"><a href="#cb8-2"></a>combinatorial_distances <span class="ot">=</span> <span class="fu">list</span>(</span>
<span id="cb8-3"><a href="#cb8-3"></a>  <span class="st">"cp_mt_conc_2_genes"</span> <span class="ot">=</span> <span class="cn">NULL</span>,</span>
<span id="cb8-4"><a href="#cb8-4"></a>  <span class="st">"cp_mt_conc_3_genes"</span> <span class="ot">=</span> <span class="cn">NULL</span>,</span>
<span id="cb8-5"><a href="#cb8-5"></a>  <span class="st">"cp_mt_conc_4_genes"</span> <span class="ot">=</span> <span class="cn">NULL</span>,</span>
<span id="cb8-6"><a href="#cb8-6"></a>  <span class="st">"cp_mt_conc_5_genes"</span> <span class="ot">=</span> <span class="cn">NULL</span>,</span>
<span id="cb8-7"><a href="#cb8-7"></a>  <span class="st">"cp_mt_conc_6_genes"</span> <span class="ot">=</span> <span class="cn">NULL</span>,</span>
<span id="cb8-8"><a href="#cb8-8"></a>  <span class="st">"cp_mt_conc_7_genes"</span> <span class="ot">=</span> <span class="cn">NULL</span>,</span>
<span id="cb8-9"><a href="#cb8-9"></a>  <span class="st">"cp_mt_conc_8_genes"</span> <span class="ot">=</span> <span class="cn">NULL</span>,</span>
<span id="cb8-10"><a href="#cb8-10"></a>  <span class="st">"cp_mt_conc_9_genes"</span> <span class="ot">=</span> <span class="cn">NULL</span>,</span>
<span id="cb8-11"><a href="#cb8-11"></a>  <span class="st">"cp_mt_conc_10_genes"</span> <span class="ot">=</span> <span class="cn">NULL</span></span>
<span id="cb8-12"><a href="#cb8-12"></a>)</span>
<span id="cb8-13"><a href="#cb8-13"></a></span>
<span id="cb8-14"><a href="#cb8-14"></a><span class="co"># populate distance lists</span></span>
<span id="cb8-15"><a href="#cb8-15"></a>combinatorial_list <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">25</span>, <span class="dv">97</span>, <span class="dv">195</span>, <span class="dv">247</span>, <span class="dv">210</span>, <span class="dv">121</span>, <span class="dv">46</span>, <span class="dv">11</span>, <span class="dv">2</span>)</span>
<span id="cb8-16"><a href="#cb8-16"></a><span class="cf">for</span>(k <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(combinatorial_distances)){</span>
<span id="cb8-17"><a href="#cb8-17"></a>  </span>
<span id="cb8-18"><a href="#cb8-18"></a>  <span class="co"># get size of the results matrix</span></span>
<span id="cb8-19"><a href="#cb8-19"></a>  empty_matrix <span class="ot">=</span> <span class="fu">matrix</span>(<span class="at">nrow =</span> combinatorial_list[[k]], <span class="at">ncol =</span> combinatorial_list[[k]])</span>
<span id="cb8-20"><a href="#cb8-20"></a>  <span class="fu">colnames</span>(empty_matrix) <span class="ot">=</span> <span class="fu">names</span>(combinatorial_trees[[k]])</span>
<span id="cb8-21"><a href="#cb8-21"></a>  <span class="fu">rownames</span>(empty_matrix) <span class="ot">=</span> <span class="fu">names</span>(combinatorial_trees[[k]])</span>
<span id="cb8-22"><a href="#cb8-22"></a>  </span>
<span id="cb8-23"><a href="#cb8-23"></a>  <span class="co"># populate</span></span>
<span id="cb8-24"><a href="#cb8-24"></a>  combinatorial_distances[[k]] <span class="ot">=</span> <span class="fu">list</span>(</span>
<span id="cb8-25"><a href="#cb8-25"></a>    <span class="st">"common_nodes_corr"</span> <span class="ot">=</span> empty_matrix,</span>
<span id="cb8-26"><a href="#cb8-26"></a>    <span class="st">"RF"</span> <span class="ot">=</span> empty_matrix,</span>
<span id="cb8-27"><a href="#cb8-27"></a>    <span class="st">"KF"</span> <span class="ot">=</span> empty_matrix,</span>
<span id="cb8-28"><a href="#cb8-28"></a>    <span class="st">"MAST_%"</span> <span class="ot">=</span> empty_matrix</span>
<span id="cb8-29"><a href="#cb8-29"></a>  )</span>
<span id="cb8-30"><a href="#cb8-30"></a>  </span>
<span id="cb8-31"><a href="#cb8-31"></a>  <span class="co"># clean</span></span>
<span id="cb8-32"><a href="#cb8-32"></a>  <span class="fu">rm</span>(empty_matrix)</span>
<span id="cb8-33"><a href="#cb8-33"></a>}</span>
<span id="cb8-34"><a href="#cb8-34"></a></span>
<span id="cb8-35"><a href="#cb8-35"></a><span class="co"># get pairwise distances</span></span>
<span id="cb8-36"><a href="#cb8-36"></a><span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(combinatorial_trees)){</span>
<span id="cb8-37"><a href="#cb8-37"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(combinatorial_trees[[j]])){</span>
<span id="cb8-38"><a href="#cb8-38"></a>    <span class="cf">for</span>(k <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(combinatorial_trees[[j]])){</span>
<span id="cb8-39"><a href="#cb8-39"></a>      </span>
<span id="cb8-40"><a href="#cb8-40"></a>      <span class="co"># get tree</span></span>
<span id="cb8-41"><a href="#cb8-41"></a>      tree1_tmp <span class="ot">=</span> combinatorial_trees[[j]][[i]]</span>
<span id="cb8-42"><a href="#cb8-42"></a>      tree2_tmp <span class="ot">=</span> combinatorial_trees[[j]][[k]]</span>
<span id="cb8-43"><a href="#cb8-43"></a></span>
<span id="cb8-44"><a href="#cb8-44"></a>      <span class="co"># list of common species</span></span>
<span id="cb8-45"><a href="#cb8-45"></a>      species_list <span class="ot">=</span> tree1_tmp<span class="sc">$</span>tip.label[<span class="fu">which</span>(tree1_tmp<span class="sc">$</span>tip.label <span class="sc">%in%</span> tree2_tmp<span class="sc">$</span>tip.label)]</span>
<span id="cb8-46"><a href="#cb8-46"></a></span>
<span id="cb8-47"><a href="#cb8-47"></a>      <span class="co"># get overlapping species</span></span>
<span id="cb8-48"><a href="#cb8-48"></a>      tree1 <span class="ot">=</span> ape<span class="sc">::</span><span class="fu">keep.tip</span>(tree1_tmp, species_list)</span>
<span id="cb8-49"><a href="#cb8-49"></a>      tree2 <span class="ot">=</span> ape<span class="sc">::</span><span class="fu">keep.tip</span>(tree2_tmp, species_list)</span>
<span id="cb8-50"><a href="#cb8-50"></a>      </span>
<span id="cb8-51"><a href="#cb8-51"></a>      <span class="co"># get distances</span></span>
<span id="cb8-52"><a href="#cb8-52"></a>      distances <span class="ot">=</span> phangorn<span class="sc">::</span><span class="fu">treedist</span>(tree1, tree2)</span>
<span id="cb8-53"><a href="#cb8-53"></a>      combinatorial_distances[[j]][[<span class="st">"RF"</span>]][i, k] <span class="ot">=</span> distances[[<span class="dv">1</span>]]</span>
<span id="cb8-54"><a href="#cb8-54"></a>      combinatorial_distances[[j]][[<span class="st">"KF"</span>]][i, k] <span class="ot">=</span> distances[[<span class="dv">2</span>]]</span>
<span id="cb8-55"><a href="#cb8-55"></a>      combinatorial_distances[[j]][[<span class="st">"MAST_%"</span>]][i, k] <span class="ot">=</span> <span class="fu">length</span>(phangorn<span class="sc">::</span><span class="fu">mast</span>(tree1, tree2, <span class="at">tree =</span> <span class="cn">FALSE</span>)) <span class="sc">/</span> <span class="fu">length</span>(species_list)</span>
<span id="cb8-56"><a href="#cb8-56"></a>      </span>
<span id="cb8-57"><a href="#cb8-57"></a>      <span class="co"># clean</span></span>
<span id="cb8-58"><a href="#cb8-58"></a>      <span class="fu">rm</span>(tree1, tree2, tree1_tmp, tree2_tmp, distances)</span>
<span id="cb8-59"><a href="#cb8-59"></a>    }</span>
<span id="cb8-60"><a href="#cb8-60"></a>  }</span>
<span id="cb8-61"><a href="#cb8-61"></a>}</span>
<span id="cb8-62"><a href="#cb8-62"></a></span>
<span id="cb8-63"><a href="#cb8-63"></a></span>
<span id="cb8-64"><a href="#cb8-64"></a><span class="do">### get pairwise common_nodes_corr distances</span></span>
<span id="cb8-65"><a href="#cb8-65"></a></span>
<span id="cb8-66"><a href="#cb8-66"></a><span class="co"># iterate all combinatorial trees</span></span>
<span id="cb8-67"><a href="#cb8-67"></a><span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(combinatorial_trees)){</span>
<span id="cb8-68"><a href="#cb8-68"></a>  </span>
<span id="cb8-69"><a href="#cb8-69"></a>  <span class="co"># create CPU cluster</span></span>
<span id="cb8-70"><a href="#cb8-70"></a>  cl <span class="ot">=</span> parallel<span class="sc">::</span><span class="fu">makeCluster</span>(<span class="dv">8</span>, <span class="at">type =</span> <span class="st">"SOCK"</span>)</span>
<span id="cb8-71"><a href="#cb8-71"></a>  doSNOW<span class="sc">::</span><span class="fu">registerDoSNOW</span>(cl)</span>
<span id="cb8-72"><a href="#cb8-72"></a>  </span>
<span id="cb8-73"><a href="#cb8-73"></a>  <span class="co"># get chronos in parallel</span></span>
<span id="cb8-74"><a href="#cb8-74"></a>  chronos_list <span class="ot">=</span> <span class="fu">foreach</span>(<span class="at">k =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(combinatorial_trees[[j]]), <span class="at">.combine =</span> <span class="st">"c"</span>) <span class="sc">%dopar%</span> {</span>
<span id="cb8-75"><a href="#cb8-75"></a>    <span class="fu">library</span>(<span class="st">"ape"</span>)</span>
<span id="cb8-76"><a href="#cb8-76"></a>    <span class="co"># get tree</span></span>
<span id="cb8-77"><a href="#cb8-77"></a>    <span class="fu">tryCatch</span>(<span class="fu">chronos</span>(combinatorial_trees[[j]][[k]]), <span class="at">error =</span> <span class="cf">function</span>(e) { <span class="cn">NA</span> })</span>
<span id="cb8-78"><a href="#cb8-78"></a>  }  </span>
<span id="cb8-79"><a href="#cb8-79"></a>  parallel<span class="sc">::</span><span class="fu">stopCluster</span>(cl)</span>
<span id="cb8-80"><a href="#cb8-80"></a>  <span class="fu">names</span>(chronos_list) <span class="ot">=</span> <span class="fu">names</span>(combinatorial_trees[[j]])</span>
<span id="cb8-81"><a href="#cb8-81"></a>  </span>
<span id="cb8-82"><a href="#cb8-82"></a>  <span class="co"># get pairwise common_nodes_corr distances</span></span>
<span id="cb8-83"><a href="#cb8-83"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(combinatorial_trees[[j]])){</span>
<span id="cb8-84"><a href="#cb8-84"></a>    <span class="cf">for</span>(k <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(combinatorial_trees[[j]])){</span>
<span id="cb8-85"><a href="#cb8-85"></a>      </span>
<span id="cb8-86"><a href="#cb8-86"></a>      <span class="fu">print</span>(<span class="fu">paste</span>(j, i, k, <span class="at">sep =</span> <span class="st">","</span>))</span>
<span id="cb8-87"><a href="#cb8-87"></a>      </span>
<span id="cb8-88"><a href="#cb8-88"></a>      <span class="co"># get trees</span></span>
<span id="cb8-89"><a href="#cb8-89"></a>      tree1_tmp <span class="ot">=</span> chronos_list[[<span class="fu">names</span>(combinatorial_trees[[j]])[i]]]</span>
<span id="cb8-90"><a href="#cb8-90"></a>      tree2_tmp <span class="ot">=</span> chronos_list[[<span class="fu">names</span>(combinatorial_trees[[j]])[k]]]</span>
<span id="cb8-91"><a href="#cb8-91"></a>      </span>
<span id="cb8-92"><a href="#cb8-92"></a>      <span class="co"># check trees and get distance</span></span>
<span id="cb8-93"><a href="#cb8-93"></a>      <span class="cf">if</span>(<span class="fu">all</span>(<span class="sc">!</span><span class="fu">is.na</span>(tree1_tmp)) <span class="sc">&amp;</span> <span class="fu">all</span>(<span class="sc">!</span><span class="fu">is.na</span>(tree2_tmp))){</span>
<span id="cb8-94"><a href="#cb8-94"></a>        combinatorial_distances[[j]][[<span class="st">"common_nodes_corr"</span>]][i, k] <span class="ot">=</span> <span class="fu">tryCatch</span>(</span>
<span id="cb8-95"><a href="#cb8-95"></a>          <span class="fu">cor.dendlist</span>(<span class="fu">dendlist</span>(<span class="fu">as.dendrogram</span>(ape<span class="sc">::</span><span class="fu">root</span>(tree1_tmp, <span class="at">outgroup =</span> <span class="fu">c</span>(<span class="st">"Oviri"</span>, <span class="st">"Pakin"</span>), <span class="at">resolve.root =</span> <span class="cn">TRUE</span>)),</span>
<span id="cb8-96"><a href="#cb8-96"></a>                                <span class="fu">as.dendrogram</span>(ape<span class="sc">::</span><span class="fu">root</span>(tree2_tmp, <span class="at">outgroup =</span> <span class="fu">c</span>(<span class="st">"Oviri"</span>, <span class="st">"Pakin"</span>), <span class="at">resolve.root =</span> <span class="cn">TRUE</span>))),</span>
<span id="cb8-97"><a href="#cb8-97"></a>                       <span class="at">method =</span> <span class="st">"common_nodes"</span>)[<span class="dv">2</span>],</span>
<span id="cb8-98"><a href="#cb8-98"></a>          <span class="at">error =</span> <span class="cf">function</span>(e) { <span class="cn">NA</span> })</span>
<span id="cb8-99"><a href="#cb8-99"></a>      } <span class="cf">else</span> {</span>
<span id="cb8-100"><a href="#cb8-100"></a>        combinatorial_distances[[j]][[<span class="st">"common_nodes_corr"</span>]][i, k] <span class="ot">=</span> <span class="cn">NA</span></span>
<span id="cb8-101"><a href="#cb8-101"></a>      }</span>
<span id="cb8-102"><a href="#cb8-102"></a>    }</span>
<span id="cb8-103"><a href="#cb8-103"></a>  }</span>
<span id="cb8-104"><a href="#cb8-104"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For each combination (i.e.: 2-genes combination, 3-genes combination, etc.), a representative combination with the highest the MAST% and the common node correlation are shown in the table below. The best combination to use for the development of <em>Ulva</em> organellar universal markers should be the combination with the highest values of the two indices and the lowest number of markers in the combination.</p>
<p>The combination with five to eight markers have the highest value for the MAST% and the highest value for the common node correlation, but given the number of markers, it might pose more challenges for the development process later. The combination with four markers, <em>psbD</em>, <em>psaA</em>, <em>cox1</em>, <em>cox2</em>, which also has the highest value for the MAST% and has the value for the common node correlation slightly lower than the highest value possible, however, we believe that approximates faithfully the reconstructed <em>Ulva</em> species tree and can be used to develop universal primers for easy and cost-efficient <em>Ulva</em> species classification of field and herbarium specimens.</p>
<div class="cell">
<div class="cell-output-display">
<div class="datatables html-widget html-fill-item" id="htmlwidget-562e9d800fed88ac0695" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-562e9d800fed88ac0695">{"x":{"filter":"none","vertical":false,"extensions":["FixedColumns","FixedHeader"],"caption":"<caption>Table 2: best combinatorial tree for each gene set<\/caption>","data":[["2","3","4","5","6","7","8","9","10"],["psaA_cox1","rps2_psaA_cox1","psbD_psaA_cox1_cox2","psbD_rps2_psaA_cox1_cox2","psbB_psbD_rps2_psaA_cox1_cox2","psaB_psbB_psbD_rps2_psaA_cox1_cox2","petB_psbB_psbD_rps2_psaA_atp6_cox1_cox2","petB_psaB_psbB_psbD_rps2_psaA_atp6_cox1_rps3","petB_psaB_psbB_psbD_rps2_psaA_atp6_cox1_cox2_rps3"],["0.824","0.882","0.941","0.941","0.941","0.941","0.941","0.922","0.882"],["1","6","1","6","10","8","2","2","1"],["0.901","0.931","0.95","0.96","0.96","0.96","0.96","0.95","0.931"],["1  psaA_cox2","1 psbB_cox1_cox2","1","1 psaB_psbB_rps2_cox1_cox2","1 petB_psaB_psbB_rps2_cox1_cox2","3 petB_psbD_rps2_psaA_atp6_cox1_cox2","1","2","1"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th># genes<\/th>\n      <th>representative best genes markers<\/th>\n      <th>MAST %<\/th>\n      <th># of equivalent combinations<\/th>\n      <th>Common node correlation<\/th>\n      <th># of equivalent combinations<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":false,"pageLength":10,"columnDefs":[{"name":"# genes","targets":0},{"name":"representative best genes markers","targets":1},{"name":"MAST %","targets":2},{"name":"# of equivalent combinations","targets":3},{"name":"Common node correlation","targets":4},{"name":"# of equivalent combinations","targets":5}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
</section>
<section id="sec-lesson-learnt-markers-selection" class="level2" data-number="3.4">
<h2 data-number="3.4" class="anchored" data-anchor-id="sec-lesson-learnt-markers-selection"><span class="header-section-number">3.4</span> Lessons Learnt</h2>
<p>So far, we have learnt:</p>
<ul>
<li>chloroplast genes <em>petB</em>, <em>psaA</em>, <em>psaB</em>, <em>psbB</em>, <em>psbD</em>, <em>rps2</em> and mitochondrial genes <em>atp6</em>, <em>cox1</em>, <em>cox2</em>, and <em>rps3</em> are among the ones with the single-gene ML tree closest to the reconstructed <em>Ulva</em> species tree;</li>
<li>interestingly, commonly used markers (i.e.: <em>rbcL</em> and <em>tufA</em>) had a single-gene ML tree more distant to the reconstructed <em>Ulva</em> species tree than the other markers we have selected;</li>
<li>out of the 945 2:10 gene combinations that we have tested, the 4-genes combination <em>psbD</em>-<em>psaA</em>-<em>cox1</em>-<em>cox2</em> provides a ML phylogenetic tree that faithfully approximate the reconstructed <em>Ulva</em> species tree and could be used for the development of universal <em>Ulva</em> species markers.</li>
</ul>
</section>
<section id="session-information" class="level2" data-number="3.5">
<h2 data-number="3.5" class="anchored" data-anchor-id="session-information"><span class="header-section-number">3.5</span> Session Information</h2>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.2 (2023-10-31)
Platform: x86_64-conda-linux-gnu (64-bit)
Running under: openSUSE Tumbleweed

Matrix products: default
BLAS/LAPACK: /home/andrea/miniforge3/envs/moai/lib/libmkl_rt.so.2;  LAPACK version 3.9.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=it_IT.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Europe/Brussels
tzcode source: system (glibc)

attached base packages:
 [1] parallel  stats4    grid      stats     graphics  grDevices utils    
 [8] datasets  methods   base     

other attached packages:
 [1] treeio_1.26.0         TreeDist_2.9.2        stringr_1.5.1        
 [4] scales_1.3.0          RColorBrewer_1.1-3    reshape_0.8.9        
 [7] phytools_2.4-4        maps_3.4.2.1          phylogram_2.1.0      
[10] phangorn_2.12.1       gridExtra_2.3         ggtree_3.10.1        
[13] ggplot2_3.5.1         ggdist_3.3.2          doSNOW_1.0.20        
[16] snow_0.4-4            iterators_1.0.14      foreach_1.5.2        
[19] dendextend_1.19.0     DECIPHER_2.30.0       RSQLite_2.3.9        
[22] Biostrings_2.70.3     GenomeInfoDb_1.38.8   XVector_0.42.0       
[25] IRanges_2.36.0        S4Vectors_0.40.2      BiocGenerics_0.48.1  
[28] corrplot_0.95         ComplexHeatmap_2.18.0 circlize_0.4.16      
[31] ape_5.8-1            

loaded via a namespace (and not attached):
  [1] jsonlite_1.8.9          shape_1.4.6.1           magrittr_2.0.3         
  [4] magick_2.8.5            farver_2.1.2            rmarkdown_2.29         
  [7] GlobalOptions_0.1.2     fs_1.6.5                zlibbioc_1.48.2        
 [10] vctrs_0.6.5             memoise_2.0.1           RCurl_1.98-1.16        
 [13] htmltools_0.5.8.1       distributional_0.5.0    DEoptim_2.2-8          
 [16] gridGraphics_0.5-1      sass_0.4.9              bslib_0.8.0            
 [19] htmlwidgets_1.6.4       plyr_1.8.9              cachem_1.1.0           
 [22] igraph_2.1.4            mime_0.12               lifecycle_1.0.4        
 [25] pkgconfig_2.0.3         Matrix_1.6-5            R6_2.5.1               
 [28] fastmap_1.2.0           GenomeInfoDbData_1.2.11 rbibutils_2.3          
 [31] shiny_1.10.0            clue_0.3-66             digest_0.6.37          
 [34] numDeriv_2016.8-1.1     aplot_0.2.4             colorspace_2.1-1       
 [37] patchwork_1.3.0         crosstalk_1.2.1         clusterGeneration_1.3.8
 [40] compiler_4.3.2          bit64_4.6.0-1           withr_3.0.2            
 [43] doParallel_1.0.17       optimParallel_1.0-2     viridis_0.6.5          
 [46] DBI_1.2.3               R.utils_2.12.3          MASS_7.3-60.0.1        
 [49] rjson_0.2.23            scatterplot3d_0.3-44    tools_4.3.2            
 [52] httpuv_1.6.15           TreeTools_1.13.0        R.oo_1.27.0            
 [55] glue_1.8.0              quadprog_1.5-8          nlme_3.1-167           
 [58] R.cache_0.16.0          promises_1.3.2          cluster_2.1.8          
 [61] PlotTools_0.3.1         generics_0.1.3          gtable_0.3.6           
 [64] R.methodsS3_1.8.2       tidyr_1.3.1             pillar_1.10.1          
 [67] yulab.utils_0.2.0       later_1.4.1             dplyr_1.1.4            
 [70] lattice_0.22-6          bit_4.5.0.1             tidyselect_1.2.1       
 [73] knitr_1.49              xfun_0.50               expm_1.0-0             
 [76] matrixStats_1.5.0       DT_0.33                 stringi_1.8.4          
 [79] lazyeval_0.2.2          ggfun_0.1.8             yaml_2.3.10            
 [82] evaluate_1.0.3          codetools_0.2-20        tibble_3.2.1           
 [85] ggplotify_0.1.2         cli_3.6.3               xtable_1.8-4           
 [88] Rdpack_2.6.2            jquerylib_0.1.4         munsell_0.5.1          
 [91] Rcpp_1.0.14             coda_0.19-4.1           png_0.1-8              
 [94] blob_1.2.4              bitops_1.0-9            viridisLite_0.4.2      
 [97] tidytree_0.4.6          purrr_1.0.2             crayon_1.5.3           
[100] combinat_0.0-8          GetoptLong_1.0.5        rlang_1.1.5            
[103] fastmatch_1.1-6         mnormt_2.1.1            shinyjs_2.1.0          </code></pre>
</div>
</div>
</div>
</div>
</div>


</section>

</main> <!-- /main -->
<script data-goatcounter="https://19adc99.goatcounter.com/count" async="" src="//gc.zgo.at/count.js"></script>
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./cp-and-mt-phylogeny-reconstruction.html" class="pagination-link" aria-label="Species tree reconstruction from organellar genomes">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Species tree reconstruction from organellar genomes</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      <ul class="footer-items list-unstyled">
    <li class="nav-item">
 Copyright © 2025 Andrea Del Cortona.
  </li>  
</ul>
    </div>   
    <div class="nav-footer-center">
      <ul class="footer-items list-unstyled">
    <li class="nav-item">
    <a class="nav-link" href="https://quarto.org/">
<p>This page is built with ❤️ and <em>Quarto</em>.</p>
</a>
  </li>  
</ul>
    </div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/19ADC99">
      <i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/andrea-del-cortona-86a75474/">
      <i class="bi bi-linkedin" role="img" aria-label="Lindekin">
</i> 
    </a>
  </li>  
</ul>
    </div>
  </div>
</footer>




</body></html>